<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>systemPipe: Workflow and Visualization Toolkit • systemPipeWorkshop2021</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="systemPipe: Workflow and Visualization Toolkit">
<meta property="og:description" content="systemPipeWorkshop2021">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-158796879-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-158796879-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeWorkshop2021</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/systemPipeWorkshop2021.html">Workshop</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Slides
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="http://www.dcassol.com/bioc2021_slides/">Slides</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/spr_project.html">Project structure</a>
    </li>
    <li>
      <a href="../articles/cwl_syntax.html">CWL syntax</a>
    </li>
    <li>
      <a href="../articles/cwl_and_spr.html">SPR and CWL</a>
    </li>
    <li>
      <a href="../articles/spr_functions.html">SPR Function and Methods</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/systemPipeR/systemPipeWorkshop2021">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://hub.docker.com/r/systempipe/systempipeworkshop2021">
    <span class="fab fa-docker"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="systemPipeWorkshop2021_files/header-attrs-2.9/header-attrs.js"></script><script src="systemPipeWorkshop2021_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><link href="systemPipeWorkshop2021_files/plotwf-2.1.2/plotwf.css" rel="stylesheet">
<script src="systemPipeWorkshop2021_files/plotwf-2.1.2/viz.js"></script><script src="systemPipeWorkshop2021_files/plotwf-2.1.2/full.render.js"></script><script src="systemPipeWorkshop2021_files/plotwf-binding-1.27.21/plotwf.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>systemPipe: Workflow and Visualization Toolkit</h1>
                        <h4 class="author"></h4>
            
            <h4 class="date">Last update: 06 August, 2021</h4>
      
      
      <div class="hidden name"><code>systemPipeWorkshop2021.Rmd</code></div>

    </div>

    
    
<script>
$(()=>{
  $('a:contains("{blk}")').each(function(){
    $(this)
      .text($(this).text().replace("{blk}", ""))
      .attr("target", "_blank");
  });
});
</script><p><strong>Authors</strong>: Daniela Cassol (<a href="mailto:danielac@ucr.edu" class="email">danielac@ucr.edu</a>), Le Zhang (<a href="mailto:le.zhang001@email.ucr.edu" class="email">le.zhang001@email.ucr.edu</a>), Thomas Girke (<a href="mailto:thomas.girke@ucr.edu" class="email">thomas.girke@ucr.edu</a>).</p>
<p><strong>Institution</strong>: Institute for Integrative Genome Biology, University of California, Riverside, California, USA.</p>
<div id="overview-of-systempipe-toolkit" class="section level1">
<h1 class="hasAnchor">
<a href="#overview-of-systempipe-toolkit" class="anchor"></a>Overview of <em>systemPipe</em> toolkit</h1>
<p><code>systemPipe</code> (SP) is a generic toolkit for designing and running reproducible data analysis workflows. The environment consists of three major modules implemented as R/Bioconductor packages:</p>
<ul>
<li><p><code>systemPipeR</code> (SPR) provides core functionalities for defining, designing and running workflows, interacting with both R and/or command-line software, as well as generating publication-quality analysis reports.</p></li>
<li><p><code>systemPipeShiny</code> (SPS) integrates a graphical user interface for managing workflows and visualizing results interactively.</p></li>
<li><p><code>systemPipeWorkflow</code> (SPW) offers a collection of pre-configured workflow templates.</p></li>
</ul>
<p>For more details, please see the <a href="http://www.dcassol.com/bioc2021_slides/#1">presentation slides</a>.</p>
</div>
<div id="introduction-to-sprs-command-line-interface" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-to-sprs-command-line-interface" class="anchor"></a>Introduction to SPR’s command-line interface</h1>
<p>A central concept for designing workflows within the <a href="http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html"><code>systemPipeR</code></a> <span class="citation">(H Backman and Girke 2016)</span> environment is the use of workflow management containers (see Figure 1).</p>
<p><code>systemPipeR</code> adopted the widely used community standard <a href="https://www.commonwl.org/">Common Workflow Language</a> (CWL) <span class="citation">(Amstutz et al. 2016)</span> for describing analysis workflows in a generic and reproducible manner.</p>
<p>Using this community standard in <code>systemPipeR</code> has many advantages. For instance, <code>systemPipeR's</code> CWL interface is that it provides two options to run command-line tools and workflows based on CWL. First, one can run CWL in its native way via an R-based wrapper utility for <code>cwl-runner</code> or <code>cwl-tools</code> (CWL-based approach). Second, one can run workflows using CWL’s command-line and workflow instructions from within R (R-based approach). In the latter case the same CWL workflow definition files (e.g. <em>.cwl</em> and <em>.yml</em>) are used but rendered and executed entirely with R functions defined by <code>systemPipeR</code>, and thus use CWL mainly as a command-line and workflow definition format rather than software to run workflows. Moreover, <code>systemPipeR</code> provides several convenience functions that are useful for designing and debugging workflows, such as a command-line rendering function to retrieve the exact command-line strings for each step prior to running a command-line.</p>
<p><code>systemPipeR</code> also includes support for both command-line and R/Bioconductor software as well as resources for containerization, parallel evaluations on computer clusters along with automated generation of interactive analysis reports.</p>
<div class="figure" style="text-align: center">
<img src="https://raw.githubusercontent.com/systemPipeR/systemPipeWorkshop2021/master/inst/images/utilities.png" alt="Figure 1. Relevant features in `systemPipeR`. Workflow design concepts are illustrated under A. Examples of *systemPipeR's* visualization functionalities are given under B." width="100%"><p class="caption">
Figure 1. Relevant features in <code>systemPipeR</code>. Workflow design concepts are illustrated under A. Examples of <em>systemPipeR’s</em> visualization functionalities are given under B.
</p>
</div>
<p>The connectivity among workflow steps is defined by the <code>SYSargsList</code> workflow management class. This S4 class is a list-like container where each instance stores all the input/output paths and parameter components required for a particular data analysis step (see Figure 2).</p>
<p>The <code>SYSargsList</code> constructor function will generate the instances, using as data input initial targets files, as well as two-parameter files (for details, see below). When running preconfigured workflows, the only input the user needs to provide is the initial targets file containing the paths to the input files (e.g. FASTQ) along with unique sample labels. Subsequent targets instances are created automatically, based on the connectivity establish between the steps. The parameters required for running command-line software is provided by the parameter (<code>*.cwl</code> and <code>*.yml</code>) files described below.</p>
<p><code>SYSargsList</code> stores one or multiple steps. It defines and controls each step, allows status checking, as well as monitoring complex workflows from start to finish.</p>
<div class="figure" style="text-align: center">
<img src="https://raw.githubusercontent.com/systemPipeR/systemPipeWorkshop2021/master/inst/images/sysargslist.png" alt="Figure 2. Workflow steps with input/output file operations are controlled by `SYSargsList` instances. The only input provided by the user is the initial targets file, which can be optional. Subsequent targets instances are created automatically, from the previous output files. Any number of predefined or custom workflow steps are supported." width="100%"><p class="caption">
Figure 2. Workflow steps with input/output file operations are controlled by <code>SYSargsList</code> instances. The only input provided by the user is the initial targets file, which can be optional. Subsequent targets instances are created automatically, from the previous output files. Any number of predefined or custom workflow steps are supported.
</p>
</div>
</div>
<div id="workshop" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop" class="anchor"></a>Workshop</h1>
<div id="running-the-workshop" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-workshop" class="anchor"></a>Running the Workshop</h2>
<p>This workshop uses R <code>4.1.0</code> and Bioconductor version<code>3.14</code>. Bioconductor can be installed following <a href="https://www.bioconductor.org/developers/how-to/useDevel/">these instructions</a>.</p>
<p>During the <a href="https://bioc2021.bioconductor.org/">Bioc2021 conference</a>, the workshop can be run in the preconfigured <a href="http://app.orchestra.cancerdatasci.org/">Orchestra Platform</a>.</p>
</div>
<div id="workshop-setup-with-docker" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-setup-with-docker" class="anchor"></a>Workshop setup with Docker</h2>
<p>The Docker container used by this workshop runs with Bioconductor’s <code>development</code> version <code>3.14</code>. It includes all the necessary packages and software for running the code of the workshop. To use the Docker container, one needs to first install <a href="https://docs.docker.com/engine/install/">Docker</a> on a user’s system.</p>
<ul>
<li>The container can be downloaded and run with:</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-e</span> PASSWORD=systempipe <span class="at">-p</span> 8787:8787 systempipe/systempipeworkshop2021:latest</span></code></pre></div>
<ul>
<li><p>Log in to RStudio at <a href="http://localhost:8787">http://localhost:8787</a> using username <code>rstudio</code> and password <code>systempipe</code>.</p></li>
<li><p>If you prefer to run the workshop from the command-line, please use:</p></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-it</span> <span class="at">--user</span> rstudio systempipe/systempipeworkshop2021:latest bash </span></code></pre></div>
</div>
<div id="workshop-setup-with-github" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-setup-with-github" class="anchor"></a>Workshop setup with GitHub</h2>
<p><a href="http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html"><code>systemPipeR</code></a> and <a href="http://bioconductor.org/packages/devel/bioc/html/systemPipeShiny.html"><em><code>systemPipeShiny</code></em></a> environment can be installed from the R console using the <a href="https://cran.r-project.org/web/packages/BiocManager/index.html"><em><code>BiocManager::install</code></em></a> command. The associated data package <a href="http://www.bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html"><em><code>systemPipeRdata</code></em></a> can be installed the same way. The latter is a helper package for generating <code>systemPipeR</code> workflow environments with a single command containing all parameter files and sample data required to quickly test and run workflows.</p>
<p>To install all packages required for this workshop on a local system, one can use the following install commands.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Install workshop package</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"systemPipeR/systemPipeWorkshop2021"</span><span class="op">)</span>
<span class="co">## Install required packages</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"systemPipeR"</span>, <span class="st">"systemPipeRdata"</span>, <span class="st">"systemPipeShiny"</span><span class="op">)</span>, version<span class="op">=</span><span class="st">"3.14"</span><span class="op">)</span></code></pre></div>
<p>To access the workshop vignette:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/browseVignettes.html">browseVignettes</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"systemPipeWorkshop2021"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="getting-started" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting Started</h1>
<div id="loading-package-and-documentation" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-package-and-documentation" class="anchor"></a>Loading package and documentation</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://systempipe.org/">"systemPipeR"</a></span><span class="op">)</span> 
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://systempipe.org/sps">"systemPipeShiny"</a></span><span class="op">)</span> 
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/tgirke/systemPipeRdata">"systemPipeRdata"</a></span><span class="op">)</span> </code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span>help<span class="op">=</span><span class="st">"systemPipeR"</span><span class="op">)</span> <span class="co"># Lists package info</span>
<span class="fu"><a href="https://rdrr.io/r/utils/vignette.html">vignette</a></span><span class="op">(</span><span class="st">"systemPipeR"</span><span class="op">)</span> <span class="co"># Opens vignette</span></code></pre></div>
</div>
<div id="load-sample-data-and-directory-structure" class="section level2">
<h2 class="hasAnchor">
<a href="#load-sample-data-and-directory-structure" class="anchor"></a>Load sample data and directory structure</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">systemPipeRdata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span><span class="st">"rnaseq"</span>, mydirname <span class="op">=</span> <span class="st">"bioc2021"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Generated bioc2021 directory. Next run in rnaseq directory, the R code from *.Rmd template interactively. Alternatively, workflows can be exectued with a single command as instructed in the vignette."</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"bioc2021"</span><span class="op">)</span></code></pre></div>
<p>For more details about the project directory structure, please see <a href="https://systempipe.org/systemPipeWorkshop2021/articles/spr_project.html">here.</a></p>
</div>
<div id="how-to-find-help" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-find-help" class="anchor"></a>How to find help</h2>
<p>All questions about the package or any particular function should be posted to the Bioconductor support site <a href="https://support.bioconductor.org">https://support.bioconductor.org</a>.</p>
<p>Please add the “<code>systemPipeR</code>” tag to your question. This triggers an email alert that will be send to the authors.</p>
<p>We also appreciate receiving suggestions for improvements and/or bug reports by opening issues on <a href="https://github.com/systemPipeR">GitHub</a>.</p>
</div>
</div>
<div id="design-and-run-workflow---quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#design-and-run-workflow---quick-start" class="anchor"></a>Design and Run Workflow - Quick Start</h1>
<div id="importWF" class="section level2">
<h2 class="hasAnchor">
<a href="#importWF" class="anchor"></a>Showcase of toy RNA-Seq workflow</h2>
<p>If you desire to generate and run a RNA-seq workflow in a few (here highly simplified) steps, please follow:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span><span class="op">)</span> 
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/importWF.html">importWF</a></span><span class="op">(</span><span class="va">sal</span>, file_path <span class="op">=</span> <span class="st">"systemPipeRNAseq_importWF.Rmd"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html">runWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span>
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderLogs.html">renderLogs</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/plotWF.html">plotWF</a></span><span class="op">(</span><span class="va">sal</span>, rstudio <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="https://raw.githubusercontent.com/systemPipeR/systemPipeWorkshop2021/master/inst/images/plot.png" alt="Figure 3. *systemPipeR's* workflow structure, showing the dependency between steps and the status of each step." width="100%"><p class="caption">
Figure 3. <em>systemPipeR’s</em> workflow structure, showing the dependency between steps and the status of each step.
</p>
</div>
</div>
</div>
<div id="explore-workflow-design-and-run-parameters" class="section level1">
<h1 class="hasAnchor">
<a href="#explore-workflow-design-and-run-parameters" class="anchor"></a>Explore Workflow Design and Run Parameters</h1>
<div id="structure-of-parameters-files" class="section level2">
<h2 class="hasAnchor">
<a href="#structure-of-parameters-files" class="anchor"></a>Structure of parameters files</h2>
<p>The parameters required for running command-line software are provided by adopting the widely used CWL (<a href="https://www.commonwl.org/">Common Workflow Language</a>) community standard <span class="citation">(Amstutz et al. 2016)</span>. Parameter files are only required for command-line steps.</p>
<p>An overview of the CWL syntax is provided in the <a href="https://systempipe.org/systemPipeWorkshop2021/articles/cwl_syntax.html">article</a>, while this <a href="https://systempipe.org/systemPipeWorkshop2021/articles/cwl_and_spr.html">article</a> demonstrates how target files can be used for CWL-based workflow steps.</p>
<div id="automate-creation-of-cwl-parameters" class="section level3">
<h3 class="hasAnchor">
<a href="#automate-creation-of-cwl-parameters" class="anchor"></a>Automate creation of CWL parameters</h3>
<p>Users need to define the command-line in a pseudo-bash script format:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># "hisat2 -S ./results/M1A.sam -x ./data/tair10.fasta -k 1 -threads 4 -U ./data/SRR446027_1.fastq.gz "</span>
<span class="va">command</span> <span class="op">&lt;-</span> <span class="st">"
    hisat2 \
    -S &lt;F, out: ./results/M1A.sam&gt; \
    -x &lt;F: ./data/tair10.fasta&gt; \
     -k &lt;int: 1&gt; \
    -threads &lt;int: 4&gt; \
    -U &lt;F: ./data/SRR446027_1.fastq.gz&gt;
"</span></code></pre></div>
<div id="define-prefix-and-defaults" class="section level4">
<h4 class="hasAnchor">
<a href="#define-prefix-and-defaults" class="anchor"></a>Define prefix and defaults</h4>
<ul>
<li><p>Each line is an argument with its default value.</p></li>
<li><p>First line is the base command.</p></li>
<li><p>For argument lines (starting from the second line), any word before the first space leading <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code> or <code>--</code> will be treated as a prefix, <em>e.g.</em> <code>-S</code> or <code>--min</code>) . Any line without this first word will be treated as no prefix.</p></li>
<li><p>All defaults are placed inside <code>&lt;...&gt;</code>.</p></li>
<li><p>First argument is the input argument type. <code>F</code> for “File,” “int,” “string” are unchanged.</p></li>
<li><p>Optional: use the keyword <code>out</code> followed the type with a <code>,</code> comma separation to indicate if this argument is also a CWL output..</p></li>
<li><p>Then, use <code><a href="https://rdrr.io/r/base/Colon.html">:</a></code> to separate keywords and default values, any non-space value after the <code><a href="https://rdrr.io/r/base/Colon.html">:</a></code> will be treated as the default value.</p></li>
<li><p>If any argument has no default value, just a flag, like <code>--verbose</code>, there is no need to add any <code>&lt;...&gt;</code></p></li>
</ul>
</div>
<div id="createparam-function" class="section level4">
<h4 class="hasAnchor">
<a href="#createparam-function" class="anchor"></a><code>createParam</code> Function</h4>
<p><code>createParam</code> function requires the <code>string</code> defined above as an input.</p>
<p>First of all, the function will print the three components of the <code>cwl</code> file: - <code>BaseCommand</code>: Specifies the program to execute. - <code>Inputs</code>: Defines the input parameters of the process. - <code>Outputs</code>: Defines the parameters representing the output of the process.</p>
<p>The four component is the original command-line.</p>
<p>If in interactive mode, the function will verify that everything is correct and will ask you to proceed. Here, the user can answer “no” and provide more information at the <code>string</code> level.</p>
<p>If running the workflow in non-interactive mode or replying “yes,” the <code>createParam</code> function will return the object.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cmd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/createParamFiles.html">createParam</a></span><span class="op">(</span><span class="va">command</span>, writeParamFiles <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; *****BaseCommand*****</span>
<span class="co">#&gt; hisat2 </span>
<span class="co">#&gt; *****Inputs*****</span>
<span class="co">#&gt; S:</span>
<span class="co">#&gt;     type: File</span>
<span class="co">#&gt;     preF: -S</span>
<span class="co">#&gt;     yml: ./results/M1A.sam</span>
<span class="co">#&gt; x:</span>
<span class="co">#&gt;     type: File</span>
<span class="co">#&gt;     preF: -x</span>
<span class="co">#&gt;     yml: ./data/tair10.fasta</span>
<span class="co">#&gt; k:</span>
<span class="co">#&gt;     type: int</span>
<span class="co">#&gt;     preF: -k</span>
<span class="co">#&gt;     yml: 1</span>
<span class="co">#&gt; threads:</span>
<span class="co">#&gt;     type: int</span>
<span class="co">#&gt;     preF: -threads</span>
<span class="co">#&gt;     yml: 4</span>
<span class="co">#&gt; U:</span>
<span class="co">#&gt;     type: File</span>
<span class="co">#&gt;     preF: -U</span>
<span class="co">#&gt;     yml: ./data/SRR446027_1.fastq.gz</span>
<span class="co">#&gt; *****Outputs*****</span>
<span class="co">#&gt; output1:</span>
<span class="co">#&gt;     type: File</span>
<span class="co">#&gt;     value: ./results/M1A.sam</span>
<span class="co">#&gt; *****Parsed raw command line*****</span>
<span class="co">#&gt; hisat2 -S ./results/M1A.sam -x ./data/tair10.fasta -k 1 -threads 4 -U ./data/SRR446027_1.fastq.gz </span>
<span class="co">#&gt;   Written content of 'commandLine' to file: </span>
<span class="co">#&gt;  param/cwl/hisat2/hisat2.cwl </span>
<span class="co">#&gt;   Written content of 'commandLine' to file: </span>
<span class="co">#&gt;  param/cwl/hisat2/hisat2.yml</span></code></pre></div>
</div>
</div>
<div id="how-to-connect-cwl-description-files-within-systempiper" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-connect-cwl-description-files-within-systempiper" class="anchor"></a>How to connect CWL description files within <code>systemPipeR</code>
</h3>
<p>This <a href="https://systempipe.org/systemPipeWorkshop2021/articles/cwl_and_spr.html">article</a> explore the connectivity between the CWL parameters files and <code>systemPipeR</code> workflow management class. In addition, we will show how the workflow can be easily scalable with <code>systemPipeR</code>.</p>
</div>
</div>
<div id="project-initialization" class="section level2">
<h2 class="hasAnchor">
<a href="#project-initialization" class="anchor"></a>Project initialization</h2>
<p>To create a workflow within <code>systemPipeR</code>, we can start by defining an empty workflow container and checking the directory structure:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>projPath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span>, silent<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> </code></pre></div>
<p>Internally, <code>SPRproject</code> function creates a hidden folder called <code>.SPRproject</code>, by default, to store all the log files. A <code>YAML</code> file, here called <code>SYSargsList.yml</code>, also has been created, which initially contains the basic location of the project structure; however, every time the workflow object <code>sal</code> is updated in R, the new information will also be saved in this flat-file database for easy recovery. If you desire different names for the logs folder and the <code>YAML</code> file, these can be modified as follows:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>logs.dir<span class="op">=</span> <span class="st">".SPRproject"</span>, sys.file<span class="op">=</span><span class="st">".SPRproject/SYSargsList.yml"</span><span class="op">)</span> </code></pre></div>
<p>Also, this function will check and/or create the basic folder structure if missing, which means <code>data</code>, <code>param</code>, and <code>results</code> directories, as described <a href="https://systempipe.org/systemPipeWorkshop2021/articles/spr_project.html">here</a>. If the user wants to use a different names for these directories, can be specified as follows:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>data <span class="op">=</span> <span class="st">"data"</span>, param <span class="op">=</span> <span class="st">"param"</span>, results <span class="op">=</span> <span class="st">"results"</span><span class="op">)</span> </code></pre></div>
<p>It is possible to isolate all the R objects created within the workflow analysis from the current environment. <code>SPRproject</code> function provides the option to create a new environment, and in this way, it is not overwriting any object you may want to have at your current section.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>envir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/environment.html">new.env</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> </code></pre></div>
<p>In this stage, the object <code>sal</code> is a empty container, except for the project information. The project information can be accessed by the <code>projectInfo</code> method:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="co">#&gt; Instance of 'SYSargsList': </span>
<span class="co">#&gt;  No workflow steps added</span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">projectInfo</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<p>Also, the <code>length</code> function will return how many steps this workflow contains and in this case it is empty, as follow:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span>
<span class="co">#&gt; [1] 0</span></code></pre></div>
</div>
<div id="workflow-design" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-design" class="anchor"></a>Workflow Design</h2>
<p><code>systemPipeR</code> workflows can be designed and built from start to finish with a single command, importing from an R Markdown file or stepwise in interactive mode from the R console.</p>
<p>In the <a href="#importWF">previous section</a>, we demonstrated how to generate and execute the workflow in a few commands.</p>
<p>In the <a href="#appendstep">next section</a>, we will demonstrate how to build the workflow in an interactive mode.</p>
<p>New workflows are constructed, or existing ones modified, by connecting each step via <code>appendStep</code> method. Each <code>SYSargsList</code> instance contains instructions needed for processing a set of input files with a specific command-line or R software, as well as the paths to the corresponding outfiles generated by a particular tool/step.</p>
<p>To build an R code based step, the constructor function <code>Linewise</code> is used. For more details about this S4 class container, see <a href="http://bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#182_LineWise_Class">here</a>.</p>
<div id="appendstep" class="section level3">
<h3 class="hasAnchor">
<a href="#appendstep" class="anchor"></a>Build workflow interactive</h3>
<p>This tutorial shows a very simple example for describing and explaining all the main features available within <code>systemPipeR</code> to design, build, manage, run, and visualize the workflow. In summary, we are trimming the raw FASTQ files, building the index and mapping the read with <code>Hisat2</code> <span class="citation">(Kim, Langmead, and Salzberg 2015)</span>, and generating a table of reads and alignment counts for all samples.</p>
<p>In the previous section, we initialize the project by building the <code>sal</code> object. Until this moment, the container has no steps:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="co">#&gt; Instance of 'SYSargsList': </span>
<span class="co">#&gt;  No workflow steps added</span></code></pre></div>
<p>Next, we need to populate the object by appending the first step in the workflow.</p>
<div id="adding-the-first-step" class="section level4">
<h4 class="hasAnchor">
<a href="#adding-the-first-step" class="anchor"></a>Adding the first step</h4>
<p>First, we need to populate the object created with the first step in the workflow. Here, an example of how to perform this task using a parameter template file for trimming FASTQ files with <a href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a> software <span class="citation">(Bolger, Lohse, and Usadel 2014)</span>.</p>
<p>The constructor function creates an <code>SYSargsList</code> S4 class object using data from three input files:</p>
<pre><code>- CWL command-line specification file (`wf_file` argument);
- Input variables (`input_file` argument);
- Targets file (`targets` argument).</code></pre>
<p>In CWL, files with the extension <code>.cwl</code> define the parameters of a chosen command-line step or workflow, while files with the extension <code>.yml</code> define the input variables of command-line steps.</p>
<p>The connection between <code>input</code> variables and the <code>targets</code> file is defined under the <code>inputvars</code> argument. It is required a <code>named vector</code>, where each element name needs to match with column names in the <code>targets</code> file, and the value must match the names of the <code>input</code> variables defined in the <code>*.yml</code> files.</p>
<p>A detailed description of the dynamic between <code>input</code> variables and <code>targets</code> files can be found <a href="https://systempipe.org/systemPipeWorkshop2021/articles/cwl_and_spr.html">here</a>. In addition, the CWL syntax overview can be found [herehttps://systempipe.org/systemPipeWorkshop2021/articles/cwl_syntax.html).</p>
<p>Besides all the data form <code>targets</code>, <code>wf_file</code>, <code>input_file</code> and <code>dir_path</code> arguments, <code>SYSargsList</code> constructor function options include:</p>
<ul>
<li>
<code>step_name</code>: a unique <em>name</em> for the step. This is not mandatory; however, it is highly recommended. If no name is provided, a default <code>step_x</code>, where <code>x</code> reflects the step index, will be added.</li>
<li>
<code>dir</code>: this option allows creating an exclusive subdirectory for the step in the workflow. All the <code>outfiles</code> and log files for this particular step will be generated in the respective subfolder.</li>
<li>
<code>dependency</code>: after the first step, all the additional steps appended to the workflow require the information of the dependency tree.</li>
</ul>
<p>The <code>appendStep&lt;-</code> method is used to append a new step in the workflow.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>overwrite <span class="op">=</span> <span class="cn">TRUE</span>, silent <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="st">"targetsPE.txt"</span>
<span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">SYSargsList</a></span><span class="op">(</span>
    step_name <span class="op">=</span> <span class="st">"trimming"</span>,
    targets <span class="op">=</span> <span class="va">targetspath</span>, 
    wf_file <span class="op">=</span> <span class="st">"trimmomatic/trimmomatic-pe.cwl"</span>, input_file <span class="op">=</span> <span class="st">"trimmomatic/trimmomatic-pe.yml"</span>,
    dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>,
    inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1<span class="op">=</span><span class="st">"_FASTQ_PATH1_"</span>, FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>For a brief overview of the workflow, we can check the object as follows:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="co">#&gt; Instance of 'SYSargsList': </span>
<span class="co">#&gt;     WF Steps:</span>
<span class="co">#&gt;        1. trimming --&gt; Status: Pending </span>
<span class="co">#&gt;            Total Files: 72 | Existing: 0 | Missing: 72 </span>
<span class="co">#&gt;          1.1. trimmomatic-pe</span>
<span class="co">#&gt;              cmdlist: 18 | Pending: 18</span>
<span class="co">#&gt; </span></code></pre></div>
<p>Also, for printing and double-check the command-lines in the step, we can use the <code>cmdlist</code> method:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">cmdlist</a></span><span class="op">(</span><span class="va">sal</span>, targets <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; $trimming</span>
<span class="co">#&gt; $trimming$M1A</span>
<span class="co">#&gt; $trimming$M1A$`trimmomatic-pe`</span>
<span class="co">#&gt; [1] "trimmomatic PE -phred33  ./data/SRR446027_1.fastq.gz  ./data/SRR446027_2.fastq.gz  results/M1A_1P.trimmed.fastq.gz  results/M1A_1U.trimmed.fastq.gz  results/M1A_2P.trimmed.fastq.gz  results/M1A_2U.trimmed.fastq.gz  -threads 4 MINLEN:35 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:5"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $trimming$M1B</span>
<span class="co">#&gt; $trimming$M1B$`trimmomatic-pe`</span>
<span class="co">#&gt; [1] "trimmomatic PE -phred33  ./data/SRR446028_1.fastq.gz  ./data/SRR446028_2.fastq.gz  results/M1B_1P.trimmed.fastq.gz  results/M1B_1U.trimmed.fastq.gz  results/M1B_2P.trimmed.fastq.gz  results/M1B_2U.trimmed.fastq.gz  -threads 4 MINLEN:35 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:5"</span></code></pre></div>
</div>
<div id="adding-more-steps" class="section level4">
<h4 class="hasAnchor">
<a href="#adding-more-steps" class="anchor"></a>Adding more steps</h4>
<p>The following steps will demonstrate how to use the short read aligner <code>Hisat2</code> <span class="citation">(Kim, Langmead, and Salzberg 2015)</span> using the <em><code>systemPipeR's</code></em> new command-line interface.</p>
<p>The <code>targets</code> file is optional for workflow steps lacking <code>input</code> files.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">SYSargsList</a></span><span class="op">(</span>
  step_name <span class="op">=</span> <span class="st">"hisat2_index"</span>, 
  targets <span class="op">=</span> <span class="cn">NULL</span>, dir <span class="op">=</span> <span class="cn">FALSE</span>, 
  wf_file <span class="op">=</span> <span class="st">"hisat2/hisat2-index.cwl"</span>, 
  input_file <span class="op">=</span> <span class="st">"hisat2/hisat2-index.yml"</span>,
  dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, 
  dependency <span class="op">=</span> <span class="st">"trimming"</span>
<span class="op">)</span></code></pre></div>
<p>For a overview of the workflow, we can check the object as follows:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="co">#&gt; Instance of 'SYSargsList': </span>
<span class="co">#&gt;     WF Steps:</span>
<span class="co">#&gt;        1. trimming --&gt; Status: Pending </span>
<span class="co">#&gt;            Total Files: 72 | Existing: 0 | Missing: 72 </span>
<span class="co">#&gt;          1.1. trimmomatic-pe</span>
<span class="co">#&gt;              cmdlist: 18 | Pending: 18</span>
<span class="co">#&gt;        2. hisat2_index --&gt; Status: Pending </span>
<span class="co">#&gt;            Total Files: 8 | Existing: 0 | Missing: 8 </span>
<span class="co">#&gt;          2.1. hisat2-index</span>
<span class="co">#&gt;              cmdlist: 1 | Pending: 1</span>
<span class="co">#&gt; </span></code></pre></div>
<p>Note that we have two steps, and it is expected 16 files from the first step and 8 outfiles for the second step. Also, the workflow status is <em>Pending</em>, which means the workflow object is rendered in R; however, we did not execute the workflow yet. In addition to this summary, it can be observed this step has four command-lines.</p>
<p>For more details about the command-line rendered for each target file, it can be checked as follows:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">cmdlist</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"hisat2_index"</span><span class="op">)</span>
<span class="co">#&gt; $hisat2_index</span>
<span class="co">#&gt; $hisat2_index$defaultid</span>
<span class="co">#&gt; $hisat2_index$defaultid$`hisat2-index`</span>
<span class="co">#&gt; [1] "hisat2-build ./data/tair10.fasta  ./data/tair10.fasta "</span></code></pre></div>
</div>
<div id="using-the-outfiles-for-the-next-step" class="section level4">
<h4 class="hasAnchor">
<a href="#using-the-outfiles-for-the-next-step" class="anchor"></a>Using the <code>outfiles</code> for the next step</h4>
<p>For building the next step, all the previous procedures are being used to append the next step. However, here, we can observe powerful features that build the connectivity between steps in the workflow.</p>
<p>In this example, we would like to use the <code>outfiles</code> from <em>trimming</em> step, as input from the next step, which is the <em>hisat2_mapping</em>. In this case, let’s explore the <code>outfiles</code> from the first step:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">outfiles</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>
<span class="co">#&gt; $trimming</span>
<span class="co">#&gt; DataFrame with 18 rows and 4 columns</span>
<span class="co">#&gt;       trimmomatic_1_paired trimmomatic_1_unpaired   trimmomatic_2_paired</span>
<span class="co">#&gt;                &lt;character&gt;            &lt;character&gt;            &lt;character&gt;</span>
<span class="co">#&gt; 1   results/M1A_1P.trimm.. results/M1A_1U.trimm.. results/M1A_2P.trimm..</span>
<span class="co">#&gt; 2   results/M1B_1P.trimm.. results/M1B_1U.trimm.. results/M1B_2P.trimm..</span>
<span class="co">#&gt; 3   results/A1A_1P.trimm.. results/A1A_1U.trimm.. results/A1A_2P.trimm..</span>
<span class="co">#&gt; 4   results/A1B_1P.trimm.. results/A1B_1U.trimm.. results/A1B_2P.trimm..</span>
<span class="co">#&gt; 5   results/V1A_1P.trimm.. results/V1A_1U.trimm.. results/V1A_2P.trimm..</span>
<span class="co">#&gt; ...                    ...                    ...                    ...</span>
<span class="co">#&gt; 14  results/M12B_1P.trim.. results/M12B_1U.trim.. results/M12B_2P.trim..</span>
<span class="co">#&gt; 15  results/A12A_1P.trim.. results/A12A_1U.trim.. results/A12A_2P.trim..</span>
<span class="co">#&gt; 16  results/A12B_1P.trim.. results/A12B_1U.trim.. results/A12B_2P.trim..</span>
<span class="co">#&gt; 17  results/V12A_1P.trim.. results/V12A_1U.trim.. results/V12A_2P.trim..</span>
<span class="co">#&gt; 18  results/V12B_1P.trim.. results/V12B_1U.trim.. results/V12B_2P.trim..</span>
<span class="co">#&gt;     trimmomatic_2_unpaired</span>
<span class="co">#&gt;                &lt;character&gt;</span>
<span class="co">#&gt; 1   results/M1A_2U.trimm..</span>
<span class="co">#&gt; 2   results/M1B_2U.trimm..</span>
<span class="co">#&gt; 3   results/A1A_2U.trimm..</span>
<span class="co">#&gt; 4   results/A1B_2U.trimm..</span>
<span class="co">#&gt; 5   results/V1A_2U.trimm..</span>
<span class="co">#&gt; ...                    ...</span>
<span class="co">#&gt; 14  results/M12B_2U.trim..</span>
<span class="co">#&gt; 15  results/A12A_2U.trim..</span>
<span class="co">#&gt; 16  results/A12B_2U.trim..</span>
<span class="co">#&gt; 17  results/V12A_2U.trim..</span>
<span class="co">#&gt; 18  results/V12B_2U.trim..</span></code></pre></div>
<p>The two-column the user may want to use are “trimmomatic_1_paired” and “trimmomatic_2_paired.” For the argument <code>targets</code> in the <code>SYSargsList</code> function, it should provide the name of the correspondent step in the workflow and which <code>outfiles</code> should be incorporated in the next step.</p>
<p>The argument <code>inputvars</code> allows the connectivity between <code>outfiles</code> and the new <code>targets</code> file. Here, the name of the previous <code>outfiles</code> should be provided it. Please note that all <code>outfiles</code> column names must be unique.</p>
<p>It is possible to keep all the original columns from the <code>targets</code> files or remove some columns for a clean <code>targets</code> file. The argument <code>rm_targets_col</code> provides this flexibility, where it is possible to specify the names of the columns that should be removed. If no names are passing here, the new columns will be appended.</p>
<p>In addition, a useful utility establishes a dependency among the steps in the workflow, and this is possible with the <code>dependency</code> argument, where the names of steps are specified.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">SYSargsList</a></span><span class="op">(</span>
  step_name <span class="op">=</span> <span class="st">"hisat2_mapping"</span>,
  targets <span class="op">=</span> <span class="st">"trimming"</span>, dir <span class="op">=</span> <span class="cn">TRUE</span>, 
  wf_file <span class="op">=</span> <span class="st">"workflow-hisat2/workflow_hisat2-pe.cwl"</span>,
  input_file <span class="op">=</span> <span class="st">"workflow-hisat2/workflow_hisat2-pe.yml"</span>,
  dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>,
  inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>trimmomatic_1_paired <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, trimmomatic_2_paired <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, 
                SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>,
  rm_targets_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"FileName1"</span>, <span class="st">"FileName2"</span><span class="op">)</span>, 
  dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"trimming"</span>, <span class="st">"hisat2_index"</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p>We can check the targets automatically create for this step, based on the previous <code>targetsWF</code>:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">targetsWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; $hisat2_mapping</span>
<span class="co">#&gt; DataFrame with 18 rows and 7 columns</span>
<span class="co">#&gt;       trimmomatic_1_paired   trimmomatic_2_paired  SampleName      Factor</span>
<span class="co">#&gt;                &lt;character&gt;            &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt; 1   results/M1A_1P.trimm.. results/M1A_2P.trimm..         M1A          M1</span>
<span class="co">#&gt; 2   results/M1B_1P.trimm.. results/M1B_2P.trimm..         M1B          M1</span>
<span class="co">#&gt; 3   results/A1A_1P.trimm.. results/A1A_2P.trimm..         A1A          A1</span>
<span class="co">#&gt; 4   results/A1B_1P.trimm.. results/A1B_2P.trimm..         A1B          A1</span>
<span class="co">#&gt; 5   results/V1A_1P.trimm.. results/V1A_2P.trimm..         V1A          V1</span>
<span class="co">#&gt; ...                    ...                    ...         ...         ...</span>
<span class="co">#&gt; 14  results/M12B_1P.trim.. results/M12B_2P.trim..        M12B         M12</span>
<span class="co">#&gt; 15  results/A12A_1P.trim.. results/A12A_2P.trim..        A12A         A12</span>
<span class="co">#&gt; 16  results/A12B_1P.trim.. results/A12B_2P.trim..        A12B         A12</span>
<span class="co">#&gt; 17  results/V12A_1P.trim.. results/V12A_2P.trim..        V12A         V12</span>
<span class="co">#&gt; 18  results/V12B_1P.trim.. results/V12B_2P.trim..        V12B         V12</span>
<span class="co">#&gt;      SampleLong  Experiment        Date</span>
<span class="co">#&gt;     &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt; 1     Mock.1h.A           1 23-Mar-2012</span>
<span class="co">#&gt; 2     Mock.1h.B           1 23-Mar-2012</span>
<span class="co">#&gt; 3      Avr.1h.A           1 23-Mar-2012</span>
<span class="co">#&gt; 4      Avr.1h.B           1 23-Mar-2012</span>
<span class="co">#&gt; 5      Vir.1h.A           1 23-Mar-2012</span>
<span class="co">#&gt; ...         ...         ...         ...</span>
<span class="co">#&gt; 14   Mock.12h.B           1 23-Mar-2012</span>
<span class="co">#&gt; 15    Avr.12h.A           1 23-Mar-2012</span>
<span class="co">#&gt; 16    Avr.12h.B           1 23-Mar-2012</span>
<span class="co">#&gt; 17    Vir.12h.A           1 23-Mar-2012</span>
<span class="co">#&gt; 18    Vir.12h.B           1 23-Mar-2012</span></code></pre></div>
<p>We can also check all the expected <code>outfiles</code> for this particular step, as follows:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">outfiles</a></span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; $hisat2_mapping</span>
<span class="co">#&gt; DataFrame with 18 rows and 4 columns</span>
<span class="co">#&gt;             hisat2_sam       samtools_bam      samtools_sort_bam</span>
<span class="co">#&gt;            &lt;character&gt;        &lt;character&gt;            &lt;character&gt;</span>
<span class="co">#&gt; 1    ./results/M1A.sam  ./results/M1A.bam ./results/M1A.sorted..</span>
<span class="co">#&gt; 2    ./results/M1B.sam  ./results/M1B.bam ./results/M1B.sorted..</span>
<span class="co">#&gt; 3    ./results/A1A.sam  ./results/A1A.bam ./results/A1A.sorted..</span>
<span class="co">#&gt; 4    ./results/A1B.sam  ./results/A1B.bam ./results/A1B.sorted..</span>
<span class="co">#&gt; 5    ./results/V1A.sam  ./results/V1A.bam ./results/V1A.sorted..</span>
<span class="co">#&gt; ...                ...                ...                    ...</span>
<span class="co">#&gt; 14  ./results/M12B.sam ./results/M12B.bam ./results/M12B.sorte..</span>
<span class="co">#&gt; 15  ./results/A12A.sam ./results/A12A.bam ./results/A12A.sorte..</span>
<span class="co">#&gt; 16  ./results/A12B.sam ./results/A12B.bam ./results/A12B.sorte..</span>
<span class="co">#&gt; 17  ./results/V12A.sam ./results/V12A.bam ./results/V12A.sorte..</span>
<span class="co">#&gt; 18  ./results/V12B.sam ./results/V12B.bam ./results/V12B.sorte..</span>
<span class="co">#&gt;             samtools_index</span>
<span class="co">#&gt;                &lt;character&gt;</span>
<span class="co">#&gt; 1   ./results/M1A.sorted..</span>
<span class="co">#&gt; 2   ./results/M1B.sorted..</span>
<span class="co">#&gt; 3   ./results/A1A.sorted..</span>
<span class="co">#&gt; 4   ./results/A1B.sorted..</span>
<span class="co">#&gt; 5   ./results/V1A.sorted..</span>
<span class="co">#&gt; ...                    ...</span>
<span class="co">#&gt; 14  ./results/M12B.sorte..</span>
<span class="co">#&gt; 15  ./results/A12A.sorte..</span>
<span class="co">#&gt; 16  ./results/A12B.sorte..</span>
<span class="co">#&gt; 17  ./results/V12A.sorte..</span>
<span class="co">#&gt; 18  ./results/V12B.sorte..</span></code></pre></div>
<p>Now, we can observe that the third step has been added and contains one substep.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="co">#&gt; Instance of 'SYSargsList': </span>
<span class="co">#&gt;     WF Steps:</span>
<span class="co">#&gt;        1. trimming --&gt; Status: Pending </span>
<span class="co">#&gt;            Total Files: 72 | Existing: 0 | Missing: 72 </span>
<span class="co">#&gt;          1.1. trimmomatic-pe</span>
<span class="co">#&gt;              cmdlist: 18 | Pending: 18</span>
<span class="co">#&gt;        2. hisat2_index --&gt; Status: Pending </span>
<span class="co">#&gt;            Total Files: 8 | Existing: 0 | Missing: 8 </span>
<span class="co">#&gt;          2.1. hisat2-index</span>
<span class="co">#&gt;              cmdlist: 1 | Pending: 1</span>
<span class="co">#&gt;        3. hisat2_mapping --&gt; Status: Pending </span>
<span class="co">#&gt;            Total Files: 72 | Existing: 0 | Missing: 72 </span>
<span class="co">#&gt;          3.1. hisat2</span>
<span class="co">#&gt;              cmdlist: 18 | Pending: 18</span>
<span class="co">#&gt;          3.2. samtools-view</span>
<span class="co">#&gt;              cmdlist: 18 | Pending: 18</span>
<span class="co">#&gt;          3.3. samtools-sort</span>
<span class="co">#&gt;              cmdlist: 18 | Pending: 18</span>
<span class="co">#&gt;          3.4. samtools-index</span>
<span class="co">#&gt;              cmdlist: 18 | Pending: 18</span>
<span class="co">#&gt; </span></code></pre></div>
<p>In addition, we can access all the command-lines for each one of the substeps.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">cmdlist</a></span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="st">"hisat2_mapping"</span><span class="op">]</span>, targets <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; $hisat2_mapping</span>
<span class="co">#&gt; $hisat2_mapping$M1A</span>
<span class="co">#&gt; $hisat2_mapping$M1A$hisat2</span>
<span class="co">#&gt; [1] "hisat2 -S ./results/M1A.sam  -x ./data/tair10.fasta  -k 1  --min-intronlen 30  --max-intronlen 3000  -1 results/M1A_1P.trimmed.fastq.gz -2 results/M1A_2P.trimmed.fastq.gz --threads 4"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $hisat2_mapping$M1A$`samtools-view`</span>
<span class="co">#&gt; [1] "samtools view -bS -o ./results/M1A.bam  ./results/M1A.sam "</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $hisat2_mapping$M1A$`samtools-sort`</span>
<span class="co">#&gt; [1] "samtools sort -o ./results/M1A.sorted.bam  ./results/M1A.bam  -@ 4"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $hisat2_mapping$M1A$`samtools-index`</span>
<span class="co">#&gt; [1] "samtools index -b results/M1A.sorted.bam  results/M1A.sorted.bam.bai  ./results/M1A.sorted.bam "</span></code></pre></div>
</div>
<div id="getting-data-from-a-workflow-instance" class="section level4">
<h4 class="hasAnchor">
<a href="#getting-data-from-a-workflow-instance" class="anchor"></a>Getting data from a workflow instance</h4>
<p>The final step in this simple workflow is an R code step. For that, we are using the <code>LineWise</code> constructor function.</p>
<p>This constructor function requires the <code>step_name</code> and the R-based code under the <code>code</code> argument. The R code should be enclosed by braces (<code>{}</code>) and separated by a new line.</p>
<p>One interesting feature showed here is the <code>getColumn</code> method that allows extracting the information for a workflow instance. Those files can be used in an R code, as demonstrated below.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"hisat2_mapping"</span>, <span class="st">"outfiles"</span>, column <span class="op">=</span> <span class="st">"samtools_sort_bam"</span><span class="op">)</span>
<span class="co">#&gt;                         M1A                         M1B </span>
<span class="co">#&gt;  "./results/M1A.sorted.bam"  "./results/M1B.sorted.bam" </span>
<span class="co">#&gt;                         A1A                         A1B </span>
<span class="co">#&gt;  "./results/A1A.sorted.bam"  "./results/A1B.sorted.bam" </span>
<span class="co">#&gt;                         V1A                         V1B </span>
<span class="co">#&gt;  "./results/V1A.sorted.bam"  "./results/V1B.sorted.bam" </span>
<span class="co">#&gt;                         M6A                         M6B </span>
<span class="co">#&gt;  "./results/M6A.sorted.bam"  "./results/M6B.sorted.bam" </span>
<span class="co">#&gt;                         A6A                         A6B </span>
<span class="co">#&gt;  "./results/A6A.sorted.bam"  "./results/A6B.sorted.bam" </span>
<span class="co">#&gt;                         V6A                         V6B </span>
<span class="co">#&gt;  "./results/V6A.sorted.bam"  "./results/V6B.sorted.bam" </span>
<span class="co">#&gt;                        M12A                        M12B </span>
<span class="co">#&gt; "./results/M12A.sorted.bam" "./results/M12B.sorted.bam" </span>
<span class="co">#&gt;                        A12A                        A12B </span>
<span class="co">#&gt; "./results/A12A.sorted.bam" "./results/A12B.sorted.bam" </span>
<span class="co">#&gt;                        V12A                        V12B </span>
<span class="co">#&gt; "./results/V12A.sorted.bam" "./results/V12B.sorted.bam"</span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
  <span class="va">bampaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"hisat2_mapping"</span>, <span class="st">"outfiles"</span>, column <span class="op">=</span> <span class="st">"samtools_sort_bam"</span><span class="op">)</span>
  <span class="va">fqpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"trimming"</span>, <span class="st">"targetsWF"</span>, column <span class="op">=</span> <span class="st">"FileName1"</span><span class="op">)</span>
  <span class="va">read_statsDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/alignStats.html">alignStats</a></span><span class="op">(</span>args<span class="op">=</span><span class="va">bampaths</span>, fqpaths <span class="op">=</span> <span class="va">fqpaths</span>, pairEnd <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">read_statsDF</span>, <span class="st">"results/alignStats.xls"</span>, row.names<span class="op">=</span><span class="cn">FALSE</span>, quote<span class="op">=</span><span class="cn">FALSE</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span>
  <span class="op">}</span>,
  step_name <span class="op">=</span> <span class="st">"mapping_stats"</span>, 
  dependency <span class="op">=</span> <span class="st">"hisat2_mapping"</span><span class="op">)</span></code></pre></div>
<p>Let’s explore the workflow to check the steps:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">stepsWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">dependency</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">codeLine</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
</div>
</div>
</div>
<div id="running-the-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#running-the-workflow" class="anchor"></a>Running the workflow</h1>
<p>For running the workflow, <code>runWF</code> function will execute all the command-lines store in the workflow container.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html">runWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<p>This essential function allows the user to choose one or multiple steps to be executed using the <code>steps</code> argument. However, it is necessary to follow the workflow dependency graph. If a selected step depends on a previous step(s) that was not executed, the execution will fail.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html">runWF</a></span><span class="op">(</span><span class="va">sal</span>, steps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Also, it allows forcing the execution of the steps, even if the status of the step is <code>'Success'</code> and all the expected <code>outfiles</code> exists. Another feature of the <code>runWF</code> function is ignoring all the warnings and errors and running the workflow by the arguments <code>warning.stop</code> and <code>error.stop</code>, respectively.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html">runWF</a></span><span class="op">(</span><span class="va">sal</span>, force <span class="op">=</span> <span class="cn">TRUE</span>, warning.stop <span class="op">=</span> <span class="cn">FALSE</span>, error.stop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>When the project was initialized by <code>SPRproject</code> function, it was created an environment for all objects created during the workflow execution. This environment can be accessed as follows:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">viewEnvir</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<p>The workflow execution allows to save this environment for future recovery:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html">runWF</a></span><span class="op">(</span><span class="va">sal</span>, saveEnv <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div id="workflow-status" class="section level3">
<h3 class="hasAnchor">
<a href="#workflow-status" class="anchor"></a>Workflow status</h3>
<p>To check the summary of the workflow, we can use:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="co">#&gt; Instance of 'SYSargsList': </span>
<span class="co">#&gt;     WF Steps:</span>
<span class="co">#&gt;        1. trimming --&gt; Status: Pending </span>
<span class="co">#&gt;            Total Files: 72 | Existing: 0 | Missing: 72 </span>
<span class="co">#&gt;          1.1. trimmomatic-pe</span>
<span class="co">#&gt;              cmdlist: 18 | Pending: 18</span>
<span class="co">#&gt;        2. hisat2_index --&gt; Status: Pending </span>
<span class="co">#&gt;            Total Files: 8 | Existing: 0 | Missing: 8 </span>
<span class="co">#&gt;          2.1. hisat2-index</span>
<span class="co">#&gt;              cmdlist: 1 | Pending: 1</span>
<span class="co">#&gt;        3. hisat2_mapping --&gt; Status: Pending </span>
<span class="co">#&gt;            Total Files: 72 | Existing: 0 | Missing: 72 </span>
<span class="co">#&gt;          3.1. hisat2</span>
<span class="co">#&gt;              cmdlist: 18 | Pending: 18</span>
<span class="co">#&gt;          3.2. samtools-view</span>
<span class="co">#&gt;              cmdlist: 18 | Pending: 18</span>
<span class="co">#&gt;          3.3. samtools-sort</span>
<span class="co">#&gt;              cmdlist: 18 | Pending: 18</span>
<span class="co">#&gt;          3.4. samtools-index</span>
<span class="co">#&gt;              cmdlist: 18 | Pending: 18</span>
<span class="co">#&gt;        4. mapping_stats --&gt; Status: Pending</span>
<span class="co">#&gt; </span></code></pre></div>
<p>To access more details about the workflow instances, we can use the <code>statusWF</code> method:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">statusWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
<div id="visualize-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-workflow" class="anchor"></a>Visualize workflow</h2>
<p><code>systemPipeR</code> workflows instances can be visualized with the <code>plotWF</code> function.</p>
<p>This function will make a plot of selected workflow instance and the following information is displayed on the plot:</p>
<pre><code>- Workflow structure (dependency graphs between different steps); 
- Workflow step status, *e.g.* `Success`, `Error`, `Pending`, `Warnings`; 
- Sample status and statistics; 
- Workflow timing: running duration time. </code></pre>
<p>If no argument is provided, the basic plot will automatically detect width, height, layout, plot method, branches, <em>etc</em>.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/plotWF.html">plotWF</a></span><span class="op">(</span><span class="va">sal</span>, show_legend <span class="op">=</span> <span class="cn">TRUE</span>, width <span class="op">=</span> <span class="st">"80%"</span>, rstudio <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-409e80e51d848ca5e1ed" style="width:80%;height:432.632880098888px;" class="plotwf html-widget"></div>
<script type="application/json" data-for="htmlwidget-409e80e51d848ca5e1ed">{"x":{"dot":"digraph {\n    node[fontsize=20];\n    subgraph {\n        node[color=\"dodgerblue\"];\n        trimming -> hisat2_index -> hisat2_mapping -> mapping_stats[color=\"dodgerblue\"]\n   }\n    \n    \n    trimming[label=<<b><font color=\"black\">trimming<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">72<\/font><\/b>; <font color=\"black\">0s<\/font>> , style=\"rounded\", shape=\"box\"  tooltip=\"step trimming: 0 samples passed; 0 samples have warnings; 0 samples have errors; 72 samples in total; Start time: 2021-08-06 13:27:49; End time: 2021-08-06 13:27:49; Duration: 00:00:00\"]\n    hisat2_index[label=<<b><font color=\"black\">hisat2_index<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">8<\/font><\/b>; <font color=\"black\">0s<\/font>> , style=\"rounded\", shape=\"box\"  tooltip=\"step hisat2_index: 0 samples passed; 0 samples have warnings; 0 samples have errors; 8 samples in total; Start time: 2021-08-06 13:27:49; End time: 2021-08-06 13:27:49; Duration: 00:00:00\"]\n    hisat2_mapping[label=<<b><font color=\"black\">hisat2_mapping<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">72<\/font><\/b>; <font color=\"black\">0s<\/font>> , style=\"rounded\", shape=\"box\"  tooltip=\"step hisat2_mapping: 0 samples passed; 0 samples have warnings; 0 samples have errors; 72 samples in total; Start time: 2021-08-06 13:27:49; End time: 2021-08-06 13:27:49; Duration: 00:00:00\"]\n    mapping_stats[label=<<b><font color=\"black\">mapping_stats<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step mapping_stats: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2021-08-06 13:27:49; End time: 2021-08-06 13:27:49; Duration: 00:00:00\"]\n    subgraph cluster_legend {\n        rankdir=TB;\n        color=\"#EEEEEE\";\n        style=filled;\n        node [style=filled];\n        {rank=same; R_step; Sysargs_step; Main_branch}\n        Main_branch -> Sysargs_step -> R_step[color=\"#EEEEEE\"]\n        Main_branch[label=\"Main branch\" color=\"dodgerblue\", style=\"filled\", fillcolor=white];   Sysargs_step ->step_state[color=\"#EEEEEE\"];\n        step_state[style=\"filled\", shape=\"box\" color=white, label =<\n            <table>\n            <tr><td><b>Step Colors<\/b><\/td><\/tr>\n            <tr><td><font color=\"black\">Pending steps<\/font>; <font color=\"#5cb85c\">Successful steps<\/font>; <font color=\"#d9534f\">Failed steps<\/font><\/td><\/tr>\n            <tr><td><b>Targets Files / Code Chunk <\/b><\/td><\/tr><tr><td><font color=\"#5cb85c\">0 (pass) <\/font> | <font color=\"#f0ad4e\">0 (warning) <\/font> | <font color=\"#d9534f\">0 (error) <\/font> | <font color=\"blue\">0 (total)<\/font>; Duration<\/td><\/tr><\/table>\n            >];\n        label=\"Legends\";\n        fontsize = 30;\n        Sysargs_step[label=\"Sysargs step\" style=\"rounded, filled\", shape=\"box\", fillcolor=white];\n        R_step[label=\"R step\" style=\"rounded, filled\", fillcolor=white];\n    }\n\n}\n","plotid":"sprwf-84157632","responsive":true,"width":"80%","height":null,"plot_method":"renderSVGElement","rmd":true,"msg":""},"evals":[],"jsHooks":[]}</script><p>For more details about the <code>plotWF</code> function, please see <a href="http://bioconductor.org/packages/devel/bioc/vignettes/systemPipeR">here</a>.</p>
</div>
<div id="generation-of-technical-report" class="section level2">
<h2 class="hasAnchor">
<a href="#generation-of-technical-report" class="anchor"></a>Generation of technical report</h2>
<p><code>systemPipeR</code> compiles all the workflow execution logs in one central location, making it easier to check any standard output (<code>stdout</code>) or standard error (<code>stderr</code>) for any command-line tools used on the workflow or the R code <code>stdout</code>. Also, the workflow plot is appended at the beginning of the report, making it easier to click on the respective step.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderLogs.html">renderLogs</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
<div id="parallelization-on-clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#parallelization-on-clusters" class="anchor"></a>Parallelization on clusters</h2>
<p>This section of the tutorial provides an introduction to the usage of the <code>systemPipeR</code> features on a cluster.</p>
<p>The computation can be greatly accelerated by processing many files in parallel using several compute nodes of a cluster, where a scheduling/queuing system is used for load balancing. For this the <code>clusterRun</code> function submits the computing requests to the scheduler using the run specifications defined by <code>runWF</code>.</p>
<p>A named list provides the computational resources. By default, it can be defined the upper time limit in minutes for jobs before they get killed by the scheduler, memory limit in Mb, number of <code>CPUs</code>, and number of tasks.</p>
<p>The number of independent parallel cluster processes is defined under the <code>Njobs</code> argument. The following example will run one process in parallel using for each 4 CPU cores. If the resources available on a cluster allow running all the processes simultaneously, then the shown sample submission will utilize in total four CPU cores (<code>NJobs * ncpus</code>). Note, <code>clusterRun</code> can be used with most queueing systems as it is based on utilities from the <em><code>batchtools</code></em> package which supports the use of template files (<em><code>*.tmpl</code></em>) for defining the run parameters of different schedulers. To run the following code, one needs to have both a <code>conffile</code> (see <em><code>.batchtools.conf.R</code></em> samples <a href="https://mllg.github.io/batchtools/">here</a>) and a template file (see <em><code>*.tmpl</code></em> samples <a href="https://github.com/mllg/batchtools/tree/master/inst/templates">here</a>) for the queueing available on a system. The following example uses the sample <code>conf</code> and <code>template</code> files for the <code>Slurm</code> scheduler provided by this package.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mllg/batchtools">batchtools</a></span><span class="op">)</span>
<span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime<span class="op">=</span><span class="fl">120</span>, ntasks<span class="op">=</span><span class="fl">1</span>, ncpus<span class="op">=</span><span class="fl">4</span>, memory<span class="op">=</span><span class="fl">1024</span><span class="op">)</span>
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">sal</span>, FUN <span class="op">=</span> <span class="va">runWF</span>, 
                  more.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>,
                  conffile<span class="op">=</span><span class="st">".batchtools.conf.R"</span>, 
                  template<span class="op">=</span><span class="st">"batchtools.slurm.tmpl"</span>,
                  Njobs<span class="op">=</span><span class="fl">1</span>, runid<span class="op">=</span><span class="st">"01"</span>, resourceList<span class="op">=</span><span class="va">resources</span><span class="op">)</span></code></pre></div>
<p>Note: The example is submitting the jog to <code>short</code> partition. If you desire to use a different partition, please adjust accordingly (<code>batchtools.slurm.tmpl</code>).</p>
</div>
</div>
<div id="exported-the-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#exported-the-workflow" class="anchor"></a>Exported the workflow</h1>
<p><code>systemPipeR</code> workflow management system allows to translate and export the workflow built interactively to R Markdown format or an executable bash script. This feature advances the reusability of the workflow, as well as the flexibility for workflow execution.</p>
<div id="r-markdown-file" class="section level3">
<h3 class="hasAnchor">
<a href="#r-markdown-file" class="anchor"></a>R Markdown file</h3>
<p><code>sal2rmd</code> function takes an <code>SYSargsList</code> workflow container and translates it to SPR workflow template R markdown format. This file can be imported with the <code>importWF</code> function, as demonstrated above.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sal2rmd.html">sal2rmd</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
<div id="bash-script" class="section level3">
<h3 class="hasAnchor">
<a href="#bash-script" class="anchor"></a>Bash script</h3>
<p><code>sal2bash</code> function takes an <code>SYSargsList</code> workflow container and translates it to an executable bash script, and the script can run the workflow without loading <code>SPR</code> or using an R console.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sal2bash.html">sal2bash</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<p>It will be generated on the project root an executable bash script, called by default the <code>spr_wf.sh</code>. Also, a directory <code>./spr_wf</code> will be created and store all the R scripts based on the workflow steps. Please note that this function will “collapse” adjacent R steps into one file as much as possible.</p>
</div>
<div id="project-resume-and-restart" class="section level2">
<h2 class="hasAnchor">
<a href="#project-resume-and-restart" class="anchor"></a>Project Resume and Restart</h2>
<p>If you desire to resume or restart a project that has been initialized in the past, <code>SPRproject</code> function allows this operation.</p>
<p>With the resume option, it is possible to load the <code>SYSargsList</code> object in R and resume the analysis. Please, make sure to provide the <code>logs.dir</code> location, and the corresponded <code>YAML</code> file name. The current working directory needs to be in the project root directory.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>resume <span class="op">=</span> <span class="cn">TRUE</span>, logs.dir <span class="op">=</span> <span class="st">".SPRproject"</span>, 
                  sys.file <span class="op">=</span> <span class="st">".SPRproject/SYSargsList.yml"</span><span class="op">)</span> </code></pre></div>
<p>If you choose to save the environment in the last analysis, you can recover all the files created in that particular section. <code>SPRproject</code> function allows this with <code>load.envir</code> argument. Please note that the environment was saved only with you run the workflow in the last section (<code><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html">runWF()</a></code>).</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>resume <span class="op">=</span> <span class="cn">TRUE</span>, load.envir <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </code></pre></div>
<p>After loading the workflow at your current section, you can check the objects created in the old environment and decide if it is necessary to copy them to the current environment.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">viewEnvir</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">copyEnvir</a></span><span class="op">(</span><span class="va">sal</span>, new.env <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/environment.html">globalenv</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The <code>resume</code> option will keep all previous logs in the folder; however, if you desire to clean the execution history and restart the workflow, the <code>restart=TRUE</code> option can be used.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>restart <span class="op">=</span> <span class="cn">TRUE</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span>, load.envir <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </code></pre></div>
<p>The last and more drastic option from <code>SYSproject</code> function is to overwrite the logs and the workflow. This option will delete the hidden folder and the information on the <code>SYSargsList.yml</code> files. This will not delete any parameter file nor any results it was created in previous runs. Please use with caution.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </code></pre></div>
</div>
</div>
<div id="systempiper-workflow-management-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#systempiper-workflow-management-functions" class="anchor"></a><code>systemPipeR</code> Workflow Management Functions</h1>
<p>Please find <a href="https://systempipe.org/systemPipeWorkshop2021/articles/spr_functions.html">here</a> a complete list of systemPipeR’s functions and methods to create, configure, modify, run, visualize and export the <code>SYSargsList</code> container.</p>
</div>
<div id="overview-of-systempipeshiny-core-functionalities" class="section level1">
<h1 class="hasAnchor">
<a href="#overview-of-systempipeshiny-core-functionalities" class="anchor"></a>Overview of <em>systemPipeShiny</em> core functionalities</h1>
<p><strong><span style="color:#5DA7D6;">s</span>ystem<span style="color:#5DA7D6;">P</span>ipe<span style="color:#5DA7D6;">S</span>hiny</strong> (SPS) extends the widely used <a href="/spr/">systemPipeR</a> (SPR) workflow environment with a versatile graphical user interface provided by a <a href="https://shiny.rstudio.com">Shiny App</a>. This allows non-R users, such as experimentalists, to run many <code>systemPipeR's</code> workflow designs, control, and visualization functionalities interactively without requiring knowledge of R. Most importantly, <code>SPS</code> has been designed as a general purpose framework for interacting with other R packages in an intuitive manner. Like most Shiny Apps, SPS can be used on both local computers as well as centralized server-based deployments that can be accessed remotely as a public web service for using SPR’s functionalities with community and/or private data. The framework can integrate many core packages from the R/Bioconductor ecosystem. Examples of SPS’ current functionalities include:</p>
<ul>
<li>A workflow module to create experimental designs, visualize and customize workflow topologies with previews, and programming free workflow execution within the application.</li>
<li>An interactive module with extensive plot options to visualize downstream analysis of a RNA-Seq workflow.</li>
<li>A quick ggplot module to make all variety of scientific plots from any user defined tabular data.</li>
<li>An extendable set of visualization functionalities makes it easy to design custom Shiny Apps under SPS framework without any knowledge of Shiny.</li>
<li>A ‘Canvas Workbench’ to manage complex visual results. It allows users to organize and to compare plots in an efficient manner combined with a session screenshot feature to edit scientific and publishable figures.</li>
<li>Three other supporting packages to help all users from beginners and advanced developers to extend under current SPS framework or on their own visualization apps and Rmarkdown.</li>
</ul>
<div id="run" class="section level2">
<h2 class="hasAnchor">
<a href="#run" class="anchor"></a>Run</h2>
<p>In this workshop, we are using the most recent develop version of <code>SPS.</code> If you are using the same Rstudio server for SPR from <a href="http://app.orchestra.cancerdatasci.org/">Orchestra{blk}</a> to run <code>SPS</code>, no more action is needed. Use following code to start <code>SPS</code> in the Rstudio server.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://systempipe.org/sps">systemPipeShiny</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeShiny/man/spsInit.html">spsInit</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html">runApp</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div id="demos" class="section level2">
<h2 class="hasAnchor">
<a href="#demos" class="anchor"></a>Demos</h2>
<p>If you could not either use the <a href="http://app.orchestra.cancerdatasci.org/">Orchestra{blk}</a>, or run it locally, here are some deployed demos:</p>
<table class="table">
<colgroup>
<col width="33%">
<col width="33%">
<col width="33%">
</colgroup>
<thead><tr class="header">
<th>Type and link</th>
<th>option changed</th>
<th>notes</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny/">Default full installation{blk}</a></td>
<td>takes ~20s to load</td>
<td>full app</td>
</tr>
<tr class="even">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny/">Minimum installation{blk}</a></td>
<td></td>
<td>no modules installed</td>
</tr>
<tr class="odd">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny_loading/">Login enabled{blk}</a></td>
<td><code>login_screen = TRUE; login_theme = "empty"</code></td>
<td>no modules installed</td>
</tr>
<tr class="even">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny_loading_theme/">Login and login themes{blk}</a></td>
<td><code>login_screen = TRUE; login_theme = "random"</code></td>
<td>no modules installed</td>
</tr>
<tr class="odd">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny_loading/?admin">App admin page{blk}</a></td>
<td><code>admin_page = TRUE</code></td>
<td>or simply add “?admin” to the end of URL of demos</td>
</tr>
</tbody>
</table>
<p>For the login required demos, the app account name is <strong>“user”</strong> password <strong>“user”</strong>.</p>
<p>For the admin panel login, account name <strong>“admin”</strong>, password <strong>“admin”</strong>.</p>
<p><strong>Please DO NOT delete or change password when you are using the admin features.</strong> <em>shinyapp.io</em> will reset the app once a while, but this will affect other people who are trying the demo simultaneously.</p>
</div>
<div id="sreenshots-of-sps-features" class="section level2">
<h2 class="hasAnchor">
<a href="#sreenshots-of-sps-features" class="anchor"></a>Sreenshots of <code>SPS</code> features</h2>
<div id="sps-main-app" class="section level3">
<h3 class="hasAnchor">
<a href="#sps-main-app" class="anchor"></a><code>SPS</code> main app</h3>
<p><img src="images/main_app.png" height="500px" class="center"></p>
</div>
<div id="module-workflow" class="section level3">
<h3 class="hasAnchor">
<a href="#module-workflow" class="anchor"></a>Module workflow</h3>
<p>A step-by-step guided workplace to create -&gt; run workflows.</p>
<p><img src="images/wf_main.png" height="500px" class="center"></p>
</div>
<div id="module-workflow-create-targets" class="section level3">
<h3 class="hasAnchor">
<a href="#module-workflow-create-targets" class="anchor"></a>Module workflow create targets</h3>
<p>Create and check targets file.</p>
<p><img src="images/wf_targets.png" height="500px" class="center"></p>
</div>
<div id="module-workflow-design-workflow-steps" class="section level3">
<h3 class="hasAnchor">
<a href="#module-workflow-design-workflow-steps" class="anchor"></a>Module workflow design workflow steps</h3>
<p>Add/remove/modify/change orders of your workflows with all the visual aids.</p>
<p><img src="images/wf_wf.png" height="300px" class="center"></p>
</div>
<div id="module-workflow-run" class="section level3">
<h3 class="hasAnchor">
<a href="#module-workflow-run" class="anchor"></a>Module workflow run</h3>
<p>Execute the workflow directly in SPS with the Rstudio-like interface.</p>
<p><img src="images/wf_run.png" height="400px" class="center"></p>
</div>
<div id="module-rnaseq-clustering-and-plotting" class="section level3">
<h3 class="hasAnchor">
<a href="#module-rnaseq-clustering-and-plotting" class="anchor"></a>Module RNAseq clustering and plotting</h3>
<p>Clustering and plotting your read counts.</p>
<p><img src="images/rnaseq_heatmap.png" height="500px" class="center"></p>
</div>
<div id="module-rna-seq-deg" class="section level3">
<h3 class="hasAnchor">
<a href="#module-rna-seq-deg" class="anchor"></a>Module RNA-seq DEG</h3>
<p>DEG analysis with a few clicks.</p>
<p><img src="images/rnaseq_deg.png" height="500px" class="center"></p>
</div>
<div id="module-quick-ggplot" class="section level3">
<h3 class="hasAnchor">
<a href="#module-quick-ggplot" class="anchor"></a>Module Quick <code>ggplot</code>
</h3>
<ul>
<li><p>Powered by {<a href="https://github.com/dreamRs/esquisse">esquisse</a>} package.</p></li>
<li><p>Plot any tabular file.</p></li>
</ul>
<p><img src="images/sps_quickgg.jpg" height="450px" class="center"></p>
</div>
<div id="canvas" class="section level3">
<h3 class="hasAnchor">
<a href="#canvas" class="anchor"></a>Canvas</h3>
<p>Use <code>to canvas</code> button to send plot screenshots to Canvas for later image editing.</p>
<p><img src="images/to_canvas.png"></p>
<p>Use the interactive Canvas image tool to edit plots made from other tabs/modules or local uploads.</p>
<p><img src="images/drawer_demo.png" height="400px" class="center"></p>
</div>
<div id="create-custom-tabs" class="section level3">
<h3 class="hasAnchor">
<a href="#create-custom-tabs" class="anchor"></a>Create custom tabs</h3>
<p>Create your own plotting tab as simple as <code><a href="https://rdrr.io/pkg/systemPipeShiny/man/spsNewTab.html">systemPipeShiny::spsNewTab("TAB_ID")</a></code></p>
<ul>
<li><p>No shiny develop experience is required.</p></li>
<li><p>You only need to know how to code in R.</p></li>
<li><p>Focus on the <code>plot_code</code>, we handle the rest for you.</p></li>
</ul>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># simple template</span>
<span class="va">newtabUI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeShiny/man/spsEzUI.html">spsEzUI</a></span><span class="op">(</span>
    desc <span class="op">=</span> <span class="st">"
    Some description
    "</span>,
    tab_title <span class="op">=</span> <span class="st">"My custom plotting tab"</span>,
    plot_title <span class="op">=</span> <span class="st">"My Plot"</span>,
    plot_control <span class="op">=</span>  <span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/htmltools/man/tag.html">tagList</a></span><span class="op">(</span>
        <span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/fluidPage.html">fluidRow</a></span><span class="op">(</span>class <span class="op">=</span> <span class="st">"center-child"</span>,
            <span class="fu">clearableTextInput</span><span class="op">(</span>inputId <span class="op">=</span> <span class="fu">ns</span><span class="op">(</span><span class="st">"plot_title"</span><span class="op">)</span>, label <span class="op">=</span> <span class="st">"Plot title"</span>, value <span class="op">=</span> <span class="st">"Example plot"</span><span class="op">)</span>
        <span class="op">)</span> <span class="op">%&gt;%</span>
            <span class="fu">bsHoverPopover</span><span class="op">(</span><span class="st">"Plot title"</span>, <span class="st">"Type your plot title"</span>, placement <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span>
    <span class="op">)</span>
<span class="op">)</span>
<span class="va">newtabServer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeShiny/man/spsEzUI.html">spsEzServer</a></span><span class="op">(</span>
    plot_code <span class="op">=</span> <span class="op">{</span>
        <span class="va">plot_data</span> <span class="op">&lt;-</span> <span class="va">mydata</span><span class="op">$</span><span class="va">data</span>
        <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">plot_data</span><span class="op">)</span> <span class="op">+</span>
            <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sepal.Length</span>, y <span class="op">=</span> <span class="va">Sepal.Width</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
            <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="va">input</span><span class="op">$</span><span class="va">plot_title</span><span class="op">)</span>
    <span class="op">}</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="custom-tabs-when-loaded" class="section level3">
<h3 class="hasAnchor">
<a href="#custom-tabs-when-loaded" class="anchor"></a>Custom tabs when loaded</h3>
<p><img src="images/sps_custom_tab_ui.png" height="500px" class="center"></p>
</div>
<div id="deep-customize-the-app" class="section level3">
<h3 class="hasAnchor">
<a href="#deep-customize-the-app" class="anchor"></a>Deep customize the app</h3>
<p>Customize almost everything of the app with these options. Read more on our <a href="https://systempipe.org/sps/adv_features/config/#app-options">website{blk}</a></p>
<p><img src="images/app_options.png" height="550px" class="center"></p>
</div>
<div id="app-protection-with-user-login-and-loading-animations" class="section level3">
<h3 class="hasAnchor">
<a href="#app-protection-with-user-login-and-loading-animations" class="anchor"></a>App protection with user login and loading animations</h3>
<p>Limit access with user accounts.</p>
<p><img src="images/user_login.png" height="300px" class="center"></p>
<p>Adding loading themes before users login.</p>
<p><img src="images/loading_theme.gif" height="300px" class="center"></p>
</div>
<div id="admin" class="section level3">
<h3 class="hasAnchor">
<a href="#admin" class="anchor"></a>Admin</h3>
<p>Customizable hidden url to the admin page.</p>
<p><img src="images/admin_login.png" height="400px" class="center"></p>
</div>
<div id="admin-server-stats" class="section level3">
<h3 class="hasAnchor">
<a href="#admin-server-stats" class="anchor"></a>Admin server stats</h3>
<p>Monitor system usage.</p>
<p><img src="images/admin_server_info.png" height="400px" class="center"></p>
</div>
<div id="admin-user-control" class="section level3">
<h3 class="hasAnchor">
<a href="#admin-user-control" class="anchor"></a>Admin user control</h3>
<p>Add/change/remove users for your SPS app.</p>
<p><img src="images/admin_user_control.png" height="400px" class="center"></p>
</div>
</div>
<div id="sps-supporting-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#sps-supporting-packages" class="anchor"></a>SPS Supporting packages</h2>
<p>Use these supporting packages in your own Rmarkdown, Shiny apps!</p>
<table class="table">
<thead><tr class="header">
<th>Package</th>
<th>Description</th>
<th>Documents</th>
<th align="center">Function reference</th>
<th>Demo</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<img src="https://github.com/systemPipeR/systemPipeR.github.io/blob/main/static/images/sps_small.png?raw=true" align="right" height="30" width="30"><a href="https://github.com/systemPipeR/systemPipeShiny">systemPipeShiny</a>
</td>
<td>SPS main package</td>
<td><a href="https://systempipe.org/sps/">website</a></td>
<td align="center"><a href="https://systempipe.org/sps/funcs/sps/reference/">link</a></td>
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny/">demo</a></td>
</tr>
<tr class="even">
<td>
<img src="https://github.com/systemPipeR/systemPipeR.github.io/blob/main/static/images/spscomps.png?raw=true" align="right" height="30" width="30"><a href="https://github.com/lz100/spsComps">spsComps</a>
</td>
<td>SPS UI and server components</td>
<td><a href="https://systempipe.org/sps/dev/spscomps/">website</a></td>
<td align="center"><a href="https://systempipe.org/sps/funcs/spscomps/reference/">link</a></td>
<td><a href="https://lezhang.shinyapps.io/spsComps">demo</a></td>
</tr>
<tr class="odd">
<td>
<img src="https://github.com/systemPipeR/systemPipeR.github.io/blob/main/static/images/drawer.png?raw=true" align="right" height="30" width="30"><a href="https://github.com/lz100/drawer">drawer</a>
</td>
<td>SPS interactive image editing tool</td>
<td><a href="https://systempipe.org/sps/dev/drawer/">website</a></td>
<td align="center"><a href="https://systempipe.org/sps/funcs/drawer/reference/">link</a></td>
<td><a href="https://lezhang.shinyapps.io/drawer">demo</a></td>
</tr>
<tr class="even">
<td>
<img src="https://github.com/systemPipeR/systemPipeR.github.io/blob/main/static/images/spsutil.png?raw=true" align="right" height="30" width="30"><a href="https://github.com/lz100/spsUtil">spsUtil</a>
</td>
<td>SPS utility functions</td>
<td><a href="https://systempipe.org/sps/dev/spsutil/">website</a></td>
<td align="center"><a href="https://systempipe.org/sps/funcs/spsutil/reference/">link</a></td>
<td>NA</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="more-information" class="section level1">
<h1 class="hasAnchor">
<a href="#more-information" class="anchor"></a>More information</h1>
<ul>
<li><a href="https://systempipe.org/">systemPipe Workflow Environment WebPage{blk}</a></li>
<li><a href="https://systempipe.org/sp/">systemPipeR Manuals{blk}</a></li>
<li><a href="https://systempipe.org/sps/">systemPipeShiny Manuals{blk}</a></li>
<li><a href="https://github.com/systemPipeR">systemPipe Github{blk}</a></li>
</ul>
</div>
<div id="version-information" class="section level1">
<h1 class="hasAnchor">
<a href="#version-information" class="anchor"></a>Version information</h1>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.0 (2021-05-18)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.2 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0</span>
<span class="co">#&gt; LAPACK: /home/dcassol/src/R-4.1.0/lib/libRlapack.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] systemPipeRdata_1.21.3      systemPipeShiny_1.3.2      </span>
<span class="co">#&gt;  [3] drawer_0.1.0                spsComps_0.3.0             </span>
<span class="co">#&gt;  [5] spsUtil_0.1.2               shiny_1.6.0                </span>
<span class="co">#&gt;  [7] systemPipeR_1.27.21         ShortRead_1.51.0           </span>
<span class="co">#&gt;  [9] GenomicAlignments_1.29.0    SummarizedExperiment_1.23.1</span>
<span class="co">#&gt; [11] Biobase_2.53.0              MatrixGenerics_1.5.2       </span>
<span class="co">#&gt; [13] matrixStats_0.60.0          BiocParallel_1.27.2        </span>
<span class="co">#&gt; [15] Rsamtools_2.9.1             Biostrings_2.61.2          </span>
<span class="co">#&gt; [17] XVector_0.33.0              GenomicRanges_1.45.0       </span>
<span class="co">#&gt; [19] GenomeInfoDb_1.29.3         IRanges_2.27.0             </span>
<span class="co">#&gt; [21] S4Vectors_0.31.0            BiocGenerics_0.39.1        </span>
<span class="co">#&gt; [23] BiocStyle_2.21.3           </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] GOstats_2.59.0           backports_1.2.1          BiocFileCache_2.1.1     </span>
<span class="co">#&gt;   [4] systemfonts_1.0.2        lazyeval_0.2.2           shinydashboard_0.7.1    </span>
<span class="co">#&gt;   [7] GSEABase_1.55.1          splines_4.1.0            ggplot2_3.3.5           </span>
<span class="co">#&gt;  [10] digest_0.6.27            htmltools_0.5.1.1        GO.db_3.13.0            </span>
<span class="co">#&gt;  [13] fansi_0.5.0              magrittr_2.0.1           checkmate_2.0.0         </span>
<span class="co">#&gt;  [16] memoise_2.0.0            BSgenome_1.61.0          base64url_1.4           </span>
<span class="co">#&gt;  [19] remotes_2.4.0            tzdb_0.1.2               limma_3.49.2            </span>
<span class="co">#&gt;  [22] shinyFiles_0.9.0         annotate_1.71.0          vroom_1.5.4             </span>
<span class="co">#&gt;  [25] askpass_1.1              pkgdown_1.6.1            prettyunits_1.1.1       </span>
<span class="co">#&gt;  [28] jpeg_0.1-9               colorspace_2.0-2         blob_1.2.2              </span>
<span class="co">#&gt;  [31] rappdirs_0.3.3           textshaping_0.3.5        xfun_0.24               </span>
<span class="co">#&gt;  [34] dplyr_1.0.7              crayon_1.4.1             RCurl_1.98-1.3          </span>
<span class="co">#&gt;  [37] jsonlite_1.7.2           graph_1.71.2             genefilter_1.75.0       </span>
<span class="co">#&gt;  [40] brew_1.0-6               survival_3.2-11          VariantAnnotation_1.39.0</span>
<span class="co">#&gt;  [43] glue_1.4.2               gtable_0.3.0             zlibbioc_1.39.0         </span>
<span class="co">#&gt;  [46] DelayedArray_0.19.1      Rgraphviz_2.37.0         scales_1.1.1            </span>
<span class="co">#&gt;  [49] pheatmap_1.0.12          DBI_1.1.1                edgeR_3.35.0            </span>
<span class="co">#&gt;  [52] Rcpp_1.0.7               viridisLite_0.4.0        xtable_1.8-4            </span>
<span class="co">#&gt;  [55] progress_1.2.2           bit_4.0.4                DT_0.18                 </span>
<span class="co">#&gt;  [58] AnnotationForge_1.35.0   shinyjqui_0.4.0          htmlwidgets_1.5.3       </span>
<span class="co">#&gt;  [61] httr_1.4.2               shinyAce_0.4.1           RColorBrewer_1.1-2      </span>
<span class="co">#&gt;  [64] ellipsis_0.3.2           shinydashboardPlus_2.0.2 pkgconfig_2.0.3         </span>
<span class="co">#&gt;  [67] XML_3.99-0.6             sass_0.4.0               dbplyr_2.1.1            </span>
<span class="co">#&gt;  [70] locfit_1.5-9.4           utf8_1.2.2               tidyselect_1.1.1        </span>
<span class="co">#&gt;  [73] rlang_0.4.11             later_1.2.0              AnnotationDbi_1.55.1    </span>
<span class="co">#&gt;  [76] munsell_0.5.0            tools_4.1.0              cachem_1.0.5            </span>
<span class="co">#&gt;  [79] generics_0.1.0           RSQLite_2.2.7            evaluate_0.14           </span>
<span class="co">#&gt;  [82] stringr_1.4.0            fastmap_1.1.0            yaml_2.2.1              </span>
<span class="co">#&gt;  [85] ragg_1.1.3               knitr_1.33               bit64_4.0.5             </span>
<span class="co">#&gt;  [88] fs_1.5.0                 purrr_0.3.4              KEGGREST_1.33.0         </span>
<span class="co">#&gt;  [91] RBGL_1.69.0              mime_0.11                xml2_1.3.2              </span>
<span class="co">#&gt;  [94] biomaRt_2.49.2           rstudioapi_0.13          debugme_1.1.0           </span>
<span class="co">#&gt;  [97] compiler_4.1.0           plotly_4.9.4.1           filelock_1.0.2          </span>
<span class="co">#&gt; [100] curl_4.3.2               png_0.1-7                tibble_3.1.3            </span>
<span class="co">#&gt; [103] bslib_0.2.5.1            stringi_1.7.3            highr_0.9               </span>
<span class="co">#&gt; [106] GenomicFeatures_1.45.0   desc_1.3.0               lattice_0.20-44         </span>
<span class="co">#&gt; [109] Matrix_1.3-4             styler_1.5.1             shinyjs_2.0.0           </span>
<span class="co">#&gt; [112] vctrs_0.3.8              pillar_1.6.2             lifecycle_1.0.0         </span>
<span class="co">#&gt; [115] BiocManager_1.30.16      jquerylib_0.1.4          data.table_1.14.0       </span>
<span class="co">#&gt; [118] bitops_1.0-7             httpuv_1.6.1             rtracklayer_1.53.0      </span>
<span class="co">#&gt; [121] R6_2.5.0                 BiocIO_1.3.0             latticeExtra_0.6-29     </span>
<span class="co">#&gt; [124] hwriter_1.3.2            promises_1.2.0.1         bookdown_0.22           </span>
<span class="co">#&gt; [127] assertthat_0.2.1         openssl_1.4.4            Category_2.59.0         </span>
<span class="co">#&gt; [130] rprojroot_2.0.2          rjson_0.2.20             shinyWidgets_0.6.0      </span>
<span class="co">#&gt; [133] withr_2.4.2              batchtools_0.9.15        GenomeInfoDbData_1.2.6  </span>
<span class="co">#&gt; [136] parallel_4.1.0           hms_1.1.0                grid_4.1.0              </span>
<span class="co">#&gt; [139] tidyr_1.1.3              rmarkdown_2.9            shinytoastr_2.1.1       </span>
<span class="co">#&gt; [142] restfulr_0.0.13</span></code></pre></div>
</div>
<div id="funding" class="section level1">
<h1 class="hasAnchor">
<a href="#funding" class="anchor"></a>Funding</h1>
<p>This project is funded by NSF award <a href="https://www.nsf.gov/awardsearch/showAward?AWD_ID=1661152">ABI-1661152</a>.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Amstutz2016-ka" class="csl-entry">
Amstutz, Peter, Michael R Crusoe, Nebojša Tijanić, Brad Chapman, John Chilton, Michael Heuer, Andrey Kartashov, et al. 2016. <span>“Common Workflow Language, V1.0,”</span> July. <a href="https://doi.org/10.6084/m9.figshare.3115156.v2">https://doi.org/10.6084/m9.figshare.3115156.v2</a>.
</div>
<div id="ref-Bolger2014-yr" class="csl-entry">
Bolger, Anthony M, Marc Lohse, and Bjoern Usadel. 2014. <span>“Trimmomatic: A Flexible Trimmer for Illumina Sequence Data.”</span> <em>Bioinformatics</em> 30 (15): 2114–20.
</div>
<div id="ref-H_Backman2016-bt" class="csl-entry">
H Backman, Tyler W, and Thomas Girke. 2016. <span>“<span class="nocase">systemPipeR: NGS workflow and report generation environment</span>.”</span> <em>BMC Bioinformatics</em> 17 (1): 388. <a href="https://doi.org/10.1186/s12859-016-1241-0">https://doi.org/10.1186/s12859-016-1241-0</a>.
</div>
<div id="ref-Kim2015-ve" class="csl-entry">
Kim, Daehwan, Ben Langmead, and Steven L Salzberg. 2015. <span>“<span>HISAT</span>: A Fast Spliced Aligner with Low Memory Requirements.”</span> <em>Nat. Methods</em> 12 (4): 357–60.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Daniela Cassol, Le Zhang, Thomas Girke.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
