<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>systemPipe: Workflow and Visualization Toolkit • systemPipeWorkshop2021</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="systemPipe: Workflow and Visualization Toolkit">
<meta property="og:description" content="systemPipeWorkshop2021">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeWorkshop2021</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/systemPipeWorkshop2021.html">Workshop</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/systemPipeR/systemPipeWorkshop2021">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://hub.docker.com/r/systempipe/systempipeworkshop2021">
    <span class="fab fa-docker"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="systemPipeWorkshop2021_files/header-attrs-2.9/header-attrs.js"></script><script src="systemPipeWorkshop2021_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="systemPipeWorkshop2021_files/jquery-3.5.1/jquery.min.js"></script><link href="systemPipeWorkshop2021_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="systemPipeWorkshop2021_files/datatables-binding-0.18/datatables.js"></script><link href="systemPipeWorkshop2021_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="systemPipeWorkshop2021_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="systemPipeWorkshop2021_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="systemPipeWorkshop2021_files/dt-ext-fixedcolumns-1.10.20/css/fixedColumns.dataTables.min.css" rel="stylesheet">
<script src="systemPipeWorkshop2021_files/dt-ext-fixedcolumns-1.10.20/js/dataTables.fixedColumns.min.js"></script><link href="systemPipeWorkshop2021_files/dt-ext-scroller-1.10.20/css/scroller.dataTables.min.css" rel="stylesheet">
<script src="systemPipeWorkshop2021_files/dt-ext-scroller-1.10.20/js/dataTables.scroller.min.js"></script><link href="systemPipeWorkshop2021_files/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="systemPipeWorkshop2021_files/crosstalk-1.1.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>systemPipe: Workflow and Visualization Toolkit</h1>
                        <h4 class="author"></h4>
            
            <h4 class="date">Last update: 01 August, 2021</h4>
      
      
      <div class="hidden name"><code>systemPipeWorkshop2021.Rmd</code></div>

    </div>

    
    
<p><strong>Authors</strong>: Daniela Cassol (<a href="mailto:danielac@ucr.edu" class="email">danielac@ucr.edu</a>), Le Zhang (<a href="mailto:le.zhang001@email.ucr.edu" class="email">le.zhang001@email.ucr.edu</a>), Thomas Girke (<a href="mailto:thomas.girke@ucr.edu" class="email">thomas.girke@ucr.edu</a>).</p>
<p><strong>Institution</strong>: Institute for Integrative Genome Biology, University of California, Riverside, California, USA.</p>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<div id="workshop-description" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h2>
<p>This workshop introduces <em>systemPipe</em> (SP), a generic toolkit for designing and running reproducible data analysis workflows. The environment consists of three major modules implemented as R/Bioconductor packages. <em>systemPipeR</em> (SPR) provides core functionalities for defining workflows, interacting with command-line software, and executing both R and/or command-line software, as well as generating publication-quality analysis reports. <em>systemPipeShiny</em> (SPS) integrates a graphical user interface for managing workflows and visualizing results interactively. <em>systemPipeWorkflow</em> (SPW) offers a collection of pre-configured workflow templates. This hand-on event will include the following topics: (1) brief overview of the design principles and functionalities of the SP toolkit; (2) design and usage of SPR’s command-line interface based on an object-oriented R implementation of CWL; (3) configuration and execution of workflows; (4) construction of custom workflows; (5) configuration and execution of a pre-configured workflow example from start to finish, e.g. RNA-Seq template; (6) parallel execution of workflows on HPC and cloud systems with and without schedulers; (7) generation of technical and scientific analysis reports including visualization; and (8) demonstration of SPS’ core functionalities, the project’s Shiny App.</p>
</div>
<div id="pre-requisites" class="section level2">
<h2 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h2>
<ul>
<li>Basic knowledge of R and usage of Bioconductor packages for NGS analysis</li>
<li>Basic knowledge of running command-line software</li>
<li>Basic knowledge of parallelization concepts</li>
</ul>
<p>Non-essential background reading:</p>
<ul>
<li><a href="https://systempipe.org/">systemPipe Workflow Environment WebPage</a></li>
<li><a href="https://bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html">systemPipeR vignette</a></li>
<li><a href="https://bioconductor.org/packages/devel/bioc/vignettes/systemPipeShiny/inst/doc/systemPipeShiny.html">systemPipeShiny vignette</a></li>
<li><a href="https://bioconductor.org/packages/release/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRdata.html">systemPipeRdata vignette</a></li>
<li><a href="https://rmarkdown.rstudio.com/lesson-2.html">R Markdown tutorial</a></li>
</ul>
</div>
<div id="workshop-participation" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h2>
<p>Participants will be able to perform all analysis components of this workshop hands-on. Active user participation throughout the event is highly encouraged, including but not limited to lecture material, hands-on sections, and final discussion about package improvements. Participants are encouraged to ask questions at any time during the workshop.</p>
</div>
<div id="r-bioconductor-packages-used" class="section level2">
<h2 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h2>
<ul>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/systemPipeR.html"><code>systemPipeR</code></a></li>
<li><a href="https://bioconductor.org/packages/devel/bioc/html/systemPipeShiny.html"><code>systemPipeShiny</code></a></li>
<li><a href="http://www.bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html"><code>systemPipeRdata</code></a></li>
</ul>
</div>
<div id="time-outline" class="section level2">
<h2 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h2>
<p>1h 45m total</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Overview of <em>systemPipe</em> toolkit</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Introduction to SPR’s command-line interface</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Configuration and execution of workflows</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Construction of custom workflows</td>
<td>10m</td>
</tr>
<tr class="odd">
<td>Showcase RNA-Seq workflow</td>
<td>20m</td>
</tr>
<tr class="even">
<td>Parallelization on single machines and clusters</td>
<td>10m</td>
</tr>
<tr class="odd">
<td>Generation of technical and scientific analysis reports</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Overview of <em>systemPipeShiny</em> core functionalities</td>
<td>20m</td>
</tr>
</tbody>
</table>
</div>
<div id="workshop-goals-and-objectives" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h2>
<div id="learning-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h3>
<ul>
<li>Recognize the benefits of a generic R-based workflow construction environment that is both scalable and reproducible</li>
<li>Integration of command-line tools via the CWL community standard</li>
<li>Rendering of R markdown reports and critical assessment of scientific analysis reports</li>
<li>Parallelization of big data analysis tasks</li>
</ul>
</div>
<div id="learning-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h3>
<ul>
<li>Identify and practice how to make analysis workflows more robust, reproducible, and portable across heterogeneous computing systems</li>
<li>Usage of new workflow control class for designing, configuring, and running workflows</li>
<li>Optimize and debug workflows</li>
<li>Inspection of technical reports and log files</li>
<li>Design of new and fully customized workflows</li>
<li>Practice interactive workflow management and visualization</li>
</ul>
</div>
</div>
</div>
<div id="workshop" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop" class="anchor"></a>Workshop</h1>
<div id="running-the-workshop" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-workshop" class="anchor"></a>Running the Workshop</h2>
<p>This workshop uses R <code>4.1.0</code> and Bioconductor version<code>3.14</code>. Bioconductor can be installed following <a href="https://www.bioconductor.org/developers/how-to/useDevel/">these instructions</a>.</p>
<p>During the <a href="https://bioc2021.bioconductor.org/">Bioc2021 conference</a>, the workshop can be run in the <a href="http://app.orchestra.cancerdatasci.org/">cloud</a>.</p>
<div id="workshop-setup-with-docker" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-setup-with-docker" class="anchor"></a>Workshop setup with Docker</h3>
<p>The workshop Docker container runs with Bioconductor <code>devel</code> version <code>3.14</code> and has all the necessary packages and software for running the workshop vignettes. To use Docker container, you need to first install <a href="https://docs.docker.com/engine/install/">Docker</a> on your system.</p>
<ul>
<li>The container can be downloaded and run with:</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-e</span> PASSWORD=systempipe <span class="at">-p</span> 8787:8787 systempipe/systempipeworkshop2021:latest</span></code></pre></div>
<ul>
<li><p>Log in to RStudio at <a href="http://localhost:8787">http://localhost:8787</a> using username <code>rstudio</code> and password <code>systempipe</code>.http://vbdukvrh.orchestra.cancerdatasci.org/</p></li>
<li><p>If you prefer to run the workshop from the command line:</p></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-it</span> <span class="at">--user</span> rstudio systempipe/systempipeworkshop2021:latest bash </span></code></pre></div>
</div>
<div id="workshop-setup-with-github" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-setup-with-github" class="anchor"></a>Workshop setup with GitHub</h3>
<p><a href="http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html"><em><code>systemPipeR</code></em></a> and <a href="http://bioconductor.org/packages/devel/bioc/html/systemPipeShiny.html"><em><code>systemPipeShiny</code></em></a> environment can be installed from the R console using the <a href="https://cran.r-project.org/web/packages/BiocManager/index.html"><em><code>BiocManager::install</code></em></a> command. The associated data package <a href="http://www.bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html"><em><code>systemPipeRdata</code></em></a> can be installed the same way. The latter is a helper package for generating <em><code>systemPipeR</code></em> workflow environments with a single command containing all parameter files and sample data required to quickly test and run workflows.</p>
<p>You can install locally the necessary packages for this workshop as follows:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Install workshop package</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"systemPipeR/systemPipeWorkshop2021"</span><span class="op">)</span>
<span class="co">## Install required packages</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"systemPipeR"</span>, <span class="st">"systemPipeRdata"</span>, <span class="st">"systemPipeShiny"</span><span class="op">)</span>, version<span class="op">=</span><span class="st">"3.14"</span><span class="op">)</span></code></pre></div>
<p>To access the vignette:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/browseVignettes.html">browseVignettes</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"systemPipeWorkshop2021"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting Started</h2>
<div id="loading-package-and-documentation" class="section level3">
<h3 class="hasAnchor">
<a href="#loading-package-and-documentation" class="anchor"></a>Loading package and documentation</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://systempipe.org/">"systemPipeR"</a></span><span class="op">)</span> 
<span class="co">#&gt; Loading required package: Rsamtools</span>
<span class="co">#&gt; Loading required package: GenomeInfoDb</span>
<span class="co">#&gt; Loading required package: BiocGenerics</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'BiocGenerics'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     IQR, mad, sd, var, xtabs</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
<span class="co">#&gt;     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
<span class="co">#&gt;     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
<span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
<span class="co">#&gt;     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
<span class="co">#&gt;     union, unique, unsplit, which.max, which.min</span>
<span class="co">#&gt; Loading required package: S4Vectors</span>
<span class="co">#&gt; Loading required package: stats4</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'S4Vectors'</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand.grid, I, unname</span>
<span class="co">#&gt; Loading required package: IRanges</span>
<span class="co">#&gt; Loading required package: GenomicRanges</span>
<span class="co">#&gt; Loading required package: Biostrings</span>
<span class="co">#&gt; Loading required package: XVector</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'Biostrings'</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     strsplit</span>
<span class="co">#&gt; Loading required package: ShortRead</span>
<span class="co">#&gt; Loading required package: BiocParallel</span>
<span class="co">#&gt; Loading required package: GenomicAlignments</span>
<span class="co">#&gt; Loading required package: SummarizedExperiment</span>
<span class="co">#&gt; Loading required package: MatrixGenerics</span>
<span class="co">#&gt; Loading required package: matrixStats</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'MatrixGenerics'</span>
<span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span>
<span class="co">#&gt; Loading required package: Biobase</span>
<span class="co">#&gt; Welcome to Bioconductor</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Vignettes contain introductory material; view with</span>
<span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'Biobase'</span>
<span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     rowMedians</span>
<span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyMissing, rowMedians</span>
<span class="co">#&gt; Loading required package: DT</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'systemPipeR'</span>
<span class="co">#&gt; The following object is masked from 'package:BiocStyle':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     output</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://systempipe.org/sps">"systemPipeShiny"</a></span><span class="op">)</span> 
<span class="co">#&gt; Loading required package: shiny</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'shiny'</span>
<span class="co">#&gt; The following objects are masked from 'package:DT':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     dataTableOutput, renderDataTable</span>
<span class="co">#&gt; The following object is masked from 'package:BiocStyle':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     markdown</span>
<span class="co">#&gt; Loading required package: spsUtil</span>
<span class="co">#&gt; Loading required package: spsComps</span>
<span class="co">#&gt; Loading required package: drawer</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/tgirke/systemPipeRdata">"systemPipeRdata"</a></span><span class="op">)</span> </code></pre></div>
<ul>
<li>Documentation</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span>help<span class="op">=</span><span class="st">"systemPipeR"</span><span class="op">)</span> <span class="co"># Lists package info</span>
<span class="fu"><a href="https://rdrr.io/r/utils/vignette.html">vignette</a></span><span class="op">(</span><span class="st">"systemPipeR"</span><span class="op">)</span> <span class="co"># Opens vignette</span></code></pre></div>
</div>
<div id="set-working-directory" class="section level3">
<h3 class="hasAnchor">
<a href="#set-working-directory" class="anchor"></a>Set working directory</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"vignettes"</span><span class="op">)</span></code></pre></div>
</div>
<div id="how-to-get-help-for-systempipersystempipeshiny" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-get-help-for-systempipersystempipeshiny" class="anchor"></a>How to get help for systemPipeR/systemPipeShiny</h3>
<p>All questions about the package or any particular function should be posted to the Bioconductor support site <a href="https://support.bioconductor.org">https://support.bioconductor.org</a>.</p>
<p>Please add the “<em><code>systemPipeR</code></em>” tag to your question, and the package authors will automatically receive an alert.</p>
<p>We appreciate receiving reports of bugs in the functions or documentation and suggestions for improvement. For that, please consider opening an issue at <a href="https://github.com/systemPipeR">GitHub</a>.</p>
</div>
</div>
<div id="overview-of-systempipe-toolkit" class="section level2">
<h2 class="hasAnchor">
<a href="#overview-of-systempipe-toolkit" class="anchor"></a>Overview of <em>systemPipe</em> toolkit</h2>
<p><code>systemPipe</code> (SP) is a generic toolkit for designing and running reproducible data analysis workflows. The environment consists of three major modules implemented as R/Bioconductor packages:</p>
<ul>
<li><p><code>systemPipeR</code> (SPR) provides core functionalities for defining workflows, interacting with command-line software, and executing both R and/or command-line software, as well as generating publication-quality analysis reports.</p></li>
<li><p><code>systemPipeShiny</code> (SPS) integrates a graphical user interface for managing workflows and visualizing results interactively.</p></li>
<li><p><code>systemPipeWorkflow</code> (SPW) offers a collection of pre-configured workflow templates.</p></li>
</ul>
</div>
<div id="introduction-to-sprs-command-line-interface" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction-to-sprs-command-line-interface" class="anchor"></a>Introduction to SPR’s command-line interface</h2>
<p>A central concept for designing workflows within the <code>systemPipeR</code> environment is the use of workflow management containers. <code>systemPipeR</code> adopted the widely used community standard <a href="https://www.commonwl.org/">Common Workflow Language</a> (CWL) <span class="citation">[@Amstutz2016-ka]</span> for describing analysis workflows in a generic and reproducible manner. Using this community standard in <code>systemPipeR</code> has many advantages. For instance, the integration of CWL allows running <code>systemPipeR</code> workflows from a single specification instance either entirely from within R, from various command line wrappers (e.g., cwl-runner) or from other languages (, e.g., Bash or Python). <code>systemPipeR</code> includes support for both command line and R/Bioconductor software as well as resources for containerization, parallel evaluations on computer clusters along with the automated generation of interactive analysis reports.</p>
<p>An important feature of <code>systemPipeR's</code> CWL interface is that it provides two options to run command line tools and workflows based on CWL. First, one can run CWL in its native way via an R-based wrapper utility for <code>cwl-runner</code> or <code>cwl-tools</code> (CWL-based approach). Second, one can run workflows using CWL’s command line and workflow instructions from within R (R-based approach). In the latter case the same CWL workflow definition files (e.g. <em>.cwl</em> and <em>.yml</em>) are used but rendered and executed entirely with R functions defined by <code>systemPipeR</code>, and thus use CWL mainly as a command line and workflow definition format rather than software to run workflows. In this regard <code>systemPipeR</code> also provides several convenience functions that are useful for designing and debugging workflows, such as a command line rendering function to retrieve the exact command line strings for each data set and processing step prior to running a command line.</p>
</div>
<div id="configuration-and-execution-of-workflows" class="section level2">
<h2 class="hasAnchor">
<a href="#configuration-and-execution-of-workflows" class="anchor"></a>Configuration and execution of workflows</h2>
<div id="load-sample-data-and-directory-structure" class="section level3">
<h3 class="hasAnchor">
<a href="#load-sample-data-and-directory-structure" class="anchor"></a>Load sample data and directory structure</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">systemPipeRdata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span><span class="st">"new"</span>, mydirname <span class="op">=</span> <span class="st">"bioc2021"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Generated bioc2021 directory. Next run in new directory, the R code from *.Rmd template interactively. Alternatively, workflows can be exectued with a single command as instructed in the vignette."</span></code></pre></div>
</div>
<div id="project-structure" class="section level3">
<h3 class="hasAnchor">
<a href="#project-structure" class="anchor"></a>Project structure</h3>
<p><em><code>systemPipeR</code></em> expects a project directory structure that consists of a directory where users may store all the raw data, the results directory that will be reserved for all the outfiles files or new output folders, and the parameters directory.</p>
<p>This structure allows reproducibility and collaboration across the data science team since internally relative paths are used. Users could transfer this project to a different location and still be able to run the entire workflow. Also, it increases efficiency and data management once the raw data is kept in a separate folder and avoids duplication.</p>
<div id="dir" class="section level4">
<h4 class="hasAnchor">
<a href="#dir" class="anchor"></a>Directory Structure</h4>
<p><a href="http://bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html"><em><code>systemPipeRdata</code></em></a>, helper package, provides pre-configured workflows, reporting templates, and sample data loaded as demonstrated below. With a single command, the package allows creating the workflow environment containing the structure described here (see Figure @ref(fig:dir)).</p>
<p>Directory names are indicated in <span style="color:grey"><strong><em>green</em></strong></span>. Users can change this structure as needed, but need to adjust the code in their workflows accordingly.</p>
<ul>
<li>
<span style="color:green"><em><strong>workflow/</strong></em></span> (<em>e.g.</em> <em>myproject/</em>)
<ul>
<li>This is the root directory of the R session running the workflow.</li>
<li>Run script ( <em>*.Rmd</em>) and sample annotation (<em>targets.txt</em>) files are located here.</li>
<li>Note, this directory can have any name (<em>e.g.</em> <span style="color:green"><em><strong>myproject</strong></em></span>). Changing its name does not require any modifications in the run script(s).</li>
<li>
<strong>Important subdirectories</strong>:
<ul>
<li>
<span style="color:green"><em><strong>param/</strong></em></span>
<ul>
<li>
<span style="color:green"><em><strong>param/cwl/</strong></em></span>: This subdirectory stores all the parameter and configuration files. To organize workflows, each can have its own subdirectory, where all <code>*.cwl</code> and <code>*input.yml</code> files need to be in the same subdirectory.</li>
</ul>
</li>
<li>
<span style="color:green"><em><strong>data/</strong></em> </span>
<ul>
<li>Raw data (<em>e.g.</em> FASTQ files)</li>
<li>FASTA file of reference (<em>e.g.</em> reference genome)</li>
<li>Annotation files</li>
<li>Metadata</li>
<li>etc.</li>
</ul>
</li>
<li>
<span style="color:green"><em><strong>results/</strong></em></span>
<ul>
<li>Analysis results are usually written to this directory, including: alignment, variant and peak files (BAM, VCF, BED); tabular result files; and image/plot files</li>
<li>Note, the user has the option to organize results files for a given sample and analysis step in a separate subdirectory.</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="figure" style="text-align: center">
<img src="../../../../usr/local/lib/R/site-library/systemPipeR/extdata/images/spr_project.png" alt="*systemPipeR's* preconfigured directory structure." width="100%"><p class="caption">
<em>systemPipeR’s</em> preconfigured directory structure.
</p>
</div>
</div>
<div id="structure-of-initial-targets-file" class="section level4">
<h4 class="hasAnchor">
<a href="#structure-of-initial-targets-file" class="anchor"></a>Structure of initial <em><code>targets</code></em> file</h4>
<p>The <em><code>targets</code></em> file defines all input files (<em>e.g.</em> FASTQ, BAM, BCF) and sample comparisons of an analysis workflow. The following shows the format of a sample <em><code>targets</code></em> file included in the package. It also can be viewed and downloaded from <em><code>systemPipeR</code></em>’s GitHub repository <a href="https://github.com/tgirke/systemPipeR/blob/master/inst/extdata/targets.txt">here</a>. In a target file with a single type of input files, here FASTQ files of single-end (SE) reads, the first column describe the path and the second column represents a unique id name for each sample. The third column called <code>Factor</code> represents the biological replicates. All subsequent columns are additional information, and any number of extra columns can be added as needed.</p>
<p>Users should note here, the usage of targets files is optional when using <em><code>systemPipeR's</code></em> new workflow management interface. They can be replaced by a standard YAML input file used by CWL. Since for organizing experimental variables targets files are extremely useful and user-friendly. Thus, we encourage users to keep using them.</p>
<ul>
<li>Structure of <em><code>targets</code></em> file for single-end (SE) samples</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="st">"targets.txt"</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/showDF.html">showDF</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">targetspath</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-444ddbe69993eb15e12c" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-444ddbe69993eb15e12c">{"x":{"filter":"none","extensions":["FixedColumns","Scroller"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18"],["./data/SRR446027_1.fastq.gz","./data/SRR446028_1.fastq.gz","./data/SRR446029_1.fastq.gz","./data/SRR446030_1.fastq.gz","./data/SRR446031_1.fastq.gz","./data/SRR446032_1.fastq.gz","./data/SRR446033_1.fastq.gz","./data/SRR446034_1.fastq.gz","./data/SRR446035_1.fastq.gz","./data/SRR446036_1.fastq.gz","./data/SRR446037_1.fastq.gz","./data/SRR446038_1.fastq.gz","./data/SRR446039_1.fastq.gz","./data/SRR446040_1.fastq.gz","./data/SRR446041_1.fastq.gz","./data/SRR446042_1.fastq.gz","./data/SRR446043_1.fastq.gz","./data/SRR446044_1.fastq.gz"],["M1A","M1B","A1A","A1B","V1A","V1B","M6A","M6B","A6A","A6B","V6A","V6B","M12A","M12B","A12A","A12B","V12A","V12B"],["M1","M1","A1","A1","V1","V1","M6","M6","A6","A6","V6","V6","M12","M12","A12","A12","V12","V12"],["Mock.1h.A","Mock.1h.B","Avr.1h.A","Avr.1h.B","Vir.1h.A","Vir.1h.B","Mock.6h.A","Mock.6h.B","Avr.6h.A","Avr.6h.B","Vir.6h.A","Vir.6h.B","Mock.12h.A","Mock.12h.B","Avr.12h.A","Avr.12h.B","Vir.12h.A","Vir.12h.B"],[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],["23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>FileName<\/th>\n      <th>SampleName<\/th>\n      <th>Factor<\/th>\n      <th>SampleLong<\/th>\n      <th>Experiment<\/th>\n      <th>Date<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"fixedColumns":true,"deferRender":true,"scrollY":200,"scroller":true,"columnDefs":[{"className":"dt-right","targets":5},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>To work with custom data, users need to generate a <em><code>targets</code></em> file containing the paths to their own FASTQ files and then provide under <em><code>targetspath</code></em> the path to the corresponding <em><code>targets</code></em> file.</p>
<ul>
<li>Structure of <em><code>targets</code></em> file for “Hello World” example</li>
</ul>
<p>In this example, <em><code>targets</code></em> file presents only two columns, which the first one are the different phrases used by the <code>echo</code> command-line and the second column it is the sample <code>id</code>. The <code>id</code> column is required, and each sample id should be unique.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/example/targets_example.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/showDF.html">showDF</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">targetspath</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-dd056ff38c8bc56d8908" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-dd056ff38c8bc56d8908">{"x":{"filter":"none","extensions":["FixedColumns","Scroller"],"data":[["1","2","3"],["Hello World!","Hello USA!","Hello Bioconductor!"],["M1","M2","M3"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Message<\/th>\n      <th>SampleName<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"fixedColumns":true,"deferRender":true,"scrollY":200,"scroller":true,"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</div>
<div id="structure-of-the-new-parameters-files" class="section level2">
<h2 class="hasAnchor">
<a href="#structure-of-the-new-parameters-files" class="anchor"></a>Structure of the new parameters files</h2>
<p>The parameters and configuration required for running command-line software are provided by the widely used community standard <a href="https://www.commonwl.org/">Common Workflow Language</a> (CWL) <span class="citation">[@Amstutz2016-ka]</span>, which describes parameters analysis workflows in a generic and reproducible manner. For R-based workflow steps, param files are not required. For a complete overview of the CWL syntax, please see the <a href="#cwl">section</a> below. Also, we have a dedicated section explain how to <em><code>systemPipeR</code></em> establish the connection between the CWL parameters files and the targets files. Please see <a href="#cwl_targets">here</a>.</p>
<div id="creating-the-cwl-param-files-from-the-command-line" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-the-cwl-param-files-from-the-command-line" class="anchor"></a>Creating the CWL param files from the command-line</h3>
<p>Users need to define the command-line in a pseudo-bash script format:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># "hisat2 -S ./results/M1A.sam -x ./data/tair10.fasta -k 1 -threads 4 -U ./data/SRR446027_1.fastq.gz "</span>
<span class="va">command</span> <span class="op">&lt;-</span> <span class="st">"
    hisat2 \
    -S &lt;F, out: ./results/M1A.sam&gt; \
    -x &lt;F: ./data/tair10.fasta&gt; \
     -k &lt;int: 1&gt; \
    -threads &lt;int: 4&gt; \
    -U &lt;F: ./data/SRR446027_1.fastq.gz&gt;
"</span></code></pre></div>
<div id="define-prefix-and-defaults" class="section level4">
<h4 class="hasAnchor">
<a href="#define-prefix-and-defaults" class="anchor"></a>Define prefix and defaults</h4>
<ul>
<li><p>First line is the base command. Each line is an argument with its default value.</p></li>
<li><p>For argument lines (starting from the second line), any word before the first space with leading <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code> or <code>--</code> in each will be treated as a prefix, like <code>-S</code> or <code>--min</code>. Any line without this first word will be treated as no prefix.</p></li>
<li><p>All defaults are placed inside <code>&lt;...&gt;</code>.</p></li>
<li><p>First argument is the input argument type. <code>F</code> for “File”, “int”, “string” are unchanged.</p></li>
<li><p>Optional: use the keyword <code>out</code> followed the type with a <code>,</code> comma separation to indicate if this argument is also an CWL output.</p></li>
<li><p>Then, use <code><a href="https://rdrr.io/r/base/Colon.html">:</a></code> to separate keywords and default values, any non-space value after the <code><a href="https://rdrr.io/r/base/Colon.html">:</a></code> will be treated as the default value.</p></li>
<li><p>If any argument has no default value, just a flag, like <code>--verbose</code>, there is no need to add any <code>&lt;...&gt;</code></p></li>
</ul>
</div>
<div id="createparam-function" class="section level4">
<h4 class="hasAnchor">
<a href="#createparam-function" class="anchor"></a><code>createParam</code> Function</h4>
<p><code>createParam</code> function requires the <code>string</code> as defined above as an input.</p>
<p>First of all, the function will print the three components of the <code>cwl</code> file: - <code>BaseCommand</code>: Specifies the program to execute. - <code>Inputs</code>: Defines the input parameters of the process. - <code>Outputs</code>: Defines the parameters representing the output of the process.</p>
<p>The four component is the original command-line.</p>
<p>If in interactive mode, the function will verify that everything is correct and will ask you to proceed. Here, the user can answer “no” and provide more information at the string level. Another question is to save the param created here.</p>
<p>If running the workflow in non-interactive mode, the <code>createParam</code> function will consider “yes” and returning the container.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cmd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/createParamFiles.html">createParam</a></span><span class="op">(</span><span class="va">command</span>, writeParamFiles <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; *****BaseCommand*****</span>
<span class="co">#&gt; hisat2 </span>
<span class="co">#&gt; *****Inputs*****</span>
<span class="co">#&gt; S:</span>
<span class="co">#&gt;     type: File</span>
<span class="co">#&gt;     preF: -S</span>
<span class="co">#&gt;     yml: ./results/M1A.sam</span>
<span class="co">#&gt; x:</span>
<span class="co">#&gt;     type: File</span>
<span class="co">#&gt;     preF: -x</span>
<span class="co">#&gt;     yml: ./data/tair10.fasta</span>
<span class="co">#&gt; k:</span>
<span class="co">#&gt;     type: int</span>
<span class="co">#&gt;     preF: -k</span>
<span class="co">#&gt;     yml: 1</span>
<span class="co">#&gt; threads:</span>
<span class="co">#&gt;     type: int</span>
<span class="co">#&gt;     preF: -threads</span>
<span class="co">#&gt;     yml: 4</span>
<span class="co">#&gt; U:</span>
<span class="co">#&gt;     type: File</span>
<span class="co">#&gt;     preF: -U</span>
<span class="co">#&gt;     yml: ./data/SRR446027_1.fastq.gz</span>
<span class="co">#&gt; *****Outputs*****</span>
<span class="co">#&gt; output1:</span>
<span class="co">#&gt;     type: File</span>
<span class="co">#&gt;     value: ./results/M1A.sam</span>
<span class="co">#&gt; *****Parsed raw command line*****</span>
<span class="co">#&gt; hisat2 -S ./results/M1A.sam -x ./data/tair10.fasta -k 1 -threads 4 -U ./data/SRR446027_1.fastq.gz </span>
<span class="co">#&gt;   Written content of 'commandLine' to file: </span>
<span class="co">#&gt;  param/cwl/hisat2/hisat2.cwl </span>
<span class="co">#&gt;   Written content of 'commandLine' to file: </span>
<span class="co">#&gt;  param/cwl/hisat2/hisat2.yml</span></code></pre></div>
</div>
</div>
</div>
</div>
<div id="cwl_targets" class="section level1">
<h1 class="hasAnchor">
<a href="#cwl_targets" class="anchor"></a>How to connect CWL description files within <em><code>systemPipeR</code></em>
</h1>
<p>This section will demonstrate how to connect CWL parameters files to create workflows. In addition, we will show how the workflow can be easily scalable with <em><code>systemPipeR</code></em>.</p>
<div class="figure" style="text-align: center">
<img src="../../../../usr/local/lib/R/site-library/systemPipeR/extdata/images/SPR_CWL_hello.png" alt="WConnectivity between CWL param files and targets files." width="100%"><p class="caption">
WConnectivity between CWL param files and targets files.
</p>
</div>
<p>To create a Workflow within <code>systemPipeR</code>, we can start by defining an empty container and checking the directory structure:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span><span class="op">)</span> </code></pre></div>
<p>Internally, <code>SPRproject</code> function will check and/or create the basic folder structure, which means <code>data</code>, <code>param</code>, and <code>results</code> folder. If the user wants to use a different names for these directories, can be specified as follows:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>projPath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span>, data <span class="op">=</span> <span class="st">"data"</span>, param <span class="op">=</span> <span class="st">"param"</span>, results <span class="op">=</span> <span class="st">"results"</span><span class="op">)</span> </code></pre></div>
<p>Also, this function allows creating a hidden folder called <code>.SPRproject</code>, by default, to store all the log files. A yaml file, here called <code>SYSargsList.yml</code>, has been created, which initially contains the basic location of the project structure; however, every time the object in R is updated, the new information will also be store in this file, for easy recover. If you desire a different names for the logs folder and the yaml files, those can be modify as follows:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>logs.dir<span class="op">=</span> <span class="st">"SPRproject"</span>, sys.file<span class="op">=</span><span class="st">"SPRproject/SYSargsList.yml"</span><span class="op">)</span> </code></pre></div>
<p>In this stage, the object <code>sal</code> is a empty container, except the project information, as the project, data, results and param folder paths, as can be access but the <code>projectInfo</code> accessory method:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">projectInfo</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<p>Also, the length function will return how many steps this workflow contains and in this case it is empty, as follow:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<div id="adding-a-new-step-from-template" class="section level3">
<h3 class="hasAnchor">
<a href="#adding-a-new-step-from-template" class="anchor"></a>Adding a new step from template</h3>
<p>Next, we need to populate the object created with the first step in the workflow. Here, an example of how to perform this task using param template files for trimming FASTQ files with <code>Hisat2</code> software.</p>
<p>The constructor functions create an <code>SYSargsList</code> S4 class object from three input files:</p>
<pre><code>- CWL command-line specification file (`wf_file` argument);
- Input variables (`input_file` argument);
- Targets file (`targets` argument).</code></pre>
<p>The latter is optional for workflow steps lacking input files. The connection between input variables and the targets file are defined under the <code>inputvars</code> argument. It is required a named vector, where each element name needs to match with column names in the targets file and the value must match the names of the <em>.yml</em> variables.</p>
<p>In CWL, files with the extension <code>.cwl</code> define the parameters of a chosen command-line step or workflow, while files with the extension <code>.yml</code> define the input variables of command-line steps. Note, input variables provided by a targets file can be passed on to a instance via the <code>inputvars</code> argument of the <code>SYSargsList</code> function.</p>
<p>For more information of each one of those files, please check here. (Link to tutorial).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"Index"</span>, dir <span class="op">=</span> <span class="cn">FALSE</span>,
                               targets<span class="op">=</span><span class="cn">NULL</span>, 
                               wf_file <span class="op">=</span> <span class="st">"hisat2/hisat2-index.cwl"</span>, input_file<span class="op">=</span><span class="st">"hisat2/hisat2-index.yml"</span>,
                               dir_path<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>,
                               dependency<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">cmdlist</a></span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">SYSargsList</a></span><span class="op">(</span>targets<span class="op">=</span><span class="va">targetspath</span>,
                               step_name<span class="op">=</span><span class="st">"Mapping"</span>, dir<span class="op">=</span><span class="cn">TRUE</span>, 
                               wf_file <span class="op">=</span> <span class="st">"workflow-hisat2/workflow_hisat2-pe.cwl"</span>,
                               input_file<span class="op">=</span><span class="st">"workflow-hisat2/workflow_hisat2-pe.yml"</span>,
                               dir_path<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>,
                               inputvars<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1<span class="op">=</span><span class="st">"_FASTQ_PATH1_"</span>,
                                           FileName2<span class="op">=</span><span class="st">"_FASTQ_PATH2_"</span>, SampleName<span class="op">=</span><span class="st">"_SampleName_"</span><span class="op">)</span>,
                               dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Index"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="accessing-the-details" class="section level3">
<h3 class="hasAnchor">
<a href="#accessing-the-details" class="anchor"></a>Accessing the details</h3>
<p>Let’s explore our object:</p>
<ul>
<li>Accessor Methods:</li>
</ul>
<p>Several accessor methods are available that are named after the slot names of the <code>SYSargsList</code> object.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<ul>
<li>Check the length of the <code>sal</code> object:</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<ul>
<li>The <code>SYSargsList</code> class and its subsetting operator <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code>:</li>
</ul>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">stepsWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<ul>
<li>The <code>SYSargsList</code> class and its subsetting input samples:</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal_sub</span> <span class="op">&lt;-</span> <span class="fu">subsetTargets</span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">stepsWF</a></span><span class="op">(</span><span class="va">sal_sub</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">targetsWF</a></span><span class="op">(</span><span class="va">sal_sub</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">outfiles</a></span><span class="op">(</span><span class="va">sal_sub</span><span class="op">)</span></code></pre></div>
<ul>
<li>The <code>SYSargsList</code> class and its operator <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code>:</li>
</ul>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="va">sal</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></code></pre></div>
<ul>
<li>Checking the expected output files:</li>
</ul>
<p>The <code>outfiles</code> components of <code>SYSargsList</code> define the expected output files for each step in the workflow; some of which are the input for the next workflow step.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">outfiles</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></code></pre></div>
<ul>
<li>Checking the input files by the <code><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">targetsWF()</a></code>:</li>
</ul>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">targetsWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<ul>
<li>Checking the command-line for each sample:</li>
</ul>
<p><code><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">cmdlist()</a></code> method constructs the system commands for running command-line software as specified by a given <code>.cwl</code> file combined with the paths to the input samples (e.g. FASTQ files) provided by a <code>targets</code> file. The example below shows the <code><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">cmdlist()</a></code> output for running Trimmomatic on the first PE read sample. Evaluating the output of <code><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">cmdlist()</a></code> can be very helpful for designing and debugging <code>.cwl</code> files of new command-line software or changing the parameter settings of existing ones.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">cmdlist</a></span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
<ul>
<li>Rename a Step</li>
</ul>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">renameStep</span><span class="op">(</span><span class="va">sal</span>, <span class="fl">1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"newStep"</span>
<span class="fu">renameStep</span><span class="op">(</span><span class="va">sal</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Index"</span>, <span class="st">"Mapping"</span><span class="op">)</span>
<span class="va">sal</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">outfiles</a></span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">targetsWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<ul>
<li>Replace a Step</li>
</ul>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">replaceStep</span><span class="op">(</span><span class="va">sal</span>, <span class="fl">1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sal</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>
<span class="fu">replaceStep</span><span class="op">(</span><span class="va">sal</span>, <span class="fl">1</span>, step_name<span class="op">=</span><span class="st">"Quality"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sal</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>
<span class="va">sal</span></code></pre></div>
<ul>
<li>Removing a Step</li>
</ul>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal_sub</span> <span class="op">&lt;-</span> <span class="va">sal</span><span class="op">[</span><span class="op">-</span><span class="fl">2</span><span class="op">]</span>
<span class="va">sal_sub</span></code></pre></div>
</div>
<div id="adding-a-r-code-as-new-step" class="section level2">
<h2 class="hasAnchor">
<a href="#adding-a-r-code-as-new-step" class="anchor"></a>Adding a R code as new Step</h2>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html">LineWise</a></span><span class="op">(</span><span class="st">"log &lt;- log(1)"</span>, <span class="st">"R_code"</span><span class="op">)</span>
<span class="va">sal</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">stepsWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<ul>
<li>Replacement and Append methods for R Code Steps</li>
</ul>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendCodeLine</span><span class="op">(</span><span class="va">sal</span>, step<span class="op">=</span><span class="fl">3</span>, after<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"log &lt;- log(10)"</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">codeLine</a></span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span>

<span class="fu">replaceCodeLine</span><span class="op">(</span><span class="va">sal</span>, step<span class="op">=</span><span class="fl">3</span>, line<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"log &lt;- log(6)"</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">codeLine</a></span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<div id="running-the-workflow" class="section level3">
<h3 class="hasAnchor">
<a href="#running-the-workflow" class="anchor"></a>Running the workflow</h3>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html">runWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">statusWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="construction-of-custom-workflows" class="section level2">
<h2 class="hasAnchor">
<a href="#construction-of-custom-workflows" class="anchor"></a>Construction of custom workflows</h2>
</div>
<div id="showcase-small-rna-seq-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#showcase-small-rna-seq-workflow" class="anchor"></a>Showcase small RNA-Seq workflow</h2>
</div>
<div id="parallelization-on-single-machines-and-clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#parallelization-on-single-machines-and-clusters" class="anchor"></a>Parallelization on single machines and clusters</h2>
</div>
<div id="generation-of-technical-and-scientific-analysis-reports" class="section level2">
<h2 class="hasAnchor">
<a href="#generation-of-technical-and-scientific-analysis-reports" class="anchor"></a>Generation of technical and scientific analysis reports</h2>
</div>
<div id="overview-of-systempipeshiny-core-functionalities" class="section level2">
<h2 class="hasAnchor">
<a href="#overview-of-systempipeshiny-core-functionalities" class="anchor"></a>Overview of <em>systemPipeShiny</em> core functionalities</h2>
<p><strong><span style="color:#5DA7D6;">s</span>ystem<span style="color:#5DA7D6;">P</span>ipe<span style="color:#5DA7D6;">S</span>hiny</strong> (SPS) extends the widely used <a href="/spr/">systemPipeR</a> (SPR) workflow environment with a versatile graphical user interface provided by a <a href="https://shiny.rstudio.com">Shiny App</a>. This allows non-R users, such as experimentalists, to run many systemPipeR’s workflow designs, control, and visualization functionalities interactively without requiring knowledge of R. Most importantly, <code>SPS</code> has been designed as a general purpose framework for interacting with other R packages in an intuitive manner. Like most Shiny Apps, SPS can be used on both local computers as well as centralized server-based deployments that can be accessed remotely as a public web service for using SPR’s functionalities with community and/or private data. The framework can integrate many core packages from the R/Bioconductor ecosystem. Examples of SPS’ current functionalities include:</p>
<ul>
<li>A default interactive workflow module to create experimental designs, visualize and customize workflow topologies with previews, and programming free workflow execution within the application.</li>
<li>An interactive module with extensive plot options to visualize downstream analysis of a RNA-Seq workflow.</li>
<li>A quick ggplot module to make all variety of scientific plots from any user defined tabular data.</li>
<li>An extendable set of visualization functionalities makes it easy to design custom Shiny Apps under SPS framework without any knowledge of Shiny.</li>
<li>A ‘Canvas Workbench’ to manage complex visual results. It allows users to organize and to compare plots in an efficient manner combined with a session screenshot feature to edit scientific and publishable figures.</li>
<li>Three other supporting packages to help all users from beginners and advanced developers to extend under current SPS framework or on their own visualization apps.</li>
</ul>
</div>
<div id="demo" class="section level2">
<h2 class="hasAnchor">
<a href="#demo" class="anchor"></a>Demo</h2>
<p>View our online demo app:</p>
<table class="table">
<colgroup>
<col width="33%">
<col width="33%">
<col width="33%">
</colgroup>
<thead><tr class="header">
<th>Type and link</th>
<th>option changed</th>
<th>notes</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny/">Default full installation{blk}</a></td>
<td><a href="#installation">See installation</a></td>
<td>full app</td>
</tr>
<tr class="even">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny/">Minimum installation{blk}</a></td>
<td><a href="#installation">See installation</a></td>
<td>no modules installed</td>
</tr>
<tr class="odd">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny_loading/">Login enabled{blk}</a></td>
<td><code>login_screen = TRUE; login_theme = "empty"</code></td>
<td>no modules installed</td>
</tr>
<tr class="even">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny_loading_theme/">Login and login themes{blk}</a></td>
<td><code>login_screen = TRUE; login_theme = "random"</code></td>
<td>no modules installed</td>
</tr>
<tr class="odd">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny_loading/?admin">App admin page{blk}</a></td>
<td><code>admin_page = TRUE</code></td>
<td>or simply add “?admin” to the end of URL of demos</td>
</tr>
</tbody>
</table>
<p>For the login required demos, the app account name is <strong>“user”</strong> password <strong>“user”</strong>.</p>
<p>For the admin panel login, account name <strong>“admin”</strong>, password <strong>“admin”</strong>.</p>
<p><strong>Please DO NOT delete or change password when you are using the admin features.</strong> <em>shinyapp.io</em> will reset the app once a while, but this will affect other people who are trying the demo simultaneously.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Daniela Cassol, Le Zhang, Thomas Girke.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
