<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>systemPipe: Workflow and Visualization Toolkit • systemPipeWorkshop2021</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="systemPipe: Workflow and Visualization Toolkit">
<meta property="og:description" content="systemPipeWorkshop2021">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeWorkshop2021</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/systemPipeWorkshop2021.html">Workshop</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/systemPipeR/systemPipeWorkshop2021">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://hub.docker.com/r/systempipe/systempipeworkshop2021">
    <span class="fab fa-docker"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="systemPipeWorkshop2021_files/header-attrs-2.9/header-attrs.js"></script><script src="systemPipeWorkshop2021_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="systemPipeWorkshop2021_files/jquery-3.5.1/jquery.min.js"></script><link href="systemPipeWorkshop2021_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="systemPipeWorkshop2021_files/datatables-binding-0.18/datatables.js"></script><link href="systemPipeWorkshop2021_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="systemPipeWorkshop2021_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="systemPipeWorkshop2021_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="systemPipeWorkshop2021_files/dt-ext-fixedcolumns-1.10.20/css/fixedColumns.dataTables.min.css" rel="stylesheet">
<script src="systemPipeWorkshop2021_files/dt-ext-fixedcolumns-1.10.20/js/dataTables.fixedColumns.min.js"></script><link href="systemPipeWorkshop2021_files/dt-ext-scroller-1.10.20/css/scroller.dataTables.min.css" rel="stylesheet">
<script src="systemPipeWorkshop2021_files/dt-ext-scroller-1.10.20/js/dataTables.scroller.min.js"></script><link href="systemPipeWorkshop2021_files/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="systemPipeWorkshop2021_files/crosstalk-1.1.1/js/crosstalk.min.js"></script><link href="systemPipeWorkshop2021_files/plotwf-2.1.2/plotwf.css" rel="stylesheet">
<script src="systemPipeWorkshop2021_files/plotwf-2.1.2/viz.js"></script><script src="systemPipeWorkshop2021_files/plotwf-2.1.2/full.render.js"></script><script src="systemPipeWorkshop2021_files/plotwf-binding-1.27.19/plotwf.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>systemPipe: Workflow and Visualization Toolkit</h1>
                        <h4 class="author"></h4>
            
            <h4 class="date">Last update: 03 August, 2021</h4>
      
      
      <div class="hidden name"><code>systemPipeWorkshop2021.Rmd</code></div>

    </div>

    
    
<p><strong>Authors</strong>: Daniela Cassol (<a href="mailto:danielac@ucr.edu" class="email">danielac@ucr.edu</a>), Le Zhang (<a href="mailto:le.zhang001@email.ucr.edu" class="email">le.zhang001@email.ucr.edu</a>), Thomas Girke (<a href="mailto:thomas.girke@ucr.edu" class="email">thomas.girke@ucr.edu</a>).</p>
<p><strong>Institution</strong>: Institute for Integrative Genome Biology, University of California, Riverside, California, USA.</p>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<div id="workshop-description" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h2>
<p>This workshop introduces <em>systemPipe</em> (SP), a generic toolkit for designing and running reproducible data analysis workflows. The environment consists of three major modules implemented as R/Bioconductor packages. <em>systemPipeR</em> (SPR) provides core functionalities for defining workflows, interacting with command-line software, and executing both R and/or command-line software, as well as generating publication-quality analysis reports. <em>systemPipeShiny</em> (SPS) integrates a graphical user interface for managing workflows and visualizing results interactively. <em>systemPipeWorkflow</em> (SPW) offers a collection of pre-configured workflow templates. This hands-on event will include the following topics: (1) brief overview of the design principles and functionalities of the SP toolkit; (2) design and usage of SPR’s command-line interface based on an object-oriented R implementation of CWL; (3) configuration and execution of workflows; (4) construction of custom workflows; (5) configuration and execution of a pre-configured workflow example from start to finish, e.g. RNA-Seq template; (6) parallel execution of workflows on HPC and cloud systems with and without schedulers; (7) generation of technical and scientific analysis reports including visualization; and (8) demonstration of SPS’ core functionalities, the project’s Shiny App.</p>
</div>
<div id="pre-requisites" class="section level2">
<h2 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h2>
<ul>
<li>Basic knowledge of R and usage of Bioconductor packages for NGS analysis</li>
<li>Basic knowledge of running command-line software</li>
<li>Basic knowledge of parallelization concepts</li>
</ul>
<p>Non-essential background reading:</p>
<ul>
<li><a href="https://systempipe.org/">systemPipe Workflow Environment WebPage</a></li>
<li><a href="https://bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html">systemPipeR vignette</a></li>
<li><a href="https://bioconductor.org/packages/devel/bioc/vignettes/systemPipeShiny/inst/doc/systemPipeShiny.html">systemPipeShiny vignette</a></li>
<li><a href="https://bioconductor.org/packages/release/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRdata.html">systemPipeRdata vignette</a></li>
<li><a href="https://rmarkdown.rstudio.com/lesson-2.html">R Markdown tutorial</a></li>
</ul>
</div>
<div id="workshop-participation" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h2>
<p>Participants will be able to perform all analysis components of this workshop hands-on. Active user participation throughout the event is highly encouraged, including but not limited to lecture material, hands-on sections, and final discussion about package improvements. Participants are encouraged to ask questions preferentially at the end of the workshop.</p>
</div>
<div id="r-bioconductor-packages-used" class="section level2">
<h2 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h2>
<ul>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/systemPipeR.html"><code>systemPipeR</code></a></li>
<li><a href="https://bioconductor.org/packages/devel/bioc/html/systemPipeShiny.html"><code>systemPipeShiny</code></a></li>
<li><a href="http://www.bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html"><code>systemPipeRdata</code></a></li>
</ul>
</div>
<div id="time-outline" class="section level2">
<h2 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h2>
<p>1h 45m total</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Overview of <em>systemPipe</em> toolkit</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Introduction to SPR’s command-line interface</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Configuration and execution of workflows</td>
<td>20m</td>
</tr>
<tr class="even">
<td>Showcase RNA-Seq workflow</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Parallelization on single machines and clusters</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Generation of technical and scientific analysis reports</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>Overview of <em>systemPipeShiny</em> core functionalities</td>
<td>20m</td>
</tr>
</tbody>
</table>
</div>
<div id="workshop-goals-and-objectives" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h2>
<div id="learning-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h3>
<ul>
<li>Recognize the benefits of a generic R-based workflow construction environment that is both scalable and reproducible</li>
<li>Integration of command-line tools via the CWL community standard</li>
<li>Rendering of R markdown reports and critical assessment of scientific analysis reports</li>
<li>Parallelization of big data analysis tasks</li>
</ul>
</div>
<div id="learning-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h3>
<ul>
<li>Identify and practice how to make analysis workflows more robust, reproducible, and portable across heterogeneous computing systems</li>
<li>Usage of new workflow control class for designing, configuring, and running workflows</li>
<li>Optimize and debug workflows</li>
<li>Inspection of technical reports and log files</li>
<li>Design of new and fully customized workflows</li>
<li>Practice interactive workflow management and visualization</li>
</ul>
</div>
</div>
</div>
<div id="workshop" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop" class="anchor"></a>Workshop</h1>
<div id="running-the-workshop" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-workshop" class="anchor"></a>Running the Workshop</h2>
<p>This workshop uses R <code>4.1.0</code> and Bioconductor version<code>3.14</code>. Bioconductor can be installed following <a href="https://www.bioconductor.org/developers/how-to/useDevel/">these instructions</a>.</p>
<p>During the <a href="https://bioc2021.bioconductor.org/">Bioc2021 conference</a>, the workshop can be run in the <a href="http://app.orchestra.cancerdatasci.org/">cloud</a>.</p>
<div id="workshop-setup-with-docker" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-setup-with-docker" class="anchor"></a>Workshop setup with Docker</h3>
<p>The Docker container used by this workshop runs with Bioconductor’s <code>development</code> version <code>3.14</code>. I includes all the necessary packages and software for running the code of in the workshop vignettes. To use the Docker container, one needs to first install <a href="https://docs.docker.com/engine/install/">Docker</a> on a user’s system.</p>
<ul>
<li>The container can be downloaded and run with:</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-e</span> PASSWORD=systempipe <span class="at">-p</span> 8787:8787 systempipe/systempipeworkshop2021:latest</span></code></pre></div>
<ul>
<li><p>Log in to RStudio at <a href="http://localhost:8787">http://localhost:8787</a> using username <code>rstudio</code> and password <code>systempipe</code>.</p></li>
<li><p>If you prefer to run the workshop from the command-line:</p></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-it</span> <span class="at">--user</span> rstudio systempipe/systempipeworkshop2021:latest bash </span></code></pre></div>
</div>
<div id="workshop-setup-with-github" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-setup-with-github" class="anchor"></a>Workshop setup with GitHub</h3>
<p><a href="http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html"><em><code>systemPipeR</code></em></a> and <a href="http://bioconductor.org/packages/devel/bioc/html/systemPipeShiny.html"><em><code>systemPipeShiny</code></em></a> environment can be installed from the R console using the <a href="https://cran.r-project.org/web/packages/BiocManager/index.html"><em><code>BiocManager::install</code></em></a> command. The associated data package <a href="http://www.bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html"><em><code>systemPipeRdata</code></em></a> can be installed the same way. The latter is a helper package for generating <em><code>systemPipeR</code></em> workflow environments with a single command containing all parameter files and sample data required to quickly test and run workflows.</p>
<p>You can install locally the necessary packages for this workshop as follows:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Install workshop package</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"systemPipeR/systemPipeWorkshop2021"</span><span class="op">)</span>
<span class="co">## Install required packages</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"systemPipeR"</span>, <span class="st">"systemPipeRdata"</span>, <span class="st">"systemPipeShiny"</span><span class="op">)</span>, version<span class="op">=</span><span class="st">"3.14"</span><span class="op">)</span></code></pre></div>
<p>To access the vignette:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/browseVignettes.html">browseVignettes</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"systemPipeWorkshop2021"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting Started</h2>
<div id="loading-package-and-documentation" class="section level3">
<h3 class="hasAnchor">
<a href="#loading-package-and-documentation" class="anchor"></a>Loading package and documentation</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://systempipe.org/">"systemPipeR"</a></span><span class="op">)</span> 
<span class="co">#&gt; Loading required package: Rsamtools</span>
<span class="co">#&gt; Loading required package: GenomeInfoDb</span>
<span class="co">#&gt; Loading required package: BiocGenerics</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'BiocGenerics'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     IQR, mad, sd, var, xtabs</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
<span class="co">#&gt;     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
<span class="co">#&gt;     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
<span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
<span class="co">#&gt;     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
<span class="co">#&gt;     union, unique, unsplit, which.max, which.min</span>
<span class="co">#&gt; Loading required package: S4Vectors</span>
<span class="co">#&gt; Loading required package: stats4</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'S4Vectors'</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand.grid, I, unname</span>
<span class="co">#&gt; Loading required package: IRanges</span>
<span class="co">#&gt; Loading required package: GenomicRanges</span>
<span class="co">#&gt; Loading required package: Biostrings</span>
<span class="co">#&gt; Loading required package: XVector</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'Biostrings'</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     strsplit</span>
<span class="co">#&gt; Loading required package: ShortRead</span>
<span class="co">#&gt; Loading required package: BiocParallel</span>
<span class="co">#&gt; Loading required package: GenomicAlignments</span>
<span class="co">#&gt; Loading required package: SummarizedExperiment</span>
<span class="co">#&gt; Loading required package: MatrixGenerics</span>
<span class="co">#&gt; Loading required package: matrixStats</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'MatrixGenerics'</span>
<span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span>
<span class="co">#&gt; Loading required package: Biobase</span>
<span class="co">#&gt; Welcome to Bioconductor</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Vignettes contain introductory material; view with</span>
<span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'Biobase'</span>
<span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     rowMedians</span>
<span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyMissing, rowMedians</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'systemPipeR'</span>
<span class="co">#&gt; The following object is masked from 'package:BiocStyle':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     output</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://systempipe.org/sps">"systemPipeShiny"</a></span><span class="op">)</span> 
<span class="co">#&gt; Loading required package: shiny</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'shiny'</span>
<span class="co">#&gt; The following object is masked from 'package:BiocStyle':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     markdown</span>
<span class="co">#&gt; Loading required package: spsUtil</span>
<span class="co">#&gt; Loading required package: spsComps</span>
<span class="co">#&gt; Loading required package: drawer</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/tgirke/systemPipeRdata">"systemPipeRdata"</a></span><span class="op">)</span> </code></pre></div>
<ul>
<li>Documentation</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span>help<span class="op">=</span><span class="st">"systemPipeR"</span><span class="op">)</span> <span class="co"># Lists package info</span>
<span class="fu"><a href="https://rdrr.io/r/utils/vignette.html">vignette</a></span><span class="op">(</span><span class="st">"systemPipeR"</span><span class="op">)</span> <span class="co"># Opens vignette</span></code></pre></div>
</div>
<div id="set-working-directory" class="section level3">
<h3 class="hasAnchor">
<a href="#set-working-directory" class="anchor"></a>Set working directory</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"vignettes"</span><span class="op">)</span></code></pre></div>
</div>
<div id="how-to-get-help-for-systempipersystempipeshiny" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-get-help-for-systempipersystempipeshiny" class="anchor"></a>How to get help for systemPipeR/systemPipeShiny</h3>
<p>All questions about the package or any particular function should be posted to the Bioconductor support site <a href="https://support.bioconductor.org">https://support.bioconductor.org</a>.</p>
<p>Please add the “<em><code>systemPipeR</code></em>” tag to your question, and the package authors will automatically receive an alert.</p>
<p>We appreciate receiving reports of bugs in the functions or documentation and suggestions for improvement. For that, please consider opening an issue at <a href="https://github.com/systemPipeR">GitHub</a>.</p>
</div>
</div>
<div id="overview-of-systempipe-toolkit" class="section level2">
<h2 class="hasAnchor">
<a href="#overview-of-systempipe-toolkit" class="anchor"></a>Overview of <em>systemPipe</em> toolkit</h2>
<p><code>systemPipe</code> (SP) is a generic toolkit for designing and running reproducible data analysis workflows. The environment consists of three major modules implemented as R/Bioconductor packages:</p>
<ul>
<li><p><code>systemPipeR</code> (SPR) provides core functionalities for defining workflows, interacting with command-line software, and executing both R and/or command-line software, as well as generating publication-quality analysis reports.</p></li>
<li><p><code>systemPipeShiny</code> (SPS) integrates a graphical user interface for managing workflows and visualizing results interactively.</p></li>
<li><p><code>systemPipeWorkflow</code> (SPW) offers a collection of pre-configured workflow templates.</p></li>
</ul>
</div>
<div id="introduction-to-sprs-command-line-interface" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction-to-sprs-command-line-interface" class="anchor"></a>Introduction to SPR’s command-line interface</h2>
<p>A central concept for designing workflows within the <code>systemPipeR</code> environment is the use of workflow management containers. <code>systemPipeR</code> adopted the widely used community standard <a href="https://www.commonwl.org/">Common Workflow Language</a> (CWL) <span class="citation">(Amstutz et al. 2016)</span> for describing analysis workflows in a generic and reproducible manner. Using this community standard in <code>systemPipeR</code> has many advantages. For instance, the integration of CWL allows running <code>systemPipeR</code> workflows from a single specification instance either entirely from within R, from various command line wrappers (e.g., cwl-runner) or from other languages (, e.g., Bash or Python). <code>systemPipeR</code> includes support for both command line and R/Bioconductor software as well as resources for containerization, parallel evaluations on computer clusters along with the automated generation of interactive analysis reports.</p>
<p>An important feature of <code>systemPipeR's</code> CWL interface is that it provides two options to run command line tools and workflows based on CWL. First, one can run CWL in its native way via an R-based wrapper utility for <code>cwl-runner</code> or <code>cwl-tools</code> (CWL-based approach). Second, one can run workflows using CWL’s command line and workflow instructions from within R (R-based approach). In the latter case the same CWL workflow definition files (e.g. <em>.cwl</em> and <em>.yml</em>) are used but rendered and executed entirely with R functions defined by <code>systemPipeR</code>, and thus use CWL mainly as a command line and workflow definition format rather than software to run workflows. In this regard <code>systemPipeR</code> also provides several convenience functions that are useful for designing and debugging workflows, such as a command line rendering function to retrieve the exact command line strings for each data set and processing step prior to running a command line.</p>
</div>
<div id="configuration-and-execution-of-workflows" class="section level2">
<h2 class="hasAnchor">
<a href="#configuration-and-execution-of-workflows" class="anchor"></a>Configuration and execution of workflows</h2>
<div id="load-sample-data-and-directory-structure" class="section level3">
<h3 class="hasAnchor">
<a href="#load-sample-data-and-directory-structure" class="anchor"></a>Load sample data and directory structure</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">systemPipeRdata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span><span class="st">"new"</span>, mydirname <span class="op">=</span> <span class="st">"bioc2021"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Generated bioc2021 directory. Next run in new directory, the R code from *.Rmd template interactively. Alternatively, workflows can be exectued with a single command as instructed in the vignette."</span></code></pre></div>
</div>
<div id="project-structure" class="section level3">
<h3 class="hasAnchor">
<a href="#project-structure" class="anchor"></a>Project structure</h3>
<p><em><code>systemPipeR</code></em> expects a project directory structure that consists of a directory where users may store all the raw data, the results directory that will be reserved for all the outfiles files or new output folders, and the parameters directory.</p>
<p>This structure allows reproducibility and collaboration across the data science team since internally relative paths are used. Users could transfer this project to a different location and still be able to run the entire workflow. Also, it increases efficiency and data management once the raw data is kept in a separate folder and avoids duplication.</p>
<div id="dir" class="section level4">
<h4 class="hasAnchor">
<a href="#dir" class="anchor"></a>Directory Structure</h4>
<p><a href="http://bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html"><em><code>systemPipeRdata</code></em></a>, helper package, provides pre-configured workflows, reporting templates, and sample data loaded as demonstrated below. With a single command, the package allows creating the workflow environment containing the structure described here (see Figure @ref(fig:dir)).</p>
<p>Directory names are indicated in <span style="color:grey"><strong><em>green</em></strong></span>. Users can change this structure as needed, but need to adjust the code in their workflows accordingly.</p>
<ul>
<li>
<span style="color:green"><em><strong>workflow/</strong></em></span> (<em>e.g.</em> <em>myproject/</em>)
<ul>
<li>This is the root directory of the R session running the workflow.</li>
<li>Run script ( <em>*.Rmd</em>) and sample annotation (<em>targets.txt</em>) files are located here.</li>
<li>Note, this directory can have any name (<em>e.g.</em> <span style="color:green"><em><strong>myproject</strong></em></span>). Changing its name does not require any modifications in the run script(s).</li>
<li>
<strong>Important subdirectories</strong>:
<ul>
<li>
<span style="color:green"><em><strong>param/</strong></em></span>
<ul>
<li>
<span style="color:green"><em><strong>param/cwl/</strong></em></span>: This subdirectory stores all the parameter and configuration files. To organize workflows, each can have its own subdirectory, where all <code>*.cwl</code> and <code>*input.yml</code> files need to be in the same subdirectory.</li>
</ul>
</li>
<li>
<span style="color:green"><em><strong>data/</strong></em> </span>
<ul>
<li>Raw data (<em>e.g.</em> FASTQ files)</li>
<li>FASTA file of reference (<em>e.g.</em> reference genome)</li>
<li>Annotation files</li>
<li>Metadata</li>
<li>etc.</li>
</ul>
</li>
<li>
<span style="color:green"><em><strong>results/</strong></em></span>
<ul>
<li>Analysis results are usually written to this directory, including: alignment, variant and peak files (BAM, VCF, BED); tabular result files; and image/plot files</li>
<li>Note, the user has the option to organize results files for a given sample and analysis step in a separate subdirectory.</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="figure" style="text-align: center">
<img src="../../../../usr/local/lib/R/site-library/systemPipeR/extdata/images/spr_project.png" alt="*systemPipeR's* preconfigured directory structure." width="100%"><p class="caption">
<em>systemPipeR’s</em> preconfigured directory structure.
</p>
</div>
</div>
<div id="structure-of-initial-targets-file" class="section level4">
<h4 class="hasAnchor">
<a href="#structure-of-initial-targets-file" class="anchor"></a>Structure of initial <em><code>targets</code></em> file</h4>
<p>The <em><code>targets</code></em> file defines all input files (<em>e.g.</em> FASTQ, BAM, BCF) and sample comparisons of an analysis workflow. The following shows the format of a sample <em><code>targets</code></em> file included in the package. It also can be viewed and downloaded from <em><code>systemPipeR</code></em>’s GitHub repository <a href="https://github.com/tgirke/systemPipeR/blob/master/inst/extdata/targets.txt">here</a>. In a target file with a single type of input files, here FASTQ files of single-end (SE) reads, the first column describe the path and the second column represents a unique id name for each sample. The third column called <code>Factor</code> represents the biological replicates. All subsequent columns are additional information, and any number of extra columns can be added as needed.</p>
<p>Users should note here, the usage of targets files is optional when using <em><code>systemPipeR's</code></em> new workflow management interface. They can be replaced by a standard YAML input file used by CWL. Since for organizing experimental variables targets files are extremely useful and user-friendly. Thus, we encourage users to keep using them.</p>
<ul>
<li>Structure of <em><code>targets</code></em> file for single-end (SE) samples</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="st">"targets.txt"</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/showDF.html">showDF</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">targetspath</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-6d085f39b286c43d4c9b" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-6d085f39b286c43d4c9b">{"x":{"filter":"none","extensions":["FixedColumns","Scroller"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18"],["./data/SRR446027_1.fastq.gz","./data/SRR446028_1.fastq.gz","./data/SRR446029_1.fastq.gz","./data/SRR446030_1.fastq.gz","./data/SRR446031_1.fastq.gz","./data/SRR446032_1.fastq.gz","./data/SRR446033_1.fastq.gz","./data/SRR446034_1.fastq.gz","./data/SRR446035_1.fastq.gz","./data/SRR446036_1.fastq.gz","./data/SRR446037_1.fastq.gz","./data/SRR446038_1.fastq.gz","./data/SRR446039_1.fastq.gz","./data/SRR446040_1.fastq.gz","./data/SRR446041_1.fastq.gz","./data/SRR446042_1.fastq.gz","./data/SRR446043_1.fastq.gz","./data/SRR446044_1.fastq.gz"],["M1A","M1B","A1A","A1B","V1A","V1B","M6A","M6B","A6A","A6B","V6A","V6B","M12A","M12B","A12A","A12B","V12A","V12B"],["M1","M1","A1","A1","V1","V1","M6","M6","A6","A6","V6","V6","M12","M12","A12","A12","V12","V12"],["Mock.1h.A","Mock.1h.B","Avr.1h.A","Avr.1h.B","Vir.1h.A","Vir.1h.B","Mock.6h.A","Mock.6h.B","Avr.6h.A","Avr.6h.B","Vir.6h.A","Vir.6h.B","Mock.12h.A","Mock.12h.B","Avr.12h.A","Avr.12h.B","Vir.12h.A","Vir.12h.B"],[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],["23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>FileName<\/th>\n      <th>SampleName<\/th>\n      <th>Factor<\/th>\n      <th>SampleLong<\/th>\n      <th>Experiment<\/th>\n      <th>Date<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"fixedColumns":true,"deferRender":true,"scrollY":200,"scroller":true,"columnDefs":[{"className":"dt-right","targets":5},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>To work with custom data, users need to generate a <em><code>targets</code></em> file containing the paths to their own FASTQ files and then provide under <em><code>targetspath</code></em> the path to the corresponding <em><code>targets</code></em> file.</p>
<ul>
<li>Structure of <em><code>targets</code></em> file for “Hello World” example</li>
</ul>
<p>In this example, <em><code>targets</code></em> file presents only two columns, which the first one are the different phrases used by the <code>echo</code> command-line and the second column it is the sample <code>id</code>. The <code>id</code> column is required, and each sample id should be unique.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/example/targets_example.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/showDF.html">showDF</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">targetspath</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-0b581b65ccc5f48da138" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-0b581b65ccc5f48da138">{"x":{"filter":"none","extensions":["FixedColumns","Scroller"],"data":[["1","2","3"],["Hello World!","Hello USA!","Hello Bioconductor!"],["M1","M2","M3"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Message<\/th>\n      <th>SampleName<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"fixedColumns":true,"deferRender":true,"scrollY":200,"scroller":true,"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<div id="structure-of-the-new-parameters-files" class="section level3">
<h3 class="hasAnchor">
<a href="#structure-of-the-new-parameters-files" class="anchor"></a>Structure of the new parameters files</h3>
<p>The parameters and configuration required for running command-line software are provided by the widely used community standard <a href="https://www.commonwl.org/">Common Workflow Language</a> (CWL) <span class="citation">(Amstutz et al. 2016)</span>, which describes parameters analysis workflows in a generic and reproducible manner. For R-based workflow steps, param files are not required. For a complete overview of the CWL syntax, please see the <a href="#cwl">section</a> below. Also, we have a dedicated section explain how to <em><code>systemPipeR</code></em> establish the connection between the CWL parameters files and the targets files. Please see <a href="#cwl_targets">here</a>.</p>
<div id="automate-creation-of-cwl-parameters" class="section level4">
<h4 class="hasAnchor">
<a href="#automate-creation-of-cwl-parameters" class="anchor"></a>Automate creation of CWL parameters</h4>
<p>Users need to define the command-line in a pseudo-bash script format:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># "hisat2 -S ./results/M1A.sam -x ./data/tair10.fasta -k 1 -threads 4 -U ./data/SRR446027_1.fastq.gz "</span>
<span class="va">command</span> <span class="op">&lt;-</span> <span class="st">"
    hisat2 \
    -S &lt;F, out: ./results/M1A.sam&gt; \
    -x &lt;F: ./data/tair10.fasta&gt; \
     -k &lt;int: 1&gt; \
    -threads &lt;int: 4&gt; \
    -U &lt;F: ./data/SRR446027_1.fastq.gz&gt;
"</span></code></pre></div>
<div id="define-prefix-and-defaults" class="section level5">
<h5 class="hasAnchor">
<a href="#define-prefix-and-defaults" class="anchor"></a>Define prefix and defaults</h5>
<ul>
<li><p>First line is the base command. Each line is an argument with its default value.</p></li>
<li><p>For argument lines (starting from the second line), any word before the first space with leading <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code> or <code>--</code> in each will be treated as a prefix, like <code>-S</code> or <code>--min</code>. Any line without this first word will be treated as no prefix.</p></li>
<li><p>All defaults are placed inside <code>&lt;...&gt;</code>.</p></li>
<li><p>First argument is the input argument type. <code>F</code> for “File,” “int,” “string” are unchanged.</p></li>
<li><p>Optional: use the keyword <code>out</code> followed the type with a <code>,</code> comma separation to indicate if this argument is also an CWL output.</p></li>
<li><p>Then, use <code><a href="https://rdrr.io/r/base/Colon.html">:</a></code> to separate keywords and default values, any non-space value after the <code><a href="https://rdrr.io/r/base/Colon.html">:</a></code> will be treated as the default value.</p></li>
<li><p>If any argument has no default value, just a flag, like <code>--verbose</code>, there is no need to add any <code>&lt;...&gt;</code></p></li>
</ul>
</div>
<div id="createparam-function" class="section level5">
<h5 class="hasAnchor">
<a href="#createparam-function" class="anchor"></a><code>createParam</code> Function</h5>
<p><code>createParam</code> function requires the <code>string</code> as defined above as an input.</p>
<p>First of all, the function will print the three components of the <code>cwl</code> file: - <code>BaseCommand</code>: Specifies the program to execute. - <code>Inputs</code>: Defines the input parameters of the process. - <code>Outputs</code>: Defines the parameters representing the output of the process.</p>
<p>The four component is the original command-line.</p>
<p>If in interactive mode, the function will verify that everything is correct and will ask you to proceed. Here, the user can answer “no” and provide more information at the string level. Another question is to save the param created here.</p>
<p>If running the workflow in non-interactive mode, the <code>createParam</code> function will consider “yes” and returning the container.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cmd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/createParamFiles.html">createParam</a></span><span class="op">(</span><span class="va">command</span>, writeParamFiles <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; *****BaseCommand*****</span>
<span class="co">#&gt; hisat2 </span>
<span class="co">#&gt; *****Inputs*****</span>
<span class="co">#&gt; S:</span>
<span class="co">#&gt;     type: File</span>
<span class="co">#&gt;     preF: -S</span>
<span class="co">#&gt;     yml: ./results/M1A.sam</span>
<span class="co">#&gt; x:</span>
<span class="co">#&gt;     type: File</span>
<span class="co">#&gt;     preF: -x</span>
<span class="co">#&gt;     yml: ./data/tair10.fasta</span>
<span class="co">#&gt; k:</span>
<span class="co">#&gt;     type: int</span>
<span class="co">#&gt;     preF: -k</span>
<span class="co">#&gt;     yml: 1</span>
<span class="co">#&gt; threads:</span>
<span class="co">#&gt;     type: int</span>
<span class="co">#&gt;     preF: -threads</span>
<span class="co">#&gt;     yml: 4</span>
<span class="co">#&gt; U:</span>
<span class="co">#&gt;     type: File</span>
<span class="co">#&gt;     preF: -U</span>
<span class="co">#&gt;     yml: ./data/SRR446027_1.fastq.gz</span>
<span class="co">#&gt; *****Outputs*****</span>
<span class="co">#&gt; output1:</span>
<span class="co">#&gt;     type: File</span>
<span class="co">#&gt;     value: ./results/M1A.sam</span>
<span class="co">#&gt; *****Parsed raw command line*****</span>
<span class="co">#&gt; hisat2 -S ./results/M1A.sam -x ./data/tair10.fasta -k 1 -threads 4 -U ./data/SRR446027_1.fastq.gz </span>
<span class="co">#&gt;   Written content of 'commandLine' to file: </span>
<span class="co">#&gt;  param/cwl/hisat2/hisat2.cwl </span>
<span class="co">#&gt;   Written content of 'commandLine' to file: </span>
<span class="co">#&gt;  param/cwl/hisat2/hisat2.yml</span></code></pre></div>
</div>
</div>
<div id="cwl_targets" class="section level4">
<h4 class="hasAnchor">
<a href="#cwl_targets" class="anchor"></a>How to connect CWL description files within <em><code>systemPipeR</code></em>
</h4>
<p>This section will demonstrate how to connect CWL parameters files to create workflows. In addition, we will show how the workflow can be easily scalable with <em><code>systemPipeR</code></em>.</p>
<div class="figure" style="text-align: center">
<img src="../../../../usr/local/lib/R/site-library/systemPipeR/extdata/images/SPR_CWL_hello.png" alt="WConnectivity between CWL param files and targets files." width="100%"><p class="caption">
WConnectivity between CWL param files and targets files.
</p>
</div>
</div>
</div>
</div>
<div id="project-initialization" class="section level2">
<h2 class="hasAnchor">
<a href="#project-initialization" class="anchor"></a>Project initialization</h2>
<p>To create a Workflow within <em><code>systemPipeR</code></em>, we can start by defining an empty container and checking the directory structure:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>projPath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> 
<span class="co">#&gt; Creating directory '/__w/systemPipeWorkshop2021/systemPipeWorkshop2021/vignettes/bioc2021/.SPRproject'</span>
<span class="co">#&gt; Creating file '/__w/systemPipeWorkshop2021/systemPipeWorkshop2021/vignettes/bioc2021/.SPRproject/SYSargsList.yml'</span></code></pre></div>
<p>Internally, <code>SPRproject</code> function will create a hidden folder called <code>.SPRproject</code>, by default, to store all the log files. A <code>YAML</code> file, here called <code>SYSargsList.yml</code>, has been created, which initially contains the basic location of the project structure; however, every time the workflow object <code>sal</code> is updated in R, the new information will also be store in this flat-file database for easy recovery. If you desire different names for the logs folder and the <code>YAML</code> file, these can be modified as follows:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>logs.dir<span class="op">=</span> <span class="st">".SPRproject"</span>, sys.file<span class="op">=</span><span class="st">".SPRproject/SYSargsList.yml"</span><span class="op">)</span> </code></pre></div>
<p>Also, this function will check and/or create the basic folder structure if missing, which means <code>data</code>, <code>param</code>, and <code>results</code> folder, as described <a href="#dir">here</a>. If the user wants to use a different names for these directories, can be specified as follows:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>data <span class="op">=</span> <span class="st">"data"</span>, param <span class="op">=</span> <span class="st">"param"</span>, results <span class="op">=</span> <span class="st">"results"</span><span class="op">)</span> </code></pre></div>
<p>It is possible to separate all the R objects created within the workflow analysis from the current environment. <code>SPRproject</code> function provides the option to create a new environment, and in this way, it is not overwriting any object you may want to have at your current section.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>envir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/environment.html">new.env</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> </code></pre></div>
<p>In this stage, the object <code>sal</code> is a empty container, except for the project information. The project information can be accessed by the <code>projectInfo</code> method:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="co">#&gt; Instance of 'SYSargsList': </span>
<span class="co">#&gt;  No workflow steps added</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">projectInfo</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span>
<span class="co">#&gt; $project</span>
<span class="co">#&gt; [1] "/__w/systemPipeWorkshop2021/systemPipeWorkshop2021/vignettes/bioc2021"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $data</span>
<span class="co">#&gt; [1] "data"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $param</span>
<span class="co">#&gt; [1] "param"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $results</span>
<span class="co">#&gt; [1] "results"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $logsDir</span>
<span class="co">#&gt; [1] ".SPRproject"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $sysargslist</span>
<span class="co">#&gt; [1] ".SPRproject/SYSargsList.yml"</span></code></pre></div>
<p>Also, the <code>length</code> function will return how many steps this workflow contains and in this case it is empty, as follow:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span>
<span class="co">#&gt; [1] 0</span></code></pre></div>
</div>
<div id="workflow-design" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-design" class="anchor"></a>Workflow Design</h2>
<p><em><code>systemPipeR</code></em> workflows can be designed and built from start to finish with a single command, importing from an R Markdown file or stepwise in interactive mode from the R console. In the <a href="#appendstep">next section</a>, we will demonstrate how to build the workflow in an interactive mode, and in the <a href="#importWF">following section</a>, we will show how to build from a file.</p>
<p>New workflows are constructed, or existing ones modified, by connecting each step via <code>appendStep</code> method. Each <code>SYSargsList</code> instance contains instructions needed for processing a set of input files with a specific command-line or R software, as well as the paths to the corresponding outfiles generated by a particular tool/step.</p>
<p>To build R code based step, the constructor function <code>Linewise</code> is used. For more details about this S4 class container, see <a href="#linewise">here</a>.</p>
<div id="appendstep" class="section level3">
<h3 class="hasAnchor">
<a href="#appendstep" class="anchor"></a>Build workflow interactive</h3>
<p>This tutorial shows a very simple example for describing and explaining all main features available within systemPipeR to design, build, manage, run, and visualize the workflow. In summary, we are exporting a dataset to multiple files, compressing and decompressing each one of the files, and importing to R, and finally performing a statistical analysis.</p>
<p>In the previous section, we initialize the project by building the <code>sal</code> object. Until this moment, the container has no steps:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="co">#&gt; Instance of 'SYSargsList': </span>
<span class="co">#&gt;  No workflow steps added</span></code></pre></div>
<p>Next, we need to populate the object created with the first step in the workflow.</p>
<div id="adding-the-first-step" class="section level4">
<h4 class="hasAnchor">
<a href="#adding-the-first-step" class="anchor"></a>Adding the first step</h4>
<p>The first step is R code based, and we are splitting the <code>iris</code> dataset by <code>Species</code> and for each <code>Species</code> will be saved on file. Please note that this code will not be executed now; it is just store in the container for further execution.</p>
<p>This constructor function requires the <code>step_name</code> and the R-based code under the <code>code</code> argument. The R code should be enclosed by braces (<code>{}</code>) and separated by a new line.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
                              <span class="fu"><a href="https://rdrr.io/r/base/mapply.html">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.csv</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span>,
                                     <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">iris</span>, <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">iris</span><span class="op">$</span><span class="va">Species</span><span class="op">)</span><span class="op">)</span>,
                                     <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="st">"results"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">iris</span>, <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">iris</span><span class="op">$</span><span class="va">Species</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, <span class="st">".csv"</span><span class="op">)</span><span class="op">)</span>
                                     <span class="op">)</span> 
                            <span class="op">}</span>,
                            step_name <span class="op">=</span> <span class="st">"export_iris"</span><span class="op">)</span></code></pre></div>
<p>For a brief overview of the workflow, we can check the object as follows:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="co">#&gt; Instance of 'SYSargsList': </span>
<span class="co">#&gt;     WF Steps:</span>
<span class="co">#&gt;        1. export_iris --&gt; Status: Pending</span>
<span class="co">#&gt; </span></code></pre></div>
<p>Also, for printing and double-check the R code in the step, we can use the <code>codeLine</code> method:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">codeLine</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span>
<span class="co">#&gt; export_iris</span>
<span class="co">#&gt;     mapply(function(x, y) write.csv(x, y), split(iris, factor(iris$Species)), file.path("results", paste0(names(split(iris, factor(iris$Species))), ".csv")))</span></code></pre></div>
</div>
<div id="adding-more-steps" class="section level4">
<h4 class="hasAnchor">
<a href="#adding-more-steps" class="anchor"></a>Adding more steps</h4>
<p>Next, an example of how to compress the exported files using <a href="https://www.gnu.org/software/gzip/"><code>gzip</code></a> command-line.</p>
<p>The constructor function creates an <code>SYSargsList</code> S4 class object using data from three input files:</p>
<pre><code>- CWL command-line specification file (`wf_file` argument);
- Input variables (`input_file` argument);
- Targets file (`targets` argument).</code></pre>
<p>In CWL, files with the extension <code>.cwl</code> define the parameters of a chosen command-line step or workflow, while files with the extension <code>.yml</code> define the input variables of command-line steps.</p>
<p>The <code>targets</code> file is optional for workflow steps lacking <code>input</code> files. The connection between <code>input</code> variables and the <code>targets</code> file is defined under the <code>inputvars</code> argument. It is required a <code>named vector</code>, where each element name needs to match with column names in the <code>targets</code> file, and the value must match the names of the <code>input</code> variables defined in the <code>*.yml</code> files (see Figure @ref(fig:sprCWL)).</p>
<p>A detailed description of the dynamic between <code>input</code> variables and <code>targets</code> files can be found <a href="#cwl_targets">here</a>. In addition, the CWL syntax overview can be found <a href="#cwl">here</a>.</p>
<p>Besides all the data form <code>targets</code>, <code>wf_file</code>, <code>input_file</code> and <code>dir_path</code> arguments, <code>SYSargsList</code> constructor function options include:</p>
<ul>
<li>
<code>step_name</code>: a unique <em>name</em> for the step. This is not mandatory; however, it is highly recommended. If no name is provided, a default <code>step_x</code>, where <code>x</code> reflects the step index, will be added.</li>
<li>
<code>dir</code>: this option allows creating an exclusive subdirectory for the step in the workflow. All the outfiles and log files for this particular step will be generated in the respective folders.</li>
<li>
<code>dependency</code>: after the first step, all the additional steps appended to the workflow require the information of the dependency tree.</li>
</ul>
<p>The <code>appendStep&lt;-</code> method is used to append a new step in the workflow.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/gunzip"</span>, <span class="st">"targets_gunzip.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"gzip"</span>, 
                      targets <span class="op">=</span> <span class="va">targetspath</span>, dir <span class="op">=</span> <span class="cn">TRUE</span>,
                      wf_file <span class="op">=</span> <span class="st">"gunzip/workflow_gzip.cwl"</span>, input_file <span class="op">=</span> <span class="st">"gunzip/gzip.yml"</span>,
                      dir_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>,
                      inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FILE_PATH_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>, 
                      dependency <span class="op">=</span> <span class="st">"export_iris"</span><span class="op">)</span></code></pre></div>
<p>Note: This will not work if the <code>gzip</code> is not available on your system (installed and exported to PATH) and may only work on Windows systems using PowerShell.</p>
<p>For a overview of the workflow, we can check the object as follows:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="co">#&gt; Instance of 'SYSargsList': </span>
<span class="co">#&gt;     WF Steps:</span>
<span class="co">#&gt;        1. export_iris --&gt; Status: Pending</span>
<span class="co">#&gt;        2. gzip --&gt; Status: Pending </span>
<span class="co">#&gt;            Total Files: 3 | Existing: 0 | Missing: 3 </span>
<span class="co">#&gt;          2.1. gzip</span>
<span class="co">#&gt;              cmdlist: 3 | Pending: 3</span>
<span class="co">#&gt; </span></code></pre></div>
<p>Note that we have two steps, and it is expected three files from the second step. Also, the workflow status is <em>Pending</em>, which means the workflow object is rendered in R; however, we did not execute the workflow yet. In addition to this summary, it can be observed this step has three command lines.</p>
<p>For more details about the command-line rendered for each target file, it can be checked as follows:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">cmdlist</a></span><span class="op">(</span><span class="va">sal</span>, step<span class="op">=</span><span class="st">"gzip"</span><span class="op">)</span>
<span class="co">#&gt; $gzip</span>
<span class="co">#&gt; $gzip$SE</span>
<span class="co">#&gt; $gzip$SE$gzip</span>
<span class="co">#&gt; [1] "gzip -c  results/setosa.csv &gt; results/SE.csv.gz"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $gzip$VE</span>
<span class="co">#&gt; $gzip$VE$gzip</span>
<span class="co">#&gt; [1] "gzip -c  results/versicolor.csv &gt; results/VE.csv.gz"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $gzip$VI</span>
<span class="co">#&gt; $gzip$VI$gzip</span>
<span class="co">#&gt; [1] "gzip -c  results/virginica.csv &gt; results/VI.csv.gz"</span></code></pre></div>
</div>
<div id="using-the-outfiles-for-the-next-step" class="section level4">
<h4 class="hasAnchor">
<a href="#using-the-outfiles-for-the-next-step" class="anchor"></a>Using the <code>outfiles</code> for the next step</h4>
<p>For building this step, all the previous procedures are being used to append the next step. However, here, we can observe power features that build the connectivity between steps in the workflow.</p>
<p>In this example, we would like to use the outfiles from <em>gzip</em> Step, as input from the next step, which is the <em>gunzip</em>. In this case, let’s look at the outfiles from the first step:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">outfiles</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span>
<span class="co">#&gt; $export_iris</span>
<span class="co">#&gt; DataFrame with 0 rows and 0 columns</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $gzip</span>
<span class="co">#&gt; DataFrame with 3 rows and 1 column</span>
<span class="co">#&gt;           gzip_file</span>
<span class="co">#&gt;         &lt;character&gt;</span>
<span class="co">#&gt; 1 results/SE.csv.gz</span>
<span class="co">#&gt; 2 results/VE.csv.gz</span>
<span class="co">#&gt; 3 results/VI.csv.gz</span></code></pre></div>
<p>The column we want to use is “gzip_file.” For the argument <code>targets</code> in the <code>SYSargsList</code> function, it should provide the name of the correspondent step in the Workflow and which <code>outfiles</code> you would like to be incorporated in the next step. The argument <code>inputvars</code> allows the connectivity between <code>outfiles</code> and the new <code>targets</code> file. Here, the name of the previous <code>outfiles</code> should be provided it. Please note that all <code>outfiles</code> column names must be unique.</p>
<p>It is possible to keep all the original columns from the <code>targets</code> files or remove some columns for a clean <code>targets</code> file. The argument <code>rm_targets_col</code> provides this flexibility, where it is possible to specify the names of the columns that should be removed. If no names are passing here, the new columns will be appended.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"gunzip"</span>, 
                      targets <span class="op">=</span> <span class="st">"gzip"</span>, dir <span class="op">=</span> <span class="cn">TRUE</span>,
                      wf_file <span class="op">=</span> <span class="st">"gunzip/workflow_gunzip.cwl"</span>, input_file <span class="op">=</span> <span class="st">"gunzip/gunzip.yml"</span>,
                      dir_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>,
                      inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>gzip_file <span class="op">=</span> <span class="st">"_FILE_PATH_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>, 
                      rm_targets_col <span class="op">=</span> <span class="st">"FileName"</span>, 
                      dependency <span class="op">=</span> <span class="st">"gzip"</span><span class="op">)</span></code></pre></div>
<p>We can check the targets automatically create for this step, based on the previous <code>outfiles</code>:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">targetsWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; $gunzip</span>
<span class="co">#&gt; DataFrame with 3 rows and 2 columns</span>
<span class="co">#&gt;           gzip_file  SampleName</span>
<span class="co">#&gt;         &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt; 1 results/SE.csv.gz          SE</span>
<span class="co">#&gt; 2 results/VE.csv.gz          VE</span>
<span class="co">#&gt; 3 results/VI.csv.gz          VI</span></code></pre></div>
<p>We can also check all the expected <code>outfiles</code> for this particular step, as follows:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">outfiles</a></span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; $gunzip</span>
<span class="co">#&gt; DataFrame with 3 rows and 1 column</span>
<span class="co">#&gt;      gunzip_file</span>
<span class="co">#&gt;      &lt;character&gt;</span>
<span class="co">#&gt; 1 results/SE.csv</span>
<span class="co">#&gt; 2 results/VE.csv</span>
<span class="co">#&gt; 3 results/VI.csv</span></code></pre></div>
<p>Now, we can observe that the third step has been added and contains one substep.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="co">#&gt; Instance of 'SYSargsList': </span>
<span class="co">#&gt;     WF Steps:</span>
<span class="co">#&gt;        1. export_iris --&gt; Status: Pending</span>
<span class="co">#&gt;        2. gzip --&gt; Status: Pending </span>
<span class="co">#&gt;            Total Files: 3 | Existing: 0 | Missing: 3 </span>
<span class="co">#&gt;          2.1. gzip</span>
<span class="co">#&gt;              cmdlist: 3 | Pending: 3</span>
<span class="co">#&gt;        3. gunzip --&gt; Status: Pending </span>
<span class="co">#&gt;            Total Files: 3 | Existing: 0 | Missing: 3 </span>
<span class="co">#&gt;          3.1. gunzip</span>
<span class="co">#&gt;              cmdlist: 3 | Pending: 3</span>
<span class="co">#&gt; </span></code></pre></div>
<p>In addition, we can access all the command lines for each one of the substeps.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">cmdlist</a></span><span class="op">(</span><span class="va">sal</span><span class="op">[</span><span class="st">"gzip"</span><span class="op">]</span>, targets <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; $gzip</span>
<span class="co">#&gt; $gzip$SE</span>
<span class="co">#&gt; $gzip$SE$gzip</span>
<span class="co">#&gt; [1] "gzip -c  results/setosa.csv &gt; results/SE.csv.gz"</span></code></pre></div>
</div>
<div id="getting-data-from-a-workflow-instance" class="section level4">
<h4 class="hasAnchor">
<a href="#getting-data-from-a-workflow-instance" class="anchor"></a>Getting data from a workflow instance</h4>
<p>The final step in this simple workflow is an R code step. For that, we are using the <code>LineWise</code> constructor function as demonstrated above.</p>
<p>One interesting feature showed here is the <code>getColumn</code> method that allows extracting the information for a workflow instance. Those files can be used in an R code, as demonstrated below.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"gunzip"</span>, <span class="st">'outfiles'</span><span class="op">)</span>
<span class="co">#&gt;               SE               VE               VI </span>
<span class="co">#&gt; "results/SE.csv" "results/VE.csv" "results/VI.csv"</span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
                    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"gunzip"</span>, <span class="st">'outfiles'</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">x</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
                    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">df</span><span class="op">)</span>
                    <span class="va">stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, <span class="fl">2</span>, <span class="va">mean</span><span class="op">)</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, <span class="fl">2</span>, <span class="va">sd</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
                    <span class="va">stats</span><span class="op">$</span><span class="va">species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">stats</span><span class="op">)</span>
                    
                    <span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">stats</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">species</span>, y <span class="op">=</span> <span class="va">mean</span>, fill <span class="op">=</span> <span class="va">species</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
                      <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, color <span class="op">=</span> <span class="st">"black"</span>, position <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html">position_dodge</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
                      <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_errorbar</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">mean</span><span class="op">-</span><span class="va">sd</span>, ymax <span class="op">=</span> <span class="va">mean</span><span class="op">+</span><span class="va">sd</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">.2</span>, position <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html">position_dodge</a></span><span class="op">(</span><span class="fl">.9</span><span class="op">)</span><span class="op">)</span> 
                    <span class="op">}</span>,
                    step_name <span class="op">=</span> <span class="st">"iris_stats"</span>, 
                    dependency <span class="op">=</span> <span class="st">"gzip"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="importWF" class="section level3">
<h3 class="hasAnchor">
<a href="#importWF" class="anchor"></a>Build workflow from a {R Markdown}</h3>
<p>The precisely same workflow can be created by importing the steps from an R Markdown file. As demonstrated above, it is required to initialize the project with <code>SPRproject</code> function.</p>
<p><code>importWF</code> function will scan and import all the R chunk from the R Markdown file and build all the workflow instances. Then, each R chuck in the file will be converted in a workflow step.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal_rmd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>logs.dir <span class="op">=</span> <span class="st">".SPRproject_rmd"</span><span class="op">)</span> 
<span class="co">#&gt; Creating directory '/__w/systemPipeWorkshop2021/systemPipeWorkshop2021/vignettes/bioc2021/.SPRproject_rmd'</span>
<span class="co">#&gt; Creating file '/__w/systemPipeWorkshop2021/systemPipeWorkshop2021/vignettes/bioc2021/.SPRproject_rmd/SYSargsList.yml'</span>

<span class="va">sal_rmd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/importWF.html">importWF</a></span><span class="op">(</span><span class="va">sal_rmd</span>, 
                file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"spr_simple_wf.Rmd"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Reading Rmd file</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  ---- Actions ----</span>
<span class="co">#&gt; Checking chunk SPR option</span>
<span class="co">#&gt; Ignore non-SPR chunks: 17</span>
<span class="co">#&gt; Checking chunk eval values</span>
<span class="co">#&gt; Resolve step names</span>
<span class="co">#&gt; Check duplicated step names</span>
<span class="co">#&gt; Checking chunk dependencies</span>
<span class="co">#&gt; Use the previous step as dependency for steps without 'spr.dep' options: 27</span>
<span class="co">#&gt; Parse chunk code</span>
<span class="co">#&gt; ---- Succes! Create output ----</span>
<span class="co">#&gt; Now importing step 'export_iris' </span>
<span class="co">#&gt; Now importing step 'gzip' </span>
<span class="co">#&gt; Now importing step 'gunzip' </span>
<span class="co">#&gt; Now importing step 'stats'</span></code></pre></div>
<p>Let’s explore the workflow to check the steps:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">stepsWF</a></span><span class="op">(</span><span class="va">sal_rmd</span><span class="op">)</span>
<span class="co">#&gt; $export_iris</span>
<span class="co">#&gt; Instance of 'LineWise'</span>
<span class="co">#&gt;     Code Chunk length: 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $gzip</span>
<span class="co">#&gt; Instance of 'SYSargs2':</span>
<span class="co">#&gt;    Slot names/accessors: </span>
<span class="co">#&gt;       targets: 3 (SE...VI), targetsheader: 1 (lines)</span>
<span class="co">#&gt;       modules: 0</span>
<span class="co">#&gt;       wf: 1, clt: 1, yamlinput: 4 (inputs)</span>
<span class="co">#&gt;       input: 3, output: 3</span>
<span class="co">#&gt;       cmdlist: 3</span>
<span class="co">#&gt;    Sub Steps:</span>
<span class="co">#&gt;       1. gzip (rendered: TRUE)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $gunzip</span>
<span class="co">#&gt; Instance of 'SYSargs2':</span>
<span class="co">#&gt;    Slot names/accessors: </span>
<span class="co">#&gt;       targets: 3 (SE...VI), targetsheader: 1 (lines)</span>
<span class="co">#&gt;       modules: 0</span>
<span class="co">#&gt;       wf: 1, clt: 1, yamlinput: 4 (inputs)</span>
<span class="co">#&gt;       input: 3, output: 3</span>
<span class="co">#&gt;       cmdlist: 3</span>
<span class="co">#&gt;    Sub Steps:</span>
<span class="co">#&gt;       1. gunzip (rendered: TRUE)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $stats</span>
<span class="co">#&gt; Instance of 'LineWise'</span>
<span class="co">#&gt;     Code Chunk length: 5</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">dependency</a></span><span class="op">(</span><span class="va">sal_rmd</span><span class="op">)</span>
<span class="co">#&gt; $export_iris</span>
<span class="co">#&gt; [1] ""</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $gzip</span>
<span class="co">#&gt; [1] "export_iris"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $gunzip</span>
<span class="co">#&gt; [1] "gzip"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $stats</span>
<span class="co">#&gt; [1] "gunzip"</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">codeLine</a></span><span class="op">(</span><span class="va">sal_rmd</span><span class="op">)</span>
<span class="co">#&gt; gzip AND gunzip step have been dropped because it is not a LineWise object.</span>
<span class="co">#&gt; export_iris</span>
<span class="co">#&gt;     mapply(function(x, y) write.csv(x, y), split(iris, factor(iris$Species)), file.path("results", paste0(names(split(iris, factor(iris$Species))), ".csv")))</span>
<span class="co">#&gt; stats</span>
<span class="co">#&gt;     df &lt;- lapply(getColumn(sal, step = "gunzip", "outfiles"), function(x) read.delim(x, sep = ",")[-1])</span>
<span class="co">#&gt;     df &lt;- do.call(rbind, df)</span>
<span class="co">#&gt;     stats &lt;- data.frame(cbind(mean = apply(df[, 1:4], 2, mean), sd = apply(df[, 1:4], 2, sd)))</span>
<span class="co">#&gt;     stats$species &lt;- rownames(stats)</span>
<span class="co">#&gt;     plot &lt;- ggplot2::ggplot(stats, ggplot2::aes(x = species, y = mean, fill = species)) + ggplot2::geom_bar(stat = "identity", color = "black", position = ggplot2::position_dodge()) + ggplot2::geom_errorbar(ggplot2::aes(ymin = mean - sd, ymax = mean + sd), width = 0.2, position = ggplot2::position_dodge(0.9))</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">targetsWF</a></span><span class="op">(</span><span class="va">sal_rmd</span><span class="op">)</span>
<span class="co">#&gt; $export_iris</span>
<span class="co">#&gt; DataFrame with 0 rows and 0 columns</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $gzip</span>
<span class="co">#&gt; DataFrame with 3 rows and 2 columns</span>
<span class="co">#&gt;                 FileName  SampleName</span>
<span class="co">#&gt;              &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt; 1     results/setosa.csv          SE</span>
<span class="co">#&gt; 2 results/versicolor.csv          VE</span>
<span class="co">#&gt; 3  results/virginica.csv          VI</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $gunzip</span>
<span class="co">#&gt; DataFrame with 3 rows and 2 columns</span>
<span class="co">#&gt;           gzip_file  SampleName</span>
<span class="co">#&gt;         &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt; 1 results/SE.csv.gz          SE</span>
<span class="co">#&gt; 2 results/VE.csv.gz          VE</span>
<span class="co">#&gt; 3 results/VI.csv.gz          VI</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $stats</span>
<span class="co">#&gt; DataFrame with 0 rows and 0 columns</span></code></pre></div>
<div id="rules-to-create-the-r-markdown-to-import-as-workflow" class="section level4">
<h4 class="hasAnchor">
<a href="#rules-to-create-the-r-markdown-to-import-as-workflow" class="anchor"></a>Rules to create the R Markdown to import as workflow</h4>
<p>To include a particular code chunk from the R Markdown file in the workflow analysis, please use the following code chunk options:</p>
<pre><code>-   `spr='r'`: for code chunks with R code lines;
-   `spr='sysargs'`: for code chunks with an `SYSargsList` object;
-   `spr.dep=&lt;StepName&gt;`: for specify the previous dependency.</code></pre>
<p>For example:</p>
<blockquote>
<p><em>```{r step_1, eval=TRUE, spr=‘r,’ spr.dep=‘step_0’}</em></p>
</blockquote>
<blockquote>
<p><em>```{r step_2, eval=TRUE, spr=‘sysargs,’ spr.dep=‘step_1’}</em></p>
</blockquote>
<p>For <code>spr = 'sysargs'</code>, the last object assigned must to be the <code>SYSargsList</code>, for example:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/example/targets_example.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">HW_mul</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"Example"</span>, 
                      targets <span class="op">=</span> <span class="va">targetspath</span>, 
                      wf_file <span class="op">=</span> <span class="st">"example/example.cwl"</span>, input_file <span class="op">=</span> <span class="st">"example/example.yml"</span>, 
                      dir_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>, 
                      inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>Message <span class="op">=</span> <span class="st">"_STRING_"</span>, SampleName <span class="op">=</span> <span class="st">"_SAMPLE_"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Also, note that all the required files or objects to generate one particular command-line step must be defined in a R code chunk imported. The motivation for this is that when R Markdown files are imported, the <code>spr = 'sysargs'</code> R chunk will be evaluated and stored in the workflow control class as the <code>SYSargsList</code> object, while the R code based (<code>spr = 'r'</code>) is not evaluated, and until the workflow is executed it will be store as an expression.</p>
</div>
</div>
</div>
<div id="running-the-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-workflow" class="anchor"></a>Running the workflow</h2>
<p>For running the workflow, <code>runWF</code> function will execute all the command lines store in the workflow container.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>sal <span class="ot">&lt;-</span> <span class="fu">runWF</span>(sal)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Running Step: export_iris </span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Step Status: Success </span></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Running Step: gzip </span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=======================</span>                                               <span class="er">|</span>  <span class="dv">33</span>%</span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===============================================</span>                       <span class="er">|</span>  <span class="dv">67</span>%</span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span>
<span id="cb39-19"><a href="#cb39-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ---- Summary ---- </span></span>
<span id="cb39-20"><a href="#cb39-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Targets Total_Files Existing_Files Missing_Files    gzip</span></span>
<span id="cb39-21"><a href="#cb39-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1      SE           1              1             0 Success</span></span>
<span id="cb39-22"><a href="#cb39-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2      VE           1              1             0 Success</span></span>
<span id="cb39-23"><a href="#cb39-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3      VI           1              1             0 Success</span></span>
<span id="cb39-24"><a href="#cb39-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb39-25"><a href="#cb39-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Step Status: Success </span></span>
<span id="cb39-26"><a href="#cb39-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Running Step: gunzip </span></span>
<span id="cb39-27"><a href="#cb39-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb39-28"><a href="#cb39-28" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb39-29"><a href="#cb39-29" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb39-30"><a href="#cb39-30" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb39-31"><a href="#cb39-31" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=======================</span>                                               <span class="er">|</span>  <span class="dv">33</span>%</span>
<span id="cb39-32"><a href="#cb39-32" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb39-33"><a href="#cb39-33" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===============================================</span>                       <span class="er">|</span>  <span class="dv">67</span>%</span>
<span id="cb39-34"><a href="#cb39-34" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb39-35"><a href="#cb39-35" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span>
<span id="cb39-36"><a href="#cb39-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ---- Summary ---- </span></span>
<span id="cb39-37"><a href="#cb39-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Targets Total_Files Existing_Files Missing_Files  gunzip</span></span>
<span id="cb39-38"><a href="#cb39-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1      SE           1              1             0 Success</span></span>
<span id="cb39-39"><a href="#cb39-39" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2      VE           1              1             0 Success</span></span>
<span id="cb39-40"><a href="#cb39-40" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3      VI           1              1             0 Success</span></span>
<span id="cb39-41"><a href="#cb39-41" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb39-42"><a href="#cb39-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Step Status: Success </span></span>
<span id="cb39-43"><a href="#cb39-43" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Running Step: iris_stats</span></span></code></pre></div>
<p><img src="systemPipeWorkshop2021_files/figure-html/runWF-1.png" width="700"></p>
<pre><code>#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |======================================================================| 100%
#&gt; Step Status: Success</code></pre>
<p>This essential function allows the user to choose one or multiple steps to be executed using the <code>steps</code> argument. However, it is necessary to follow the workflow dependency graph. If a selected step depends on a previous step(s) that was not executed, the execution will fail.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html">runWF</a></span><span class="op">(</span><span class="va">sal</span>, steps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Also, it allows forcing the execution of the steps, even if the status of the step is <code>'Success'</code> and all the expected <code>outfiles</code> exists. Another feature of the <code>runWF</code> function is ignoring all the warnings and errors and running the workflow by the arguments <code>warning.stop</code> and <code>error.stop</code>, respectively.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html">runWF</a></span><span class="op">(</span><span class="va">sal</span>, force <span class="op">=</span> <span class="cn">TRUE</span>, warning.stop <span class="op">=</span> <span class="cn">FALSE</span>, error.stop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>When the project was initialized by <code>SPRproject</code> function, it was created an environment for all objects created during the workflow execution. This environment can be accessed as follows:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">viewEnvir</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<p>The workflow execution allows to save this environment for future recovery:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html">runWF</a></span><span class="op">(</span><span class="va">sal</span>, saveEnv <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div id="workflow-status" class="section level3">
<h3 class="hasAnchor">
<a href="#workflow-status" class="anchor"></a>Workflow status</h3>
<p>To check the summary of the workflow, we can use:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="co">#&gt; Instance of 'SYSargsList': </span>
<span class="co">#&gt;     WF Steps:</span>
<span class="co">#&gt;        1. export_iris --&gt; Status: Success</span>
<span class="co">#&gt;        2. gzip --&gt; Status: Success </span>
<span class="co">#&gt;            Total Files: 3 | Existing: 3 | Missing: 0 </span>
<span class="co">#&gt;          2.1. gzip</span>
<span class="co">#&gt;              cmdlist: 3 | Success: 3</span>
<span class="co">#&gt;        3. gunzip --&gt; Status: Success </span>
<span class="co">#&gt;            Total Files: 3 | Existing: 3 | Missing: 0 </span>
<span class="co">#&gt;          3.1. gunzip</span>
<span class="co">#&gt;              cmdlist: 3 | Success: 3</span>
<span class="co">#&gt;        4. iris_stats --&gt; Status: Success</span>
<span class="co">#&gt; </span></code></pre></div>
<p>To access more details about the workflow instances, we can use the <code>statusWF</code> method:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">statusWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span>
<span class="co">#&gt; $export_iris</span>
<span class="co">#&gt; DataFrame with 1 row and 2 columns</span>
<span class="co">#&gt;          Step      Status</span>
<span class="co">#&gt;   &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt; 1 export_iris     Success</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $gzip</span>
<span class="co">#&gt; DataFrame with 3 rows and 5 columns</span>
<span class="co">#&gt;       Targets Total_Files Existing_Files Missing_Files    gzip</span>
<span class="co">#&gt;   &lt;character&gt;   &lt;numeric&gt;      &lt;numeric&gt;     &lt;numeric&gt;  &lt;list&gt;</span>
<span class="co">#&gt; 1          SE           1              1             0 Success</span>
<span class="co">#&gt; 2          VE           1              1             0 Success</span>
<span class="co">#&gt; 3          VI           1              1             0 Success</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $gunzip</span>
<span class="co">#&gt; DataFrame with 3 rows and 5 columns</span>
<span class="co">#&gt;       Targets Total_Files Existing_Files Missing_Files  gunzip</span>
<span class="co">#&gt;   &lt;character&gt;   &lt;numeric&gt;      &lt;numeric&gt;     &lt;numeric&gt;  &lt;list&gt;</span>
<span class="co">#&gt; 1          SE           1              1             0 Success</span>
<span class="co">#&gt; 2          VE           1              1             0 Success</span>
<span class="co">#&gt; 3          VI           1              1             0 Success</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $iris_stats</span>
<span class="co">#&gt; DataFrame with 1 row and 2 columns</span>
<span class="co">#&gt;          Step      Status</span>
<span class="co">#&gt;   &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt; 1  iris_stats     Success</span></code></pre></div>
</div>
</div>
<div id="visualize-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-workflow" class="anchor"></a>Visualize workflow</h2>
<p><em><code>systemPipeR</code></em> workflows instances can be visualized with the <code>plotWF</code> function.</p>
<p>This function will make a plot of selected workflow instance and the following information is displayed on the plot:</p>
<pre><code>- Workflow structure (dependency graphs between different steps); 
- Workflow step status, *e.g.* `Success`, `Error`, `Pending`, `Warnings`; 
- Sample status and statistics; 
- Workflow timing: running duration time. </code></pre>
<p>If no argument is provided, the basic plot will automatically detect width, height, layout, plot method, branches, <em>etc</em>.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/plotWF.html">plotWF</a></span><span class="op">(</span><span class="va">sal</span>, show_legend <span class="op">=</span> <span class="cn">TRUE</span>, width <span class="op">=</span> <span class="st">"80%"</span>, rstudio <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-edd098f3220372425dc0" style="width:80%;height:432.632880098888px;" class="plotwf html-widget"></div>
<script type="application/json" data-for="htmlwidget-edd098f3220372425dc0">{"x":{"dot":"digraph {\n    node[fontsize=20];\n    subgraph {\n        node[color=\"dodgerblue\"];\n        export_iris -> gzip -> iris_stats[color=\"dodgerblue\"]\n   }\n    gzip -> gunzip\n    \n    export_iris[label=<<b><font color=\"#5cb85c\">export_iris<\/font><br><\/br><font color=\"#5cb85c\">1<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step export_iris: 1 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2021-08-03 00:23:14; End time: 2021-08-03 00:23:14; Duration: 00:00:00\"]\n    gzip[label=<<b><font color=\"#5cb85c\">gzip<\/font><br><\/br><font color=\"#5cb85c\">3<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">3<\/font><\/b>; <font color=\"black\">0s<\/font>> , style=\"rounded\", shape=\"box\"  tooltip=\"step gzip: 3 samples passed; 0 samples have warnings; 0 samples have errors; 3 samples in total; Start time: 2021-08-03 00:23:14; End time: 2021-08-03 00:23:14; Duration: 00:00:00\"]\n    gunzip[label=<<b><font color=\"#5cb85c\">gunzip<\/font><br><\/br><font color=\"#5cb85c\">3<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">3<\/font><\/b>; <font color=\"black\">0s<\/font>> , style=\"rounded\", shape=\"box\"  tooltip=\"step gunzip: 3 samples passed; 0 samples have warnings; 0 samples have errors; 3 samples in total; Start time: 2021-08-03 00:23:15; End time: 2021-08-03 00:23:15; Duration: 00:00:00\"]\n    iris_stats[label=<<b><font color=\"#5cb85c\">iris_stats<\/font><br><\/br><font color=\"#5cb85c\">1<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0.3s<\/font>>  tooltip=\"step iris_stats: 1 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2021-08-03 00:23:15; End time: 2021-08-03 00:23:15; Duration: 00:00:0.3\"]\n    subgraph cluster_legend {\n        rankdir=TB;\n        color=\"#EEEEEE\";\n        style=filled;\n        node [style=filled];\n        {rank=same; R_step; Sysargs_step; Main_branch}\n        Main_branch -> Sysargs_step -> R_step[color=\"#EEEEEE\"]\n        Main_branch[label=\"Main branch\" color=\"dodgerblue\", style=\"filled\", fillcolor=white];   Sysargs_step ->step_state[color=\"#EEEEEE\"];\n        step_state[style=\"filled\", shape=\"box\" color=white, label =<\n            <table>\n            <tr><td><b>Step Colors<\/b><\/td><\/tr>\n            <tr><td><font color=\"black\">Pending steps<\/font>; <font color=\"#5cb85c\">Successful steps<\/font>; <font color=\"#d9534f\">Failed steps<\/font><\/td><\/tr>\n            <tr><td><b>Targets Files / Code Chunk <\/b><\/td><\/tr><tr><td><font color=\"#5cb85c\">0 (pass) <\/font> | <font color=\"#f0ad4e\">0 (warning) <\/font> | <font color=\"#d9534f\">0 (error) <\/font> | <font color=\"blue\">0 (total)<\/font>; Duration<\/td><\/tr><\/table>\n            >];\n        label=\"Legends\";\n        fontsize = 30;\n        Sysargs_step[label=\"Sysargs step\" style=\"rounded, filled\", shape=\"box\", fillcolor=white];\n        R_step[label=\"R step\" style=\"rounded, filled\", fillcolor=white];\n    }\n\n}\n","plotid":"sprwf-56217834","responsive":true,"width":"80%","height":null,"plot_method":"renderSVGElement","rmd":true,"msg":""},"evals":[],"jsHooks":[]}</script><p>For more details about the <code>plotWF</code> function, please see <a href="#plotWF">here</a>.</p>
</div>
<div id="generation-of-technical-report" class="section level2">
<h2 class="hasAnchor">
<a href="#generation-of-technical-report" class="anchor"></a>Generation of technical report</h2>
<p><em><code>systemPipeR</code></em> compiles all the workflow execution logs in one central location, making it easier to check any standard output (<code>stdout</code>) or standard error (<code>stderr</code>) for any command-line tools used on the workflow or the R code <code>stdout</code>. Also, the workflow plot is appended at the beginning of the report, making it easier to click on the respective step.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderLogs.html">renderLogs</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
<div id="parallelization-on-clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#parallelization-on-clusters" class="anchor"></a>Parallelization on clusters</h2>
<p>This section of the tutorial provides an introduction to the usage of the <em><code>systemPipeR</code></em> features on a cluster.</p>
<p>The computation can be greatly accelerated by processing many files in parallel using several compute nodes of a cluster, where a scheduling/queuing system is used for load balancing. For this the <code>clusterRun</code> function submits the computing requests to the scheduler using the run specifications defined by <code>runWF</code>.</p>
<p>A named list provides the computational resources. By default, it can be defined the upper time limit in minutes for jobs before they get killed by the scheduler, memory limit in Mb, number of <code>CPUs</code>, and number of tasks.</p>
<p>The number of independent parallel cluster processes is defined under the <code>Njobs</code> argument. The following example will run one process in parallel using for each 4 CPU cores. If the resources available on a cluster allow running all the processes simultaneously, then the shown sample submission will utilize in total four CPU cores (<code>NJobs * ncpus</code>). Note, <code>clusterRun</code> can be used with most queueing systems as it is based on utilities from the <em><code>batchtools</code></em> package which supports the use of template files (<em><code>*.tmpl</code></em>) for defining the run parameters of different schedulers. To run the following code, one needs to have both a <code>conf file</code> (see <em><code>.batchtools.conf.R</code></em> samples <a href="https://mllg.github.io/batchtools/">here</a>) and a template file (see <em><code>*.tmpl</code></em> samples <a href="https://github.com/mllg/batchtools/tree/master/inst/templates">here</a>) for the queueing available on a system. The following example uses the sample <code>conf</code> and <code>template</code> files for the <code>Slurm</code> scheduler provided by this package.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mllg/batchtools">batchtools</a></span><span class="op">)</span>
<span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime<span class="op">=</span><span class="fl">120</span>, ntasks<span class="op">=</span><span class="fl">1</span>, ncpus<span class="op">=</span><span class="fl">4</span>, memory<span class="op">=</span><span class="fl">1024</span><span class="op">)</span>
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">sal</span>, FUN <span class="op">=</span> <span class="va">runWF</span>, 
                  more.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>,
                  conffile<span class="op">=</span><span class="st">".batchtools.conf.R"</span>, 
                  template<span class="op">=</span><span class="st">"batchtools.slurm.tmpl"</span>,
                  Njobs<span class="op">=</span><span class="fl">1</span>, runid<span class="op">=</span><span class="st">"01"</span>, resourceList<span class="op">=</span><span class="va">resources</span><span class="op">)</span></code></pre></div>
<p>Note: The example is submitting the jog to <code>short</code> partition. If you desire to use a different partition, please adjust accordingly (<code>batchtools.slurm.tmpl</code>).</p>
</div>
<div id="exported-the-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#exported-the-workflow" class="anchor"></a>Exported the workflow</h2>
<p><em><code>systemPipeR</code></em> workflow management system allows to translate and export the workflow build interactively to R Markdown format or an executable bash script. This feature advances the reusability of the workflow, as well as the flexibility for workflow execution.</p>
<div id="r-markdown-file" class="section level3">
<h3 class="hasAnchor">
<a href="#r-markdown-file" class="anchor"></a>R Markdown file</h3>
<p><code>sal2rmd</code> function takes an <code>SYSargsList</code> workflow container and translates it to SPR workflow template R markdown format. This file can be imported with the <code>importWF</code> function, as demonstrated above.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sal2rmd.html">sal2rmd</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
<div id="bash-script" class="section level3">
<h3 class="hasAnchor">
<a href="#bash-script" class="anchor"></a>Bash script</h3>
<p><code>sal2bash</code> function takes an <code>SYSargsList</code> workflow container and translates it to an executable bash script, so one can run the workflow without loading <code>SPR</code> or using an R console.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sal2bash.html">sal2bash</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<p>It will be generated on the project root an executable bash script, called by default the <code>spr_wf.sh</code>. Also, a directory <code>./spr_wf</code> will be created and store all the R scripts based on the workflow steps. Please note that this function will “collapse” adjacent R steps into one file as much as possible.</p>
</div>
</div>
<div id="project-resume-and-restart" class="section level2">
<h2 class="hasAnchor">
<a href="#project-resume-and-restart" class="anchor"></a>Project Resume and Restart</h2>
<p>If you desire to resume or restart a project that has been initialized in the past, <code>SPRproject</code> function allows this operation.</p>
<p>With the resume option, it is possible to load the <code>SYSargsList</code> object in R and resume the analysis. Please, make sure to provide the <code>logs.dir</code> location, and the corresponded <code>YAML</code> file name. The current working directory needs to be in the project root directory.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>resume <span class="op">=</span> <span class="cn">TRUE</span>, logs.dir <span class="op">=</span> <span class="st">".SPRproject"</span>, 
                  sys.file <span class="op">=</span> <span class="st">".SPRproject/SYSargsList.yml"</span><span class="op">)</span> </code></pre></div>
<p>If you choose to save the environment in the last analysis, you can recover all the files created in that particular section. <code>SPRproject</code> function allows this with <code>load.envir</code> argument. Please note that the environment was saved only with you run the workflow in the last section (<code><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html">runWF()</a></code>).</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>resume <span class="op">=</span> <span class="cn">TRUE</span>, load.envir <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </code></pre></div>
<p>After loading the workflow at your current section, you can check the objects created in the old environment and decide if it is necessary to copy them to the current environment.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">viewEnvir</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html">copyEnvir</a></span><span class="op">(</span><span class="va">sal</span>, list<span class="op">=</span><span class="st">"plot"</span>, new.env <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/environment.html">globalenv</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>This option will keep all previous logs in the folder; however, if you desire to clean the execution history and restart the workflow, the <code>restart=TRUE</code> option can be used.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>restart <span class="op">=</span> <span class="cn">TRUE</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span>, load.envir <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </code></pre></div>
<p>The last and more drastic option from <code>SYSproject</code> function is to overwrite the logs and the workflow. This option will delete the hidden folder and the information on the <code>SYSargsList.yml</code> files. This will not delete any parameter file nor any results it was created in previous runs. Please use with caution.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span>overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </code></pre></div>
</div>
<div id="showcase-rna-seq-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#showcase-rna-seq-workflow" class="anchor"></a>Showcase RNA-Seq workflow</h2>
<p>If you desire to import and run a RNA-seq pipeline, please follow:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">systemPipeRdata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeRdata/man/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"rnaseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"rnaseq"</span><span class="op">)</span>
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html">SPRproject</a></span><span class="op">(</span><span class="op">)</span> 
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/importWF.html">importWF</a></span><span class="op">(</span><span class="va">sal</span>, file_path <span class="op">=</span> <span class="st">"systemPipeRNAseq_importWF.Rmd"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html">runWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/plotWF.html">plotWF</a></span><span class="op">(</span><span class="va">sal</span>, rstudio <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderLogs.html">renderLogs</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
<div id="overview-of-systempipeshiny-core-functionalities" class="section level2">
<h2 class="hasAnchor">
<a href="#overview-of-systempipeshiny-core-functionalities" class="anchor"></a>Overview of <em>systemPipeShiny</em> core functionalities</h2>
<p><strong><span style="color:#5DA7D6;">s</span>ystem<span style="color:#5DA7D6;">P</span>ipe<span style="color:#5DA7D6;">S</span>hiny</strong> (SPS) extends the widely used <a href="/spr/">systemPipeR</a> (SPR) workflow environment with a versatile graphical user interface provided by a <a href="https://shiny.rstudio.com">Shiny App</a>. This allows non-R users, such as experimentalists, to run many systemPipeR’s workflow designs, control, and visualization functionalities interactively without requiring knowledge of R. Most importantly, <code>SPS</code> has been designed as a general purpose framework for interacting with other R packages in an intuitive manner. Like most Shiny Apps, SPS can be used on both local computers as well as centralized server-based deployments that can be accessed remotely as a public web service for using SPR’s functionalities with community and/or private data. The framework can integrate many core packages from the R/Bioconductor ecosystem. Examples of SPS’ current functionalities include:</p>
<ul>
<li>A default interactive workflow module to create experimental designs, visualize and customize workflow topologies with previews, and programming free workflow execution within the application.</li>
<li>An interactive module with extensive plot options to visualize downstream analysis of a RNA-Seq workflow.</li>
<li>A quick ggplot module to make all variety of scientific plots from any user defined tabular data.</li>
<li>An extendable set of visualization functionalities makes it easy to design custom Shiny Apps under SPS framework without any knowledge of Shiny.</li>
<li>A ‘Canvas Workbench’ to manage complex visual results. It allows users to organize and to compare plots in an efficient manner combined with a session screenshot feature to edit scientific and publishable figures.</li>
<li>Three other supporting packages to help all users from beginners and advanced developers to extend under current SPS framework or on their own visualization apps.</li>
</ul>
</div>
<div id="demo" class="section level2">
<h2 class="hasAnchor">
<a href="#demo" class="anchor"></a>Demo</h2>
<p>View our online demo app:</p>
<table class="table">
<colgroup>
<col width="33%">
<col width="33%">
<col width="33%">
</colgroup>
<thead><tr class="header">
<th>Type and link</th>
<th>option changed</th>
<th>notes</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny/">Default full installation{blk}</a></td>
<td><a href="#installation">See installation</a></td>
<td>full app</td>
</tr>
<tr class="even">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny/">Minimum installation{blk}</a></td>
<td><a href="#installation">See installation</a></td>
<td>no modules installed</td>
</tr>
<tr class="odd">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny_loading/">Login enabled{blk}</a></td>
<td><code>login_screen = TRUE; login_theme = "empty"</code></td>
<td>no modules installed</td>
</tr>
<tr class="even">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny_loading_theme/">Login and login themes{blk}</a></td>
<td><code>login_screen = TRUE; login_theme = "random"</code></td>
<td>no modules installed</td>
</tr>
<tr class="odd">
<td><a href="https://tgirke.shinyapps.io/systemPipeShiny_loading/?admin">App admin page{blk}</a></td>
<td><code>admin_page = TRUE</code></td>
<td>or simply add “?admin” to the end of URL of demos</td>
</tr>
</tbody>
</table>
<p>For the login required demos, the app account name is <strong>“user”</strong> password <strong>“user”</strong>.</p>
<p>For the admin panel login, account name <strong>“admin”</strong>, password <strong>“admin”</strong>.</p>
<p><strong>Please DO NOT delete or change password when you are using the admin features.</strong> <em>shinyapp.io</em> will reset the app once a while, but this will affect other people who are trying the demo simultaneously.</p>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Amstutz2016-ka" class="csl-entry">
Amstutz, Peter, Michael R Crusoe, Nebojša Tijanić, Brad Chapman, John Chilton, Michael Heuer, Andrey Kartashov, et al. 2016. <span>“Common Workflow Language, V1.0,”</span> July. <a href="https://doi.org/10.6084/m9.figshare.3115156.v2">https://doi.org/10.6084/m9.figshare.3115156.v2</a>.
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Daniela Cassol, Le Zhang, Thomas Girke.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
